<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Data Science for Biology Workshop Series - 16S Data Analysis Part I</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../">
<script src="../../../site_libs/quarto-html/quarto.js"></script>
<script src="../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../../index.html">
    <span class="navbar-title">Data Science for Biology Workshop Series</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../../instance_ips.html"> 
<span class="menu-text">AWS Instance IPs</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../installation-instructions.html"> 
<span class="menu-text">Installation</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../datasets.html"> 
<span class="menu-text">Datasets</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../resources.html"> 
<span class="menu-text">Resources</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../past_workshops.html"> 
<span class="menu-text">Past Workshops</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/elsherbini/durban-data-science-for-biology"> <i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <a class="flex-grow-1 no-decor" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
          <h1 class="quarto-secondary-nav-title">16S Data Analysis Part I</h1>
        </a>     
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false">
 <span class="menu-text"><em>1. Reproducible Data Analysis with R</em></span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/1-workshop1/0-welcome/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 0:</b> Welcome to the workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/1-workshop1/1-data-generation/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 1</b>: How data analysis is informed by data generation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/1-workshop1/2-intro-to-r/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 2</b>: Intro to R, Rstudio, and Quarto</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/1-workshop1/3-tidyverse-101/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 3</b>: Intro to data visualization and data wrangling with the <code>tidyverse</code></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/1-workshop1/4-tidyverse-201/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 4</b>: Data wrangling and more data visualization</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/1-workshop1/5-tableone/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 5</b>: <code>tableone</code> and its Basic Statistics</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/1-workshop1/6-hypothesis-testing/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 6</b>: Data transformations and more statistics</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/1-workshop1/7-custom-data-visualizations/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 7</b>: Customizing data visualizations using <code>colorspace</code>, <code>ggplot2</code>, and <code>patchwork</code></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/1-workshop1/8-group-projects/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Group Project</b> -Applying what you’ve learned to new data</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false">
 <span class="menu-text"><em>2. Working with High Dimensional Data in R - The Human Microbiome</em></span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/2-workshop2/0-welcome/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 0:</b> Welcome to the workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/2-workshop2/1-library-prep/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 1</b>: Sequencing Library Preparation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/2-workshop2/2-r-refresher/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 2</b>: R Refresher</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/2-workshop2/3-16s_processing/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 3</b>: 16S Data Processing</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/2-workshop2/4-16s-pt1/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 4</b>: Exploring High Dimensional Data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/2-workshop2/5-16s-pt2/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 5</b>: Correlations and Ordinations</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/2-workshop2/6-virgo/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 6</b>: Shotgun Metagenomics and VIRGO</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/2-workshop2/7-keynote/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 7</b>: Keynote</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/2-workshop2/8-group-project/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Group Activity:</b> Applying what you’ve learned to new data</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false">
 <span class="menu-text"><b>3. Bioinformatics and Viral Genomics - February 22-28 2025</b></span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/3-workshop3/0-welcome/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 0:</b> Welcome to the workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/3-workshop3/1-viral-sequencing/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 1</b>: Preparing libraries for viromics</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/3-workshop3/2-intro-to-unix/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 2</b>: Intro to the Unix Shell</span></a>
  </div>
</li>
          <li class="sidebar-item">
 <span class="menu-text"><b>Module 3</b>:Intro to Bioinformatics I</span>
  </li>
          <li class="sidebar-item">
 <span class="menu-text"><b>Module 4</b>:Intro to Bioinformatics II</span>
  </li>
          <li class="sidebar-item">
 <span class="menu-text"><b>Module 5</b>:Targetted Viromics and Phylogenetics</span>
  </li>
          <li class="sidebar-item">
 <span class="menu-text"><b>Interactive 1</b>:Genomics Adventure</span>
  </li>
          <li class="sidebar-item">
 <span class="menu-text"><b>Module 6</b>:Untargetted Viromics</span>
  </li>
          <li class="sidebar-item">
 <span class="menu-text"><b>Interactive 2</b>:Genomics Adventure</span>
  </li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="99">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#eda-mindset" id="toc-eda-mindset" class="nav-link active" data-scroll-target="#eda-mindset">EDA mindset</a></li>
  <li><a href="#get-set-up" id="toc-get-set-up" class="nav-link" data-scroll-target="#get-set-up">0. Get set up</a>
  <ul class="collapse">
  <li><a href="#load-packages" id="toc-load-packages" class="nav-link" data-scroll-target="#load-packages">Load packages</a></li>
  <li><a href="#manage-file-paths" id="toc-manage-file-paths" class="nav-link" data-scroll-target="#manage-file-paths">Manage file paths</a></li>
  <li><a href="#load-instructional-data" id="toc-load-instructional-data" class="nav-link" data-scroll-target="#load-instructional-data">Load instructional data</a></li>
  </ul></li>
  <li><a href="#get-to-know-the-data" id="toc-get-to-know-the-data" class="nav-link" data-scroll-target="#get-to-know-the-data">1. Get to know the data</a>
  <ul class="collapse">
  <li><a href="#sample-data" id="toc-sample-data" class="nav-link" data-scroll-target="#sample-data">Sample data</a></li>
  <li><a href="#count-table" id="toc-count-table" class="nav-link" data-scroll-target="#count-table">Count table</a></li>
  <li><a href="#taxonomy-table" id="toc-taxonomy-table" class="nav-link" data-scroll-target="#taxonomy-table">Taxonomy table</a></li>
  </ul></li>
  <li><a href="#make-phyloseq-object" id="toc-make-phyloseq-object" class="nav-link" data-scroll-target="#make-phyloseq-object">2. Make phyloseq object</a>
  <ul class="collapse">
  <li><a href="#accessor-functions" id="toc-accessor-functions" class="nav-link" data-scroll-target="#accessor-functions">Accessor functions</a></li>
  <li><a href="#processor-functions" id="toc-processor-functions" class="nav-link" data-scroll-target="#processor-functions">Processor functions</a></li>
  </ul></li>
  <li><a href="#plot-bars" id="toc-plot-bars" class="nav-link" data-scroll-target="#plot-bars">3. Plot bars</a>
  <ul class="collapse">
  <li><a href="#zymo-mocks" id="toc-zymo-mocks" class="nav-link" data-scroll-target="#zymo-mocks">Zymo mocks</a></li>
  <li><a href="#study-samples" id="toc-study-samples" class="nav-link" data-scroll-target="#study-samples">Study samples</a></li>
  </ul></li>
  <li><a href="#alpha-diversity" id="toc-alpha-diversity" class="nav-link" data-scroll-target="#alpha-diversity">4. Alpha diversity</a>
  <ul class="collapse">
  <li><a href="#zymo-mocks-1" id="toc-zymo-mocks-1" class="nav-link" data-scroll-target="#zymo-mocks-1">Zymo mocks</a></li>
  <li><a href="#study-samples-1" id="toc-study-samples-1" class="nav-link" data-scroll-target="#study-samples-1">Study samples</a></li>
  </ul></li>
  <li><a href="#beta-diversity" id="toc-beta-diversity" class="nav-link" data-scroll-target="#beta-diversity">5. Beta diversity</a>
  <ul class="collapse">
  <li><a href="#zymo-mocks-2" id="toc-zymo-mocks-2" class="nav-link" data-scroll-target="#zymo-mocks-2">Zymo mocks</a></li>
  <li><a href="#study-samples-2" id="toc-study-samples-2" class="nav-link" data-scroll-target="#study-samples-2">Study samples</a></li>
  </ul></li>
  <li><a href="#lets-practice" id="toc-lets-practice" class="nav-link" data-scroll-target="#lets-practice">6. Let’s practice!</a>
  <ul class="collapse">
  <li><a href="#manage-paths" id="toc-manage-paths" class="nav-link" data-scroll-target="#manage-paths">Manage paths</a></li>
  <li><a href="#load-data" id="toc-load-data" class="nav-link" data-scroll-target="#load-data">Load data</a></li>
  <li><a href="#dairy-dataset" id="toc-dairy-dataset" class="nav-link" data-scroll-target="#dairy-dataset">Dairy dataset</a></li>
  <li><a href="#delivery-dataset" id="toc-delivery-dataset" class="nav-link" data-scroll-target="#delivery-dataset">Delivery dataset</a></li>
  </ul></li>
  <li><a href="#reproducibility-receipt" id="toc-reproducibility-receipt" class="nav-link" data-scroll-target="#reproducibility-receipt">7. Reproducibility receipt</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">16S Data Analysis Part I</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>We’ve inferred and classified amplicon sequence variants (ASVs) using the <code>dada2</code> package.</p>
<p>Now let’s do some exploratory data analysis (EDA) using the <code>phyloseq</code> package and other tools.</p>
<section id="eda-mindset" class="level2">
<h2 class="anchored" data-anchor-id="eda-mindset">EDA mindset</h2>
<p>Exploratory mindset:</p>
<ul>
<li>“I’m very curious about my data”</li>
<li>“It’s unlikely that my data are perfect”</li>
<li>“I’d like to familiarize myself with the data”</li>
</ul>
</section>
<section id="get-set-up" class="level2">
<h2 class="anchored" data-anchor-id="get-set-up">0. Get set up</h2>
<section id="load-packages" class="level3">
<h3 class="anchored" data-anchor-id="load-packages">Load packages</h3>
<p>Load some packages including the <code>phyloseq</code> package</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phyloseq)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rstatix)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pander)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To get Help with any function, type a question mark (?) in front of the function name at the Console. For example, type <code>?dplyr::filter</code> (or <code>?filter</code>) and hit enter to get Help with the <code>dplyr</code> function <code>filter()</code>.</p>
</section>
<section id="manage-file-paths" class="level3">
<h3 class="anchored" data-anchor-id="manage-file-paths">Manage file paths</h3>
<p>Show path to the current project</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>here<span class="sc">::</span><span class="fu">here</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "/Users/jelsherbini/dev/durban-data-science-for-biology"</code></pre>
</div>
</div>
<p>Show paths to the data we intend to analyze</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>fs<span class="sc">::</span><span class="fu">dir_ls</span>(<span class="fu">here</span>(<span class="st">"materials/2-workshop2/4-16s-pt1/data"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>/Users/jelsherbini/dev/durban-data-science-for-biology/materials/2-workshop2/4-16s-pt1/data/instructional
/Users/jelsherbini/dev/durban-data-science-for-biology/materials/2-workshop2/4-16s-pt1/data/practice</code></pre>
</div>
</div>
<p>Create an object containing the path to the instructional data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>in_path <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="fu">here</span>(<span class="st">"materials/2-workshop2/4-16s-pt1/data/instructional/"</span>))</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>in_path</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "/Users/jelsherbini/dev/durban-data-science-for-biology/materials/2-workshop2/4-16s-pt1/data/instructional/"</code></pre>
</div>
</div>
<p>Just for fun, let’s be curious about this object <code>in_path</code>. What type of object is it? How many strings does it contain? How many characters in this string?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(in_path)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "character"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(in_path)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nchar</span>(in_path)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 106</code></pre>
</div>
</div>
<p>We can answer these questions. It’s a character vector containing 1 string consisting of 106 characters.</p>
<p>We’ll use this object to construct commands for loading our data.</p>
</section>
<section id="load-instructional-data" class="level3">
<h3 class="anchored" data-anchor-id="load-instructional-data">Load instructional data</h3>
<p>List the files we intend to load</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(in_path)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "count_table_instructional.rds" "sample_data_instructional.rds"
[3] "tax_table_instructional.rds"  </code></pre>
</div>
</div>
<p>The count_table and tax_table are the outputs of <code>dada2</code>. The sample_data is a version of the file <code>05_amplicon_sample_ids_UKZN_workshop_2023.csv</code> that I’ve modified for use here.</p>
<p>Note: An <code>.rds</code> file contains a single R object which has been saved to a file. The file can be loaded back into the R environment using the <code>readRDS()</code> function.</p>
<p>Let’s load the instructional data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>sample_df <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(in_path, <span class="st">"sample_data_instructional.rds"</span>))</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>count_tab <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(in_path, <span class="st">"count_table_instructional.rds"</span>))</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>tax_tab <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(in_path, <span class="st">"tax_table_instructional.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note: The <code>paste0()</code> function is a variation of <code>paste()</code> that concatenates strings without any separator.</p>
</section>
</section>
<section id="get-to-know-the-data" class="level2">
<h2 class="anchored" data-anchor-id="get-to-know-the-data">1. Get to know the data</h2>
<p>Familiarize yourself with the data:</p>
<ul>
<li>Think about how you would describe or summarize it</li>
<li>Examine the objects for form, size, and content</li>
<li>Review and visualize your study design</li>
<li>Identify any missing data</li>
</ul>
<section id="sample-data" class="level3">
<h3 class="anchored" data-anchor-id="sample-data">Sample data</h3>
<p>Let’s begin with the sample data.</p>
<p>Glimpse the object using <code>head()</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>sample_df <span class="sc">%&gt;%</span> </span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>(<span class="at">n =</span> <span class="dv">3</span>) <span class="sc">%&gt;%</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<caption>Table continues below</caption>
<colgroup>
<col style="width: 28%">
<col style="width: 21%">
<col style="width: 22%">
<col style="width: 16%">
<col style="width: 12%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">amplicon_sample_id</th>
<th style="text-align: center;">amplicon_type</th>
<th style="text-align: center;">instrument_run</th>
<th style="text-align: center;">reads_out</th>
<th style="text-align: center;">pid</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">GUT0003</td>
<td style="text-align: center;">study_sample</td>
<td style="text-align: center;">run1</td>
<td style="text-align: center;">112769</td>
<td style="text-align: center;">pid_01</td>
</tr>
<tr class="even">
<td style="text-align: center;">VAG0142</td>
<td style="text-align: center;">study_sample</td>
<td style="text-align: center;">run2</td>
<td style="text-align: center;">73426</td>
<td style="text-align: center;">pid_01</td>
</tr>
<tr class="odd">
<td style="text-align: center;">GUT0050</td>
<td style="text-align: center;">study_sample</td>
<td style="text-align: center;">run1</td>
<td style="text-align: center;">20211</td>
<td style="text-align: center;">pid_01</td>
</tr>
</tbody>
</table>
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 13%">
<col style="width: 18%">
<col style="width: 19%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">arm</th>
<th style="text-align: center;">time_point</th>
<th style="text-align: center;">sample_type</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">baseline</td>
<td style="text-align: center;">gut_biopsy</td>
</tr>
<tr class="even">
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">baseline</td>
<td style="text-align: center;">vaginal</td>
</tr>
<tr class="odd">
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">week_1</td>
<td style="text-align: center;">gut_biopsy</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Glimpse the object using <code>tail()</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>sample_df <span class="sc">%&gt;%</span> </span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tail</span>(<span class="at">n =</span> <span class="dv">3</span>) <span class="sc">%&gt;%</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<caption>Table continues below</caption>
<colgroup>
<col style="width: 12%">
<col style="width: 30%">
<col style="width: 20%">
<col style="width: 21%">
<col style="width: 15%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">&nbsp;</th>
<th style="text-align: center;">amplicon_sample_id</th>
<th style="text-align: center;">amplicon_type</th>
<th style="text-align: center;">instrument_run</th>
<th style="text-align: center;">reads_out</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;"><strong>272</strong></td>
<td style="text-align: center;">ZYMO.DNA.CTRL.P6.run2</td>
<td style="text-align: center;">zymo_mock</td>
<td style="text-align: center;">run2</td>
<td style="text-align: center;">41915</td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>273</strong></td>
<td style="text-align: center;">ZYMO.DNA.CTRL.P7.run2</td>
<td style="text-align: center;">zymo_mock</td>
<td style="text-align: center;">run2</td>
<td style="text-align: center;">48564</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><strong>274</strong></td>
<td style="text-align: center;">ZYMO.DNA.CTRL.P8.run2</td>
<td style="text-align: center;">zymo_mock</td>
<td style="text-align: center;">run2</td>
<td style="text-align: center;">38318</td>
</tr>
</tbody>
</table>
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 13%">
<col style="width: 8%">
<col style="width: 8%">
<col style="width: 18%">
<col style="width: 19%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">&nbsp;</th>
<th style="text-align: center;">pid</th>
<th style="text-align: center;">arm</th>
<th style="text-align: center;">time_point</th>
<th style="text-align: center;">sample_type</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;"><strong>272</strong></td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">NA</td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>273</strong></td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">NA</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><strong>274</strong></td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">NA</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Does this object contain any missing data? How much?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">is.na</span>(sample_df))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 40</code></pre>
</div>
</div>
<p>Yes it does.</p>
<p>This object is a data.frame with 274 rows (observations) and 8 columns (variables). We acknowledge the presence of NA values (missing data).</p>
<p>Note: I’ve added a variable <code>sample_df$reads_out</code> containing the total number of reads output by <code>dada2</code>. This variable can be taken from the portion of the <code>dada2</code> workflow in which reads are tracked through the pipeline.</p>
<p>In our sample data, we see variables describing the amplicons. How many amplicons did we sequence on each run?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>sample_df <span class="sc">%&gt;%</span> </span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(instrument_run, amplicon_type) <span class="sc">%&gt;%</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 23%">
<col style="width: 22%">
<col style="width: 8%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">instrument_run</th>
<th style="text-align: center;">amplicon_type</th>
<th style="text-align: center;">n</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">run1</td>
<td style="text-align: center;">study_sample</td>
<td style="text-align: center;">132</td>
</tr>
<tr class="even">
<td style="text-align: center;">run1</td>
<td style="text-align: center;">zymo_mock</td>
<td style="text-align: center;">5</td>
</tr>
<tr class="odd">
<td style="text-align: center;">run2</td>
<td style="text-align: center;">study_sample</td>
<td style="text-align: center;">132</td>
</tr>
<tr class="even">
<td style="text-align: center;">run2</td>
<td style="text-align: center;">zymo_mock</td>
<td style="text-align: center;">5</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>On each instrument run, we sequenced <em>n</em> = 132 study samples and <em>n</em> = 5 mock communities (positive controls).</p>
<p>Here, the mock communities are the <a href="https://www.zymoresearch.com/collections/zymobiomics-microbial-community-standards/products/zymobiomics-microbial-community-dna-standard">ZymoBIOMICS Microbial Community DNA Standard</a>. Later, we’ll use these mocks to assess for batch effects (run1 vs run2).</p>
<p>We also have variables pertaining to the study design. Let’s report the number of participants in the study <em>without</em> counting the NA values associated with the mocks.</p>
<p>Here are two options</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(sample_df<span class="sc">$</span>pid)[<span class="sc">!</span><span class="fu">is.na</span>(<span class="fu">unique</span>(sample_df<span class="sc">$</span>pid))]) <span class="co"># Option 1 using base R</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 44</code></pre>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>sample_df <span class="sc">%&gt;%</span> <span class="fu">drop_na</span>() <span class="sc">%$%</span> <span class="fu">n_distinct</span>(pid) <span class="co"># Option 2 using tidyverse &amp; magrittr</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 44</code></pre>
</div>
</div>
<p>Which option do you find more readable?</p>
<p>Note: The <code>%$%</code> operator is known as the “exposition pipe operator”. It’s part of the <code>magrittr</code> package and is used to expose the variables in a dataframe to the environment in an expression.</p>
<p>Now let’s review the study design. Often, a visualization works well for this purpose.</p>
<p>Plot our study design</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>sample_df <span class="sc">%&gt;%</span> </span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">drop_na</span>() <span class="sc">%&gt;%</span> <span class="co"># Without this line, what happens? Why? </span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">has_reads =</span> reads_out <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> time_point, <span class="at">y =</span> pid, <span class="at">color =</span> has_reads)) <span class="sc">+</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(<span class="at">rows =</span> <span class="fu">vars</span>(arm), <span class="at">cols =</span> <span class="fu">vars</span>(sample_type), <span class="at">scales =</span> <span class="st">"free_y"</span>) <span class="sc">+</span></span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"study design"</span>, <span class="at">y =</span> <span class="st">"participant_id"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We recall that our study involves 44 participants, randomized to placebo or treatment, who contributed a gut and vaginal sample at baseline (before intervention) and at 1 and 7 weeks after intervention.</p>
<p>We also note that three samples failed to return any reads after processing through <code>dada2</code>. (What happened?)</p>
<p>Let’s make a plot that looks closely at our sequencing yields</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>sample_df <span class="sc">%&gt;%</span> </span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(reads_out <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">sample_type =</span> <span class="fu">ifelse</span>(<span class="fu">is.na</span>(sample_type), <span class="st">"zymo"</span>, sample_type)) <span class="sc">%&gt;%</span> </span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> reads_out, <span class="at">fill =</span> sample_type)) <span class="sc">+</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(.<span class="sc">~</span>instrument_run) <span class="sc">+</span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"sequencing yield"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>These histograms have very different shapes. If this is surprising or concerning to us, we might investigate why this happened. It’s also clear that we have additional samples with relatively low yield. How would you quickly modify the code above to display only those samples with fewer than 500 reads?</p>
<p>When our work involves batches, we always want to include replicate samples (here, the zymo mocks) that allow us to assess for batch effects. Looking at the plot above, why was this extra important in the context of this study?</p>
<p>Finally, we can compute some summary statistics for this variable <code>sample_df$reads_out</code> for each instrument run</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>sample_df <span class="sc">%&gt;%</span> </span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(instrument_run) <span class="sc">%&gt;%</span> </span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">get_summary_stats</span>(reads_out, <span class="at">type =</span> <span class="st">"five_number"</span>) <span class="sc">%&gt;%</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 22%">
<col style="width: 15%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 11%">
<col style="width: 10%">
<col style="width: 11%">
<col style="width: 11%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">instrument_run</th>
<th style="text-align: center;">variable</th>
<th style="text-align: center;">n</th>
<th style="text-align: center;">min</th>
<th style="text-align: center;">max</th>
<th style="text-align: center;">q1</th>
<th style="text-align: center;">median</th>
<th style="text-align: center;">q3</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">run1</td>
<td style="text-align: center;">reads_out</td>
<td style="text-align: center;">137</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">257270</td>
<td style="text-align: center;">24926</td>
<td style="text-align: center;">62910</td>
<td style="text-align: center;">124354</td>
</tr>
<tr class="even">
<td style="text-align: center;">run2</td>
<td style="text-align: center;">reads_out</td>
<td style="text-align: center;">137</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">96147</td>
<td style="text-align: center;">45022</td>
<td style="text-align: center;">57113</td>
<td style="text-align: center;">69108</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>In the code above we’ve used the function <code>get_summary_stats()</code> from the package <code>rstatix</code>. This function allows for different types of summary statistics (we’ve used type = “five number”). We note that we have a median of around 60k reads per sample.</p>
</section>
<section id="count-table" class="level3">
<h3 class="anchored" data-anchor-id="count-table">Count table</h3>
<p>Now let’s look at the count table. This table contains the count of each amplicon sequence variant (ASV) in each sample. It will become the core of our phyloseq object.</p>
<p>Describe the count table</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(count_tab)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "matrix" "array" </code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(count_tab)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 271</code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ncol</span>(count_tab)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6422</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">is.na</span>(count_tab))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0</code></pre>
</div>
</div>
<p>The count table is a matrix (a 2D array) with 271 rows and 6422 columns. There is no missing data. Based on these numbers, it seems that rows are samples and columns are ASVs.</p>
<p>Let’s make sure. How are the rows and columns named (labelled)?</p>
<p>What are the row names?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(count_tab) <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="at">n =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "GUT0001" "GUT0002" "GUT0003"</code></pre>
</div>
</div>
<p>In this count table, the rows are samples. They are named using the amplicon_sample_id.</p>
<p>Is every sample in the count table listed in the sample data?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>sample_df <span class="sc">%&gt;%</span> </span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span>amplicon_sample_id <span class="sc">%in%</span> <span class="fu">rownames</span>(count_tab)) <span class="sc">%&gt;%</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<caption>Table continues below</caption>
<colgroup>
<col style="width: 28%">
<col style="width: 21%">
<col style="width: 22%">
<col style="width: 16%">
<col style="width: 12%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">amplicon_sample_id</th>
<th style="text-align: center;">amplicon_type</th>
<th style="text-align: center;">instrument_run</th>
<th style="text-align: center;">reads_out</th>
<th style="text-align: center;">pid</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">GUT0071</td>
<td style="text-align: center;">study_sample</td>
<td style="text-align: center;">run1</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">pid_08</td>
</tr>
<tr class="even">
<td style="text-align: center;">GUT0117</td>
<td style="text-align: center;">study_sample</td>
<td style="text-align: center;">run1</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">pid_22</td>
</tr>
<tr class="odd">
<td style="text-align: center;">VAG0280</td>
<td style="text-align: center;">study_sample</td>
<td style="text-align: center;">run2</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">pid_40</td>
</tr>
</tbody>
</table>
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 16%">
<col style="width: 18%">
<col style="width: 19%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">arm</th>
<th style="text-align: center;">time_point</th>
<th style="text-align: center;">sample_type</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">week_7</td>
<td style="text-align: center;">gut_biopsy</td>
</tr>
<tr class="even">
<td style="text-align: center;">treatment</td>
<td style="text-align: center;">week_1</td>
<td style="text-align: center;">gut_biopsy</td>
</tr>
<tr class="odd">
<td style="text-align: center;">treatment</td>
<td style="text-align: center;">week_1</td>
<td style="text-align: center;">vaginal</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># And this difficult-to-read command should evaluate to TRUE</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">rownames</span>(count_tab) <span class="sc">%in%</span> sample_df<span class="sc">$</span>amplicon_sample_id) <span class="sc">==</span> <span class="fu">nrow</span>(count_tab)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
<p>Yes, all of the samples in the count table are present in the sample data, except for the three samples that failed (and that’s okay, we can leave these three samples in the sample data).</p>
<p>What are the column names?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(count_tab) <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="at">n =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "TAGGGAATCTTCCACAATGGACGCAAGTCTGATGGAGCAACGCCGCGTGAGTGAAGAAGGGTTTCGGCTCGTAAAGCTCTGTTGTTGGTGAAGAAGGACAGGGGTAGTAACTGACCTTTGTTTGACGGTAATCAATTAGAAAGTCACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGTGCAGGCGGCTCGATAAGTCTGATGTGAAAGCCTTCGGCTCAACCGGAGAATTGCATCAGAAACTGTCGAGCTTGAGTACAGAAGAGGAGAGTGGAACTCCATGTGTAGCGGTGAAATGCGTAGATATATGGAAGAACACCGGTGGCGAAGGCGGCTCTCTGGTCTGTTACTGACGCTGAGGCTCGAAAGCATGGGTAGCGAACAGG"
[2] "TGAGGAATATTGGTCAATGGGCGAGAGCCTGAACCAGCCAAGTAGCGTGAAGGATGACTGCCCTATGGGTTGTAAACTTCTTTTATAAAGGAATAAAGTCGGGTATGGATACCCGTTTGCATGTACTTTATGAATAAGGATCGGCTAACTCCGTGCCAGCAGCCGCGGTAATACGGAGGATCCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGAGCGTAGATGGATGTTTAAGTCAGTTGTGAAAGTTTGCGGCTCAACCGTAAAATTGCAGTTGATACTGGATATCTTGAGTGCAGTTGAGGCAGGCGGAATTCGTGGTGTAGCGGTGAAATGCTTAGATATCACGAAGAACTCCGATTGCGAAGGCAGCCTGCTAAGCTGCAACTGACATTGAGGCTCGAAAGTGTGGGTATCAAACAGG"     
[3] "TGGGGAATATTGCACAATGGGCGCAAGCCTGATGCAGCCATGCCGCGTGTATGAAGAAGGCCTTCGGGTTGTAAAGTACTTTCAGCGGGGAGGAAGGGAGTAAAGTTAATACCTTTGCTCATTGACGTTACCCGCAGAAGAAGCACCGGCTAACTCCGTGCCAGCAGCCGCGGTAATACGGAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCACGCAGGCGGTTTGTTAAGTCAGATGTGAAATCCCCGGGCTCAACCTGGGAACTGCATCTGATACTGGCAAGCTTGAGTCTCGTAGAGGGGGGTAGAATTCCAGGTGTAGCGGTGAAATGCGTAGAGATCTGGAGGAATACCGGTGGCGAAGGCGGCCCCCTGGACGAAGACTGACGCTCAGGTGCGAAAGCGTGGGGAGCAAACAGG"</code></pre>
</div>
</div>
<p>Wow, how long are these names?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(count_tab) <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="at">n =</span> <span class="dv">3</span>) <span class="sc">%&gt;%</span> <span class="fu">nchar</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 429 424 429</code></pre>
</div>
</div>
<p>In this count table, the columns are ASVs and they are named using … the ASV sequences themselves!! Here, these labels can be &gt;400 characters long. (Discuss pros, cons, and workarounds for using ASV sequences as labels.)</p>
<p>Finally, how “sparse” is our count table? In other words, what fraction of the elements are zero?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(count_tab <span class="sc">==</span> <span class="dv">0</span>) <span class="sc">/</span> <span class="fu">length</span>(count_tab)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.9858483</code></pre>
</div>
</div>
<p>Over 98% of the elements are zero. This count table is very sparse.</p>
<p>Can we create a plot that helps us think about this?</p>
<p>Let’s try</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(<span class="at">asv_prev =</span> <span class="fu">colSums</span>(count_tab <span class="sc">&gt;</span> <span class="dv">0</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> asv_prev)) <span class="sc">+</span></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Number of samples in which the ASV was detected"</span>,</span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Number of ASVs"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The table is sparse because most ASVs are absent from most samples. There are very few ASVs that are widespread (detected in many or most samples). How could we modify the above code to see the number of ASVs detected in many (say, &gt;50) samples? Hint: Use the <code>filter()</code> function.</p>
<p>Let’s save this ASV prevalence data in a dataframe. We can also add the total count for the ASV.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>prevalence_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">asv_sum =</span> <span class="fu">colSums</span>(count_tab),</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>                            <span class="at">asv_prev =</span> <span class="fu">colSums</span>(count_tab <span class="sc">&gt;</span> <span class="dv">0</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"asv_seq"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>How prevalent is the most prevalent ASV?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>prevalence_df <span class="sc">%&gt;%</span> </span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(asv_prev <span class="sc">==</span> <span class="fu">max</span>(asv_prev))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                                                                                                                                                                                                                                                                                                                                                                                                                        asv_seq
1 TAGGGAATCTTCCACAATGGACGCAAGTCTGATGGAGCAACGCCGCGTGAGTGAAGAAGGGTTTCGGCTCGTAAAGCTCTGTTGTTGGTGAAGAAGGACAGGGGTAGTAACTGACCTTTGTTTGACGGTAATCAATTAGAAAGTCACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGTGCAGGCGGCTCGATAAGTCTGATGTGAAAGCCTTCGGCTCAACCGGAGAATTGCATCAGAAACTGTCGAGCTTGAGTACAGAAGAGGAGAGTGGAACTCCATGTGTAGCGGTGAAATGCGTAGATATATGGAAGAACACCGGTGGCGAAGGCGGCTCTCTGGTCTGTTACTGACGCTGAGGCTCGAAAGCATGGGTAGCGAACAGG
  asv_sum asv_prev
1 2147189      228</code></pre>
</div>
</div>
<p>This ASV was detected in 228 out of 271 samples. Huh, I wonder who this is?</p>
</section>
<section id="taxonomy-table" class="level3">
<h3 class="anchored" data-anchor-id="taxonomy-table">Taxonomy table</h3>
<p>Now let’s look at the taxonomy table. This table contains the taxonomic assignments made using <code>dada2</code>.</p>
<p>Describe the taxonomy table</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(tax_tab)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "matrix" "array" </code></pre>
</div>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(tax_tab)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6422</code></pre>
</div>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ncol</span>(tax_tab)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 8</code></pre>
</div>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">is.na</span>(tax_tab))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 7388</code></pre>
</div>
</div>
<p>The taxonomy table is a matrix (a 2D array) with 6422 rows and 8 columns. Missing data are present (in cases where the taxonomic assignment was unknown or ambiguous). Based on these numbers, it seems that rows are ASVs. What are the columns?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(tax_tab)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"   "Species"
[8] "Label"  </code></pre>
</div>
</div>
<p>The columns are taxonomic ranks. Note that I’ve added a variable at far right called “Label” which contains a short label for each ASV (e.g., “asv1”, “asv2”, etc).</p>
<p>At this point, it’s generally a good idea to explore various features of the ASVs.</p>
<p>To do this, one can begin by creating a dataframe from the taxonomy table. I think of this dataframe as a place to store <em>any</em> ASV-associated information, not only taxonomic assignments.</p>
<p>So let’s do this</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>asv_df <span class="ot">&lt;-</span> tax_tab <span class="sc">%&gt;%</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"asv_seq"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(prevalence_df, <span class="at">by =</span> <span class="st">"asv_seq"</span>) <span class="sc">%&gt;%</span> <span class="co"># Add data from above </span></span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">asv_len =</span> <span class="fu">nchar</span>(asv_seq))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can explore, for example, the distribution of ASV lengths in relation to ASV prevalence</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>asv_df <span class="sc">%&gt;%</span> </span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> asv_len, <span class="at">y =</span> asv_prev, <span class="at">color =</span> Kingdom)) <span class="sc">+</span></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">size =</span> <span class="dv">3</span>, <span class="at">stroke =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"How prevalent are any length outliers?"</span>,</span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"ASV length"</span>,</span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"ASV prevalence"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We have length outliers; they are not very prevalent. After some follow-up, we might decide to filter them out. Maybe they are problematic ASVs, e.g., chimeric or non-specific (non-SSU rRNA) amplicons.</p>
<p>We also notice three non-bacterial ASVs. Interesting! Do they make sense, given the body sites we’ve sampled?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>asv_df <span class="sc">%&gt;%</span> </span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(Kingdom <span class="sc">!=</span> <span class="st">"Bacteria"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>asv_seq) <span class="sc">%&gt;%</span> </span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(Genus<span class="sc">:</span>asv_len) <span class="sc">%&gt;%</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 28%">
<col style="width: 13%">
<col style="width: 13%">
<col style="width: 13%">
<col style="width: 15%">
<col style="width: 15%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">Genus</th>
<th style="text-align: center;">Species</th>
<th style="text-align: center;">Label</th>
<th style="text-align: center;">asv_sum</th>
<th style="text-align: center;">asv_prev</th>
<th style="text-align: center;">asv_len</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">Methanobrevibacter</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">asv1136</td>
<td style="text-align: center;">670</td>
<td style="text-align: center;">4</td>
<td style="text-align: center;">388</td>
</tr>
<tr class="even">
<td style="text-align: center;">Trichomonas</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">asv1815</td>
<td style="text-align: center;">219</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">457</td>
</tr>
<tr class="odd">
<td style="text-align: center;">Methanosphaera</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">asv4889</td>
<td style="text-align: center;">8</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">388</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Our sample types are gut and vaginal. So, yes, these make sense.</p>
<p>Let’s look at a few of the oddly short ASVs</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>asv_df <span class="sc">%&gt;%</span> </span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(asv_prev <span class="sc">&gt;</span> <span class="dv">1</span>, asv_len <span class="sc">&lt;</span> <span class="dv">280</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>asv_seq) <span class="sc">%&gt;%</span> </span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>(<span class="at">n =</span> <span class="dv">3</span>) <span class="sc">%&gt;%</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<caption>Table continues below</caption>
<colgroup>
<col style="width: 14%">
<col style="width: 20%">
<col style="width: 24%">
<col style="width: 17%">
<col style="width: 11%">
<col style="width: 11%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">Kingdom</th>
<th style="text-align: center;">Phylum</th>
<th style="text-align: center;">Class</th>
<th style="text-align: center;">Order</th>
<th style="text-align: center;">Family</th>
<th style="text-align: center;">Genus</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">Bacteria</td>
<td style="text-align: center;">Cyanobacteria</td>
<td style="text-align: center;">Oxyphotobacteria</td>
<td style="text-align: center;">Chloroplast</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">NA</td>
</tr>
<tr class="even">
<td style="text-align: center;">Bacteria</td>
<td style="text-align: center;">Cyanobacteria</td>
<td style="text-align: center;">Oxyphotobacteria</td>
<td style="text-align: center;">Chloroplast</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">NA</td>
</tr>
<tr class="odd">
<td style="text-align: center;">Bacteria</td>
<td style="text-align: center;">Cyanobacteria</td>
<td style="text-align: center;">Oxyphotobacteria</td>
<td style="text-align: center;">Chloroplast</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">NA</td>
</tr>
</tbody>
</table>
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 13%">
<col style="width: 13%">
<col style="width: 13%">
<col style="width: 15%">
<col style="width: 15%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">Species</th>
<th style="text-align: center;">Label</th>
<th style="text-align: center;">asv_sum</th>
<th style="text-align: center;">asv_prev</th>
<th style="text-align: center;">asv_len</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">NA</td>
<td style="text-align: center;">asv2318</td>
<td style="text-align: center;">95</td>
<td style="text-align: center;">4</td>
<td style="text-align: center;">270</td>
</tr>
<tr class="even">
<td style="text-align: center;">NA</td>
<td style="text-align: center;">asv2338</td>
<td style="text-align: center;">93</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">271</td>
</tr>
<tr class="odd">
<td style="text-align: center;">NA</td>
<td style="text-align: center;">asv2576</td>
<td style="text-align: center;">63</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">271</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>These are chloroplast sequences. We might consider filtering out ASVs assigned to the Order <em>Chloroplast</em>. But we should state our reasoning. For example, “chloroplasts are not, in and of themselves, members of the gut or vaginal microbiota”.</p>
<p>Let’s look at a few long ASVs</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>asv_df <span class="sc">%&gt;%</span> </span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(asv_len <span class="sc">&gt;</span> <span class="dv">450</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>asv_seq) <span class="sc">%&gt;%</span> </span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(Genus<span class="sc">:</span>asv_len) <span class="sc">%&gt;%</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 19%">
<col style="width: 13%">
<col style="width: 13%">
<col style="width: 13%">
<col style="width: 15%">
<col style="width: 15%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">Genus</th>
<th style="text-align: center;">Species</th>
<th style="text-align: center;">Label</th>
<th style="text-align: center;">asv_sum</th>
<th style="text-align: center;">asv_prev</th>
<th style="text-align: center;">asv_len</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">Trichomonas</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">asv1815</td>
<td style="text-align: center;">219</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">457</td>
</tr>
<tr class="even">
<td style="text-align: center;">Bacteroides</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">asv5826</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">460</td>
</tr>
<tr class="odd">
<td style="text-align: center;">Alistipes</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">asv6152</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">460</td>
</tr>
<tr class="even">
<td style="text-align: center;">Bacteroides</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">asv6203</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">475</td>
</tr>
<tr class="odd">
<td style="text-align: center;">Ureaplasma</td>
<td style="text-align: center;">NA</td>
<td style="text-align: center;">asv6384</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">498</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Of these five, I’d guess the <em>Trichomonas</em> ASV is real (and from vaginal samples) and the others are problematic. Just my hunch.</p>
<p>In general, we’d continue to explore the ASV data until we developed an intuition of which ASVs, if any, we might want to set aside. For example, here we might retain ASVs with length &gt; 385-nt and &lt; 460-nt, that aren’t Order <em>Chloroplast</em> (or Family <em>Mitochondria</em>).</p>
<p>Let’s make a vector containing the ASVs we wish to keep</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>keepers <span class="ot">&lt;-</span> asv_df <span class="sc">%&gt;%</span> </span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(asv_len <span class="sc">&gt;</span> <span class="dv">385</span> <span class="sc">&amp;</span> asv_len <span class="sc">&lt;</span> <span class="dv">460</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Careful below! Don't drop NA unless you intend to</span></span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(Order <span class="sc">!=</span> <span class="st">"Chloroplast"</span> <span class="sc">|</span> <span class="fu">is.na</span>(Order)) <span class="sc">%&gt;%</span> </span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(Family <span class="sc">!=</span> <span class="st">"Mitochondria"</span> <span class="sc">|</span> <span class="fu">is.na</span>(Family)) <span class="sc">%$%</span></span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unique</span>(asv_seq)</span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Retaining these ASVs removes how many ASVs?</span></span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(asv_df) <span class="sc">-</span> <span class="fu">length</span>(keepers)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 146</code></pre>
</div>
</div>
<p>Btw, who was it - the most prevalent ASV in our dataset?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>asv_df <span class="sc">%&gt;%</span> </span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(asv_prev <span class="sc">==</span> <span class="fu">max</span>(asv_prev)) <span class="sc">%&gt;%</span> </span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(Genus<span class="sc">:</span>asv_len) <span class="sc">%&gt;%</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 22%">
<col style="width: 13%">
<col style="width: 11%">
<col style="width: 13%">
<col style="width: 15%">
<col style="width: 15%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">Genus</th>
<th style="text-align: center;">Species</th>
<th style="text-align: center;">Label</th>
<th style="text-align: center;">asv_sum</th>
<th style="text-align: center;">asv_prev</th>
<th style="text-align: center;">asv_len</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">Lactobacillus</td>
<td style="text-align: center;">iners</td>
<td style="text-align: center;">asv1</td>
<td style="text-align: center;">2147189</td>
<td style="text-align: center;">228</td>
<td style="text-align: center;">429</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Interesting, we think of <em>L. iners</em> as a vaginal species. Does is also appear in the gut?</p>
</section>
</section>
<section id="make-phyloseq-object" class="level2">
<h2 class="anchored" data-anchor-id="make-phyloseq-object">2. Make phyloseq object</h2>
<p>Let’s build our phyloseq object using the <code>phyloseq()</code> function from the <code>phyloseq</code> package</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>ps <span class="ot">&lt;-</span> <span class="fu">phyloseq</span>(<span class="fu">sample_data</span>(sample_df <span class="sc">%&gt;%</span> </span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>                             <span class="fu">column_to_rownames</span>(<span class="at">var =</span> <span class="st">"amplicon_sample_id"</span>)),</span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>               <span class="fu">otu_table</span>(count_tab, <span class="at">taxa_are_rows =</span> <span class="cn">FALSE</span>),</span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>               <span class="fu">tax_table</span>(tax_tab))</span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>ps <span class="co"># Prints a concise summary</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 6422 taxa and 271 samples ]
sample_data() Sample Data:       [ 271 samples by 7 sample variables ]
tax_table()   Taxonomy Table:    [ 6422 taxa by 8 taxonomic ranks ]</code></pre>
</div>
</div>
<p>A phyloseq object is a special data structure for organizing, linking, storing, and analyzing multiple related types of data from sequencing-based studies (e.g., marker gene surveys).</p>
<p>Here we’re using three slots (additional slots exist for a phylogenetic tree and reference sequences). <code>Phyloseq</code> checks that sample and ASV labels match across the different slots. Note that the three samples (in our sample data) that did not return reads have been dropped.</p>
<section id="accessor-functions" class="level3">
<h3 class="anchored" data-anchor-id="accessor-functions">Accessor functions</h3>
<p><code>Phyloseq</code> provides various helpful “accessor” functions enabling queries of the phyloseq object. We won’t demonstrate all of them here, but they are worth learning.</p>
<p>For example</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_variables</span>(ps)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "amplicon_type"  "instrument_run" "reads_out"      "pid"           
[5] "arm"            "time_point"     "sample_type"   </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rank_names</span>(ps)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"   "Species"
[8] "Label"  </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_sums</span>(ps) <span class="sc">%&gt;%</span> <span class="fu">min</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="fu">taxa_sums</span>(ps) <span class="sc">%&gt;%</span> <span class="fu">min</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1</code></pre>
</div>
</div>
<p>Use the functions <code>sample_data()</code>, <code>otu_table()</code> and <code>tax_table()</code> to access (extract) the component objects.</p>
<p>For example</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_data</span>(ps) <span class="sc">%&gt;%</span> </span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"amplicon_sample_id"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>(<span class="at">n =</span> <span class="dv">3</span>) <span class="sc">%&gt;%</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<caption>Table continues below</caption>
<colgroup>
<col style="width: 28%">
<col style="width: 21%">
<col style="width: 22%">
<col style="width: 16%">
<col style="width: 12%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">amplicon_sample_id</th>
<th style="text-align: center;">amplicon_type</th>
<th style="text-align: center;">instrument_run</th>
<th style="text-align: center;">reads_out</th>
<th style="text-align: center;">pid</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">GUT0001</td>
<td style="text-align: center;">study_sample</td>
<td style="text-align: center;">run1</td>
<td style="text-align: center;">191197</td>
<td style="text-align: center;">pid_32</td>
</tr>
<tr class="even">
<td style="text-align: center;">GUT0002</td>
<td style="text-align: center;">study_sample</td>
<td style="text-align: center;">run1</td>
<td style="text-align: center;">17206</td>
<td style="text-align: center;">pid_14</td>
</tr>
<tr class="odd">
<td style="text-align: center;">GUT0003</td>
<td style="text-align: center;">study_sample</td>
<td style="text-align: center;">run1</td>
<td style="text-align: center;">112769</td>
<td style="text-align: center;">pid_01</td>
</tr>
</tbody>
</table>
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 16%">
<col style="width: 18%">
<col style="width: 19%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">arm</th>
<th style="text-align: center;">time_point</th>
<th style="text-align: center;">sample_type</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">treatment</td>
<td style="text-align: center;">week_1</td>
<td style="text-align: center;">gut_biopsy</td>
</tr>
<tr class="even">
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">week_7</td>
<td style="text-align: center;">gut_biopsy</td>
</tr>
<tr class="odd">
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">baseline</td>
<td style="text-align: center;">gut_biopsy</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="processor-functions" class="level3">
<h3 class="anchored" data-anchor-id="processor-functions">Processor functions</h3>
<p><code>Phyloseq</code> also provides various helpful “processor” functions. These functions allow for pruning, subsetting, filtering, transforming, and glomming (aggregating) the data.</p>
<p>For example, we could retain ASVs we wish to keep and samples with adequate sequencing depth</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>ps_clean <span class="ot">&lt;-</span> ps <span class="sc">%&gt;%</span> </span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">prune_taxa</span>(keepers, .) <span class="sc">%&gt;%</span> <span class="co"># What is the dot? </span></span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">prune_samples</span>(<span class="fu">sample_sums</span>(.) <span class="sc">&gt;</span> <span class="dv">100</span>, .) <span class="sc">%&gt;%</span> </span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter_taxa</span>(., <span class="cf">function</span>(x) <span class="fu">sum</span>(x <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">&gt;</span> <span class="dv">0</span>, <span class="at">prune =</span> <span class="cn">TRUE</span>) <span class="co"># Removes empty ASVs</span></span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a>ps_clean</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 6271 taxa and 265 samples ]
sample_data() Sample Data:       [ 265 samples by 7 sample variables ]
tax_table()   Taxonomy Table:    [ 6271 taxa by 8 taxonomic ranks ]</code></pre>
</div>
</div>
<p>Check out our new phyloseq object. Are the low-yield samples gone?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="fu">min</span>(<span class="fu">sample_sums</span>(ps_clean))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 330</code></pre>
</div>
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="fu">min</span>(<span class="fu">taxa_sums</span>(ps_clean))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1</code></pre>
</div>
</div>
<p>Yes! Are the chloroplasts gone?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tax_table</span>(ps_clean) <span class="sc">%&gt;%</span> </span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(Order <span class="sc">==</span> <span class="st">"Chloroplast"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] Kingdom Phylum  Class   Order   Family  Genus   Species Label  
&lt;0 rows&gt; (or 0-length row.names)</code></pre>
</div>
</div>
<p>Yes. I miss them a little.</p>
<p>We will encounter other “processor” functions as we continue to explore our data.</p>
<p>Now let’s explore our data in three ways:</p>
<ul>
<li>Plotting taxon relative abundances (the <code>plot_bar()</code> function)</li>
<li>Within-sample diversity (also known as alpha diversity)</li>
<li>Between-sample diversity (beta diversity; distance metrics and ordination)</li>
</ul>
<p>And for each way, let’s look at our zymo mocks (e.g., run1 vs run2) and our study samples (e.g., gut vs vaginal; baseline placebo vs treatment).</p>
</section>
</section>
<section id="plot-bars" class="level2">
<h2 class="anchored" data-anchor-id="plot-bars">3. Plot bars</h2>
<p>Visualizing taxon (relative) abundances</p>
<p>Composition assessment using stacked bars works well for small numbers of samples and/or small numbers of taxa. With more than around 8-10(?) taxa, they get too busy to track visually (in my opinion). Nor do I advocate for binning samples to categories prior to plotting, as the meaning of such bins often seems vague (again, my opinion). Stacked bars should work well for our zymo mocks; for our study samples, I’m not so sure. But let’s try!</p>
<section id="zymo-mocks" class="level3">
<h3 class="anchored" data-anchor-id="zymo-mocks">Zymo mocks</h3>
<p>In theory, the Zymo mock should contain 8 bacterial genera. You can review the list of species <a href="https://www.zymoresearch.com/collections/zymobiomics-microbial-community-standards/products/zymobiomics-microbial-community-dna-standard">here</a>.</p>
<p>The following code performs three “pre-processing steps” before plotting using <code>plot_bar()</code> and adding some <code>ggplot</code> layers. We can take it step by step.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>ps_clean <span class="sc">%&gt;%</span> </span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Subset to control samples</span></span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">subset_samples</span>(., amplicon_type <span class="sc">==</span> <span class="st">"zymo_mock"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb89-5"><a href="#cb89-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb89-6"><a href="#cb89-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Transform counts to relative abundance</span></span>
<span id="cb89-7"><a href="#cb89-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">transform_sample_counts</span>(., <span class="cf">function</span>(x) x<span class="sc">/</span><span class="fu">sum</span>(x)) <span class="sc">%&gt;%</span></span>
<span id="cb89-8"><a href="#cb89-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb89-9"><a href="#cb89-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Remove ASVs present in fewer than 6 samples</span></span>
<span id="cb89-10"><a href="#cb89-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># What happens when we drop the number below 6?</span></span>
<span id="cb89-11"><a href="#cb89-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Toggle the number down, rerun, and find out</span></span>
<span id="cb89-12"><a href="#cb89-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter_taxa</span>(., <span class="cf">function</span>(x) <span class="fu">sum</span>(x <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">&gt;=</span> <span class="dv">6</span>, <span class="at">prune =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb89-13"><a href="#cb89-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb89-14"><a href="#cb89-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Plot the stacked bars</span></span>
<span id="cb89-15"><a href="#cb89-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_bar</span>(<span class="at">fill =</span> <span class="st">"Genus"</span>) <span class="sc">+</span> </span>
<span id="cb89-16"><a href="#cb89-16" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb89-17"><a href="#cb89-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add some ggplot2 layers</span></span>
<span id="cb89-18"><a href="#cb89-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(.<span class="sc">~</span>instrument_run, <span class="at">scales =</span> <span class="st">"free_x"</span>) <span class="sc">+</span></span>
<span id="cb89-19"><a href="#cb89-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Zymo mocks"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-43-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>In the above plotted bars, each division is an ASV and each color is a Genus. So there are two <em>Staphylococcus</em> ASVs, for example.</p>
<p>We observe that for the Zymo genera (those we <em>expect</em> to be present), things look good - the stacked bars are consistent within and between batches. However …</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>ps_clean <span class="sc">%&gt;%</span> </span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Subset to control samples</span></span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">subset_samples</span>(., amplicon_type <span class="sc">==</span> <span class="st">"zymo_mock"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb90-6"><a href="#cb90-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Transform counts to relative abundance</span></span>
<span id="cb90-7"><a href="#cb90-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">transform_sample_counts</span>(., <span class="cf">function</span>(x) x<span class="sc">/</span><span class="fu">sum</span>(x)) <span class="sc">%&gt;%</span></span>
<span id="cb90-8"><a href="#cb90-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb90-9"><a href="#cb90-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Subset taxa to Mycoplasma or Sneathia</span></span>
<span id="cb90-10"><a href="#cb90-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">subset_taxa</span>(., Genus <span class="sc">==</span> <span class="st">"Sneathia"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb90-11"><a href="#cb90-11" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb90-12"><a href="#cb90-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Plot the stacked bars</span></span>
<span id="cb90-13"><a href="#cb90-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_bar</span>(<span class="at">fill =</span> <span class="st">"Species"</span>) <span class="sc">+</span> </span>
<span id="cb90-14"><a href="#cb90-14" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb90-15"><a href="#cb90-15" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add some ggplot2 layers</span></span>
<span id="cb90-16"><a href="#cb90-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(.<span class="sc">~</span>instrument_run, <span class="at">scales =</span> <span class="st">"free_x"</span>) <span class="sc">+</span></span>
<span id="cb90-17"><a href="#cb90-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Zymo mocks"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-44-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>When we look at the non-Zymo genera (those we <em>don’t</em> expect to be present), which are at very low abundance, we see there may be sources of contamination that vary by batch. Reflect on what this might mean for our downstream analyses.</p>
</section>
<section id="study-samples" class="level3">
<h3 class="anchored" data-anchor-id="study-samples">Study samples</h3>
<p>For each body site, let’s compare the placebo and treatment groups at baseline (before intervention)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>ps_clean <span class="sc">%&gt;%</span> </span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Subset to vaginal at baseline</span></span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">subset_samples</span>(., sample_type <span class="sc">==</span> <span class="st">"vaginal"</span> <span class="sc">&amp;</span></span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a>                   time_point <span class="sc">==</span> <span class="st">"baseline"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb91-6"><a href="#cb91-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb91-7"><a href="#cb91-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Drop empty ASVs</span></span>
<span id="cb91-8"><a href="#cb91-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter_taxa</span>(., <span class="cf">function</span>(x) <span class="fu">sum</span>(x <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">&gt;</span> <span class="dv">0</span>, <span class="at">prune =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb91-9"><a href="#cb91-9" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb91-10"><a href="#cb91-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Transform counts to relative abundance</span></span>
<span id="cb91-11"><a href="#cb91-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">transform_sample_counts</span>(., <span class="cf">function</span>(x) x<span class="sc">/</span><span class="fu">sum</span>(x)) <span class="sc">%&gt;%</span></span>
<span id="cb91-12"><a href="#cb91-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb91-13"><a href="#cb91-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Aggregate families</span></span>
<span id="cb91-14"><a href="#cb91-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_glom</span>(., <span class="at">taxrank =</span> <span class="st">"Family"</span>) <span class="sc">%&gt;%</span> <span class="co"># What happens to NA? </span></span>
<span id="cb91-15"><a href="#cb91-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-16"><a href="#cb91-16" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Drop sporadic families</span></span>
<span id="cb91-17"><a href="#cb91-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter_taxa</span>(., <span class="cf">function</span>(x) <span class="fu">sum</span>(x <span class="sc">&gt;</span> <span class="fl">0.01</span>) <span class="sc">&gt;</span> <span class="dv">2</span>, <span class="at">prune =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb91-18"><a href="#cb91-18" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb91-19"><a href="#cb91-19" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Plot the stacked bars</span></span>
<span id="cb91-20"><a href="#cb91-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_bar</span>(<span class="at">x =</span> <span class="st">"pid"</span>, <span class="at">fill =</span> <span class="st">"Family"</span>) <span class="sc">+</span> </span>
<span id="cb91-21"><a href="#cb91-21" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb91-22"><a href="#cb91-22" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add some ggplot2 layers</span></span>
<span id="cb91-23"><a href="#cb91-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(.<span class="sc">~</span>arm, <span class="at">scales =</span> <span class="st">"free_x"</span>) <span class="sc">+</span> </span>
<span id="cb91-24"><a href="#cb91-24" aria-hidden="true" tabindex="-1"></a>  <span class="co"># theme(legend.position = "none") +</span></span>
<span id="cb91-25"><a href="#cb91-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"vaginal, baseline"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>ps_clean <span class="sc">%&gt;%</span> </span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Subset to gut at baseline</span></span>
<span id="cb92-4"><a href="#cb92-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">subset_samples</span>(., sample_type <span class="sc">==</span> <span class="st">"gut_biopsy"</span> <span class="sc">&amp;</span></span>
<span id="cb92-5"><a href="#cb92-5" aria-hidden="true" tabindex="-1"></a>                   time_point <span class="sc">==</span> <span class="st">"baseline"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb92-6"><a href="#cb92-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb92-7"><a href="#cb92-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Drop empty ASVs</span></span>
<span id="cb92-8"><a href="#cb92-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter_taxa</span>(., <span class="cf">function</span>(x) <span class="fu">sum</span>(x <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">&gt;</span> <span class="dv">0</span>, <span class="at">prune =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb92-9"><a href="#cb92-9" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb92-10"><a href="#cb92-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Transform counts to relative abundance</span></span>
<span id="cb92-11"><a href="#cb92-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">transform_sample_counts</span>(., <span class="cf">function</span>(x) x<span class="sc">/</span><span class="fu">sum</span>(x)) <span class="sc">%&gt;%</span></span>
<span id="cb92-12"><a href="#cb92-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb92-13"><a href="#cb92-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Aggregate families</span></span>
<span id="cb92-14"><a href="#cb92-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_glom</span>(., <span class="at">taxrank =</span> <span class="st">"Family"</span>) <span class="sc">%&gt;%</span> <span class="co"># What happens to NA? </span></span>
<span id="cb92-15"><a href="#cb92-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-16"><a href="#cb92-16" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Drop sporadic families</span></span>
<span id="cb92-17"><a href="#cb92-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter_taxa</span>(., <span class="cf">function</span>(x) <span class="fu">sum</span>(x <span class="sc">&gt;</span> <span class="fl">0.01</span>) <span class="sc">&gt;</span> <span class="dv">2</span>, <span class="at">prune =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb92-18"><a href="#cb92-18" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb92-19"><a href="#cb92-19" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Plot the stacked bars</span></span>
<span id="cb92-20"><a href="#cb92-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_bar</span>(<span class="at">x =</span> <span class="st">"pid"</span>, <span class="at">fill =</span> <span class="st">"Family"</span>) <span class="sc">+</span> </span>
<span id="cb92-21"><a href="#cb92-21" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb92-22"><a href="#cb92-22" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add some ggplot2 layers</span></span>
<span id="cb92-23"><a href="#cb92-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(.<span class="sc">~</span>arm, <span class="at">scales =</span> <span class="st">"free_x"</span>) <span class="sc">+</span> </span>
<span id="cb92-24"><a href="#cb92-24" aria-hidden="true" tabindex="-1"></a>  <span class="co"># theme(legend.position = "none") +</span></span>
<span id="cb92-25"><a href="#cb92-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"gut, baseline"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-46-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Do the arms look fairly similar prior to intervention? So, stacked bars for the study samples are a little unsatisfying, I think. A lot of work subsetting and aggregating for not much insight. What do you think?</p>
<p>But we can pick better questions for this <code>plot_bar()</code> function.</p>
<p>For example, we were curious if asv1, <em>L. iners</em>, appeared in the gut samples; in fact, it must, given its prevalence. This would be surprising (I think?) because <em>L. iners</em> is thought of as a vagina-specific organism.</p>
<p>So let’s look</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>ps_clean <span class="sc">%&gt;%</span> </span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">subset_samples</span>(., sample_type <span class="sc">==</span> <span class="st">"gut_biopsy"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">transform_sample_counts</span>(., <span class="cf">function</span>(x) x<span class="sc">/</span><span class="fu">sum</span>(x)) <span class="sc">%&gt;%</span> </span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">subset_taxa</span>(., Label <span class="sc">==</span> <span class="st">"asv1"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb93-5"><a href="#cb93-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_bar</span>(<span class="at">x =</span> <span class="st">"pid"</span>, <span class="at">fill =</span> <span class="st">"instrument_run"</span>) <span class="sc">+</span></span>
<span id="cb93-6"><a href="#cb93-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(.<span class="sc">~</span>time_point, <span class="at">scales =</span> <span class="st">"free_x"</span>) <span class="sc">+</span></span>
<span id="cb93-7"><a href="#cb93-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"asv1 (L. iners) in gut samples"</span>,</span>
<span id="cb93-8"><a href="#cb93-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"participant_id"</span>,</span>
<span id="cb93-9"><a href="#cb93-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"asv1_frequency"</span>) <span class="sc">+</span></span>
<span id="cb93-10"><a href="#cb93-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-47-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Yes, asv1 (<em>L. iners</em>) is present at low frequency in many of our gut_biopsy samples.</p>
<p>All of the gut samples were sequenced in run1. Do the zymo mocks from this batch (run1), which are not expected to contain <em>L. iners</em>, also harbor low levels of asv1?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>ps_clean <span class="sc">%&gt;%</span> </span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">subset_samples</span>(., amplicon_type <span class="sc">==</span> <span class="st">"zymo_mock"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">transform_sample_counts</span>(., <span class="cf">function</span>(x) x<span class="sc">/</span><span class="fu">sum</span>(x)) <span class="sc">%&gt;%</span> </span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">subset_taxa</span>(., Label <span class="sc">==</span> <span class="st">"asv1"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb94-5"><a href="#cb94-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_bar</span>() <span class="sc">+</span></span>
<span id="cb94-6"><a href="#cb94-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(.<span class="sc">~</span>instrument_run, <span class="at">scales =</span> <span class="st">"free_x"</span>) <span class="sc">+</span></span>
<span id="cb94-7"><a href="#cb94-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"asv1 (L. iners) in zymo mocks"</span>,</span>
<span id="cb94-8"><a href="#cb94-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"amplicon_sample_id"</span>,</span>
<span id="cb94-9"><a href="#cb94-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"asv1_frequency"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-48-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Indeed, they do. So what do we think is going on here with asv1 (<em>L. iners</em>)? Where did it come from? What other samples, if any, were included in this batch? What might we do differently next time?</p>
</section>
</section>
<section id="alpha-diversity" class="level2">
<h2 class="anchored" data-anchor-id="alpha-diversity">4. Alpha diversity</h2>
<p>Alpha diversity metrics are crucial in ecology and microbiome studies because they provide insights into the richness and evenness of species within a single ecosystem or community sample. This allows researchers to understand:</p>
<ul>
<li><p>Species Richness: How many different types of species exist in a single community? This can be particularly important in assessing the overall biodiversity.</p></li>
<li><p>Species Evenness: How evenly these species are distributed within the community. In other words, whether a few species dominate the community, or whether many species coexist in relatively similar numbers.</p></li>
</ul>
<p>The data can be used to compare different communities or to see how a community changes over time in response to various factors such as perturbation, disease, or interventions.</p>
<p>In general, various alpha diversity metrics differ in whether, and to what degree, they emphasize richness or evenness.</p>
<p>The <code>phyloseq</code> package provides an all-in-one function <code>plot_richness()</code> that both calculates and plots an array of alpha diversity measures. But to be honest, I don’t often use it, opting instead to calculate the metrics using the <code>estimate_richness()</code> function. This places them in a dataframe, to which we can add our sample data and create plots using <code>ggplot2</code>.</p>
<p>Look at the documentation by calling <code>?estimate_richness</code> at the console prompt. Which alpha diversity measures are available to us within the function <code>estimate_richness()</code>?</p>
<p>Note: Some estimates of alpha diversity (e.g., Chao1, ACE) must be calculated on unfiltered data. Specifically, singletons should be present. (That is, ASVs that appear only once (count of 1) within samples should be present.) This is because singletons are considered as a term in these estimates. Does our data contain singletons?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">otu_table</span>(ps_clean) <span class="sc">==</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 42</code></pre>
</div>
</div>
<p>Yes, but not many. (Think about how dada2 works and why this might be.) We won’t use Chao1 or ACE here, but it’s still a good idea to calculate whatever alpha diversity measure you choose using relatively unfiltered data.</p>
<p>Let’s calculate alpha diversity using three different measures. The result will return as a dataframe. Then let’s join our alpha diversity data to our sample data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>measures <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Observed"</span>, <span class="st">"Shannon"</span>, <span class="st">"Simpson"</span>)</span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a>alpha_df <span class="ot">&lt;-</span> ps_clean <span class="sc">%&gt;%</span> </span>
<span id="cb97-4"><a href="#cb97-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">estimate_richness</span>(<span class="at">measures =</span> measures) <span class="sc">%&gt;%</span> </span>
<span id="cb97-5"><a href="#cb97-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"amplicon_sample_id"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb97-6"><a href="#cb97-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(sample_df, <span class="at">by =</span> <span class="st">"amplicon_sample_id"</span>)</span>
<span id="cb97-7"><a href="#cb97-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb97-8"><a href="#cb97-8" aria-hidden="true" tabindex="-1"></a>alpha_df <span class="sc">%&gt;%</span> </span>
<span id="cb97-9"><a href="#cb97-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>() <span class="sc">%&gt;%</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<caption>Table continues below</caption>
<colgroup>
<col style="width: 29%">
<col style="width: 15%">
<col style="width: 13%">
<col style="width: 13%">
<col style="width: 22%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">amplicon_sample_id</th>
<th style="text-align: center;">Observed</th>
<th style="text-align: center;">Shannon</th>
<th style="text-align: center;">Simpson</th>
<th style="text-align: center;">amplicon_type</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">GUT0001</td>
<td style="text-align: center;">126</td>
<td style="text-align: center;">3.286</td>
<td style="text-align: center;">0.9296</td>
<td style="text-align: center;">study_sample</td>
</tr>
<tr class="even">
<td style="text-align: center;">GUT0002</td>
<td style="text-align: center;">112</td>
<td style="text-align: center;">3.454</td>
<td style="text-align: center;">0.9274</td>
<td style="text-align: center;">study_sample</td>
</tr>
<tr class="odd">
<td style="text-align: center;">GUT0003</td>
<td style="text-align: center;">117</td>
<td style="text-align: center;">3.277</td>
<td style="text-align: center;">0.9403</td>
<td style="text-align: center;">study_sample</td>
</tr>
<tr class="even">
<td style="text-align: center;">GUT0004</td>
<td style="text-align: center;">105</td>
<td style="text-align: center;">3.163</td>
<td style="text-align: center;">0.9113</td>
<td style="text-align: center;">study_sample</td>
</tr>
<tr class="odd">
<td style="text-align: center;">GUT0005</td>
<td style="text-align: center;">122</td>
<td style="text-align: center;">3.265</td>
<td style="text-align: center;">0.9067</td>
<td style="text-align: center;">study_sample</td>
</tr>
<tr class="even">
<td style="text-align: center;">GUT0006</td>
<td style="text-align: center;">239</td>
<td style="text-align: center;">4.333</td>
<td style="text-align: center;">0.9764</td>
<td style="text-align: center;">study_sample</td>
</tr>
</tbody>
</table>
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 22%">
<col style="width: 15%">
<col style="width: 11%">
<col style="width: 15%">
<col style="width: 16%">
<col style="width: 18%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">instrument_run</th>
<th style="text-align: center;">reads_out</th>
<th style="text-align: center;">pid</th>
<th style="text-align: center;">arm</th>
<th style="text-align: center;">time_point</th>
<th style="text-align: center;">sample_type</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">run1</td>
<td style="text-align: center;">191197</td>
<td style="text-align: center;">pid_32</td>
<td style="text-align: center;">treatment</td>
<td style="text-align: center;">week_1</td>
<td style="text-align: center;">gut_biopsy</td>
</tr>
<tr class="even">
<td style="text-align: center;">run1</td>
<td style="text-align: center;">17206</td>
<td style="text-align: center;">pid_14</td>
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">week_7</td>
<td style="text-align: center;">gut_biopsy</td>
</tr>
<tr class="odd">
<td style="text-align: center;">run1</td>
<td style="text-align: center;">112769</td>
<td style="text-align: center;">pid_01</td>
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">baseline</td>
<td style="text-align: center;">gut_biopsy</td>
</tr>
<tr class="even">
<td style="text-align: center;">run1</td>
<td style="text-align: center;">25242</td>
<td style="text-align: center;">pid_06</td>
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">baseline</td>
<td style="text-align: center;">gut_biopsy</td>
</tr>
<tr class="odd">
<td style="text-align: center;">run1</td>
<td style="text-align: center;">50533</td>
<td style="text-align: center;">pid_09</td>
<td style="text-align: center;">treatment</td>
<td style="text-align: center;">week_1</td>
<td style="text-align: center;">gut_biopsy</td>
</tr>
<tr class="even">
<td style="text-align: center;">run1</td>
<td style="text-align: center;">142644</td>
<td style="text-align: center;">pid_18</td>
<td style="text-align: center;">treatment</td>
<td style="text-align: center;">week_7</td>
<td style="text-align: center;">gut_biopsy</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Now let’s make some plots comparing alpha diversity across samples of interest.</p>
<section id="zymo-mocks-1" class="level3">
<h3 class="anchored" data-anchor-id="zymo-mocks-1">Zymo mocks</h3>
<p>In the absence of any batch effect, we would expect no difference in alpha diversity between mocks from run1 vs run2</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>) <span class="co"># For reproducibility</span></span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a>alpha_df <span class="sc">%&gt;%</span> </span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(amplicon_type <span class="sc">==</span> <span class="st">"zymo_mock"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb98-5"><a href="#cb98-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">all_of</span>(measures),</span>
<span id="cb98-6"><a href="#cb98-6" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"adiv_measure"</span>,</span>
<span id="cb98-7"><a href="#cb98-7" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"adiv_value"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb98-8"><a href="#cb98-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> instrument_run, <span class="at">y =</span> adiv_value)) <span class="sc">+</span></span>
<span id="cb98-9"><a href="#cb98-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">outlier.shape =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb98-10"><a href="#cb98-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">width =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb98-11"><a href="#cb98-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(.<span class="sc">~</span>adiv_measure, <span class="at">scales =</span> <span class="st">"free_y"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-51-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>However, they <em>are</em> different, suggesting a batch effect. There seem to be “extra” taxa in mocks from run1 (keeping in mind that the Zymo mock contains eight bacterial species). We could discuss and test whether rarefying the data or aggregating taxa change this result.</p>
<p>Note: The traditional Simpson diversity index represents the probability that two individuals randomly selected from a sample will belong to the same species. Higher values indicate lower diversity. Therefore, it is common to transform Simpson to 1 - Simpson. How would you modify the last chunk to plot 1 - Simpson in place of Simpson?</p>
</section>
<section id="study-samples-1" class="level3">
<h3 class="anchored" data-anchor-id="study-samples-1">Study samples</h3>
<p>If our allocation of participants to study arms was random, we might expect no difference in alpha diversity between the placebo and treatment groups at baseline (before the study intervention). Is this true?</p>
<p>First we plot</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggbeeswarm) <span class="co"># For function geom_quasirandom()</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>alpha_df <span class="sc">%&gt;%</span> </span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(time_point <span class="sc">==</span> <span class="st">"baseline"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">all_of</span>(measures),</span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"adiv_measure"</span>,</span>
<span id="cb100-5"><a href="#cb100-5" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"adiv_value"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb100-6"><a href="#cb100-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> arm, <span class="at">y =</span> adiv_value)) <span class="sc">+</span></span>
<span id="cb100-7"><a href="#cb100-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">alpha =</span> <span class="fl">0.4</span>, <span class="at">outlier.shape =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb100-8"><a href="#cb100-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_quasirandom</span>(<span class="at">stroke =</span> <span class="cn">FALSE</span>, <span class="at">width =</span> <span class="fl">0.3</span>) <span class="sc">+</span></span>
<span id="cb100-9"><a href="#cb100-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(sample_type<span class="sc">~</span>adiv_measure, <span class="at">scales =</span> <span class="st">"free_y"</span>) <span class="sc">+</span></span>
<span id="cb100-10"><a href="#cb100-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"baseline (pre-intervention)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-53-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We notice some small differences. If these are of interest or concern to us, we can follow up. For example, for the baseline vaginal Shannon diversity index, we might ask whether the difference between groups (arms) was significant?</p>
<p>Let’s use a Wilcoxon test using the function <code>wilcox_test()</code> from the package <code>rstatix</code>. (It’s the same as base R’s <code>wilcox.test()</code>.)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>alpha_df <span class="sc">%&gt;%</span> </span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(time_point <span class="sc">==</span> <span class="st">"baseline"</span>,</span>
<span id="cb101-3"><a href="#cb101-3" aria-hidden="true" tabindex="-1"></a>         sample_type <span class="sc">==</span> <span class="st">"vaginal"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb101-4"><a href="#cb101-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">wilcox_test</span>(Shannon <span class="sc">~</span> arm, <span class="at">paired =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 13%">
<col style="width: 13%">
<col style="width: 16%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 16%">
<col style="width: 11%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">.y.</th>
<th style="text-align: center;">group1</th>
<th style="text-align: center;">group2</th>
<th style="text-align: center;">n1</th>
<th style="text-align: center;">n2</th>
<th style="text-align: center;">statistic</th>
<th style="text-align: center;">p</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">Shannon</td>
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">treatment</td>
<td style="text-align: center;">23</td>
<td style="text-align: center;">20</td>
<td style="text-align: center;">287</td>
<td style="text-align: center;">0.171</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>At baseline, prior to intervention, there was no significant difference between the two groups (arms) in terms of vaginal bacterial diversity, as measured using the Shannon diversity index (Wilcoxon rank sum test; <em>P</em> = 0.171).</p>
<p>In the chunk above, change the time_point to “week_1” (that’s after intervention) and re-run the code. What do you notice? How about for “week_7”?</p>
<p>Let’s visualize the trends. We’ll use the function <code>stat_compare_means()</code> from the package <code>ggpubr</code> to compare between arms at each time point using the same type of test we ran above (a Wilcoxon rank sum test).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggpubr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s start with the vaginal data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>alpha_df <span class="sc">%&gt;%</span> </span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(sample_type <span class="sc">==</span> <span class="st">"vaginal"</span>) <span class="sc">%&gt;%</span>  </span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> arm, <span class="at">y =</span> Shannon)) <span class="sc">+</span></span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">alpha =</span> <span class="fl">0.4</span>, <span class="at">outlier.shape =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb103-5"><a href="#cb103-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_quasirandom</span>(<span class="at">stroke =</span> <span class="cn">FALSE</span>, <span class="at">width =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb103-6"><a href="#cb103-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(.<span class="sc">~</span>time_point) <span class="sc">+</span></span>
<span id="cb103-7"><a href="#cb103-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_compare_means</span>(<span class="at">method =</span> <span class="st">"wilcox.test"</span>, <span class="at">paired =</span> <span class="cn">FALSE</span>) <span class="sc">+</span> </span>
<span id="cb103-8"><a href="#cb103-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Vaginal samples"</span>,</span>
<span id="cb103-9"><a href="#cb103-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Study arm"</span>,</span>
<span id="cb103-10"><a href="#cb103-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Shannon diversity index"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-56-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>How would you describe these trends?</p>
<p>What do we notice if we modify the chunk above to perform the same analysis for gut samples? Or for different measures of alpha diversity?</p>
<p>Finally, there’s an alternative way to view these data, that is, viewing it as longitudinal data. In this view, participants are compared to themselves prior to intervention.</p>
<p>Let’s visualize</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>alpha_df <span class="sc">%&gt;%</span> </span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">drop_na</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> time_point, <span class="at">y =</span> Shannon)) <span class="sc">+</span></span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">group =</span> pid), <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb104-5"><a href="#cb104-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">stroke =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb104-6"><a href="#cb104-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(sample_type <span class="sc">~</span> arm, <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb104-7"><a href="#cb104-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Timepoint"</span>,</span>
<span id="cb104-8"><a href="#cb104-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Shannon diversity index"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-57-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Individuals are dynamic!</p>
<p>As a follow-up, we might be interested in comparing week_1 to baseline <em>within</em> arms using a paired test so that participants are compared to themselves.</p>
<p>Here’s an example</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get and arrange the data</span></span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a>test_df <span class="ot">&lt;-</span> alpha_df <span class="sc">%&gt;%</span> </span>
<span id="cb105-3"><a href="#cb105-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(arm <span class="sc">==</span> <span class="st">"treatment"</span>,</span>
<span id="cb105-4"><a href="#cb105-4" aria-hidden="true" tabindex="-1"></a>         time_point <span class="sc">!=</span> <span class="st">"week_7"</span>,</span>
<span id="cb105-5"><a href="#cb105-5" aria-hidden="true" tabindex="-1"></a>         sample_type <span class="sc">==</span> <span class="st">"vaginal"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb105-6"><a href="#cb105-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(pid) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(<span class="fu">n</span>() <span class="sc">==</span> <span class="dv">2</span>) <span class="sc">%&gt;%</span> <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> <span class="co"># Retain complete cases </span></span>
<span id="cb105-7"><a href="#cb105-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(pid, time_point) <span class="co"># Important step if paired tests will be performed</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Display summary statistics</span></span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a>test_df <span class="sc">%&gt;%</span> </span>
<span id="cb106-3"><a href="#cb106-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(time_point) <span class="sc">%&gt;%</span> </span>
<span id="cb106-4"><a href="#cb106-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">get_summary_stats</span>(Shannon, <span class="at">type =</span> <span class="st">"five_number"</span>) <span class="sc">%&gt;%</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 18%">
<col style="width: 15%">
<col style="width: 6%">
<col style="width: 11%">
<col style="width: 11%">
<col style="width: 11%">
<col style="width: 12%">
<col style="width: 12%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">time_point</th>
<th style="text-align: center;">variable</th>
<th style="text-align: center;">n</th>
<th style="text-align: center;">min</th>
<th style="text-align: center;">max</th>
<th style="text-align: center;">q1</th>
<th style="text-align: center;">median</th>
<th style="text-align: center;">q3</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">baseline</td>
<td style="text-align: center;">Shannon</td>
<td style="text-align: center;">18</td>
<td style="text-align: center;">0.169</td>
<td style="text-align: center;">3.438</td>
<td style="text-align: center;">1.038</td>
<td style="text-align: center;">2.033</td>
<td style="text-align: center;">2.859</td>
</tr>
<tr class="even">
<td style="text-align: center;">week_1</td>
<td style="text-align: center;">Shannon</td>
<td style="text-align: center;">18</td>
<td style="text-align: center;">0.148</td>
<td style="text-align: center;">2.893</td>
<td style="text-align: center;">0.301</td>
<td style="text-align: center;">0.875</td>
<td style="text-align: center;">2.272</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Do a paired test using base R</span></span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wilcox.test</span>(<span class="at">x =</span> test_df<span class="sc">$</span>Shannon[test_df<span class="sc">$</span>time_point <span class="sc">==</span> <span class="st">"baseline"</span>],</span>
<span id="cb107-3"><a href="#cb107-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">y =</span> test_df<span class="sc">$</span>Shannon[test_df<span class="sc">$</span>time_point <span class="sc">==</span> <span class="st">"week_1"</span>],</span>
<span id="cb107-4"><a href="#cb107-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">paired =</span> <span class="cn">TRUE</span>, <span class="at">alternative =</span> <span class="st">"two.sided"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Wilcoxon signed rank exact test

data:  test_df$Shannon[test_df$time_point == "baseline"] and test_df$Shannon[test_df$time_point == "week_1"]
V = 125, p-value = 0.08977
alternative hypothesis: true location shift is not equal to 0</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="co"># But of course there's also the friendlier way</span></span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a>test_df <span class="sc">%&gt;%</span></span>
<span id="cb109-3"><a href="#cb109-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">wilcox_test</span>(Shannon <span class="sc">~</span> time_point, <span class="at">paired =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 13%">
<col style="width: 15%">
<col style="width: 12%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 16%">
<col style="width: 12%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">.y.</th>
<th style="text-align: center;">group1</th>
<th style="text-align: center;">group2</th>
<th style="text-align: center;">n1</th>
<th style="text-align: center;">n2</th>
<th style="text-align: center;">statistic</th>
<th style="text-align: center;">p</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">Shannon</td>
<td style="text-align: center;">baseline</td>
<td style="text-align: center;">week_1</td>
<td style="text-align: center;">18</td>
<td style="text-align: center;">18</td>
<td style="text-align: center;">125</td>
<td style="text-align: center;">0.0898</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>How would you express this result? If you modified the chunks above to examine the placebo arm (or week_7, or gut samples), what do you find?</p>
</section>
</section>
<section id="beta-diversity" class="level2">
<h2 class="anchored" data-anchor-id="beta-diversity">5. Beta diversity</h2>
<p>Variation (sometimes called turnover) in species composition over space and time; in relation to any variable of interest such as perturbation, disease, or intervention</p>
<p>Beta diversity refers to between-sample diversity. At its core are measures of how much diversity is shared (or not shared) between communities. These are measures of pairwise resemblance; referred to as similarity, dissimilarity, or distance metrics. These measures differ in how they are calculated; whether and to what degree they consider abundance information; and whether they treat taxa as independent or (phylogenetically) related units. Different measures emphasize different aspects of the data and therefore may give different results. Insight may be gained from these differing results.</p>
<p>When we calculate distances (i.e., between-sample diversity) among many, many pairs of samples, we typically can’t ascertain directly (by plotting or by eye) the main patterns or structures in the data. To explore these patterns or structures, we employ unsupervised methods that reduce this “high-dimensionality”. These methods include unconstrained ordination and clustering. They are considered unsupervised because the goal is not to predict or explain the importance of one particular variable, but to visualize and understand the primary sources of variation within the data, whatever they might be. In an ordination plot, similar samples are placed closer together, and dissimilar samples are placed further away.</p>
<p>The <code>phyloseq</code> package provides functions for calculating measures of pairwise resemblance and performing ordination. Core functions include <code>distance()</code>, <code>ordinate()</code> and <code>plot_ordination()</code>. Please see the documentation for the function <code>vegdist()</code> from the <code>vegan</code> package for details on many dissimilarity measures.</p>
<p>Steps in a typical exploratory beta-diversity workflow might include:</p>
<ul>
<li>Subet samples and/or filter taxa</li>
<li>Transform count data (if necessary or desired)</li>
<li>Calculate distances</li>
<li>Ordinate</li>
<li>Plot</li>
<li>Repeat using alternative distance measures or ordination methods</li>
</ul>
<section id="zymo-mocks-2" class="level3">
<h3 class="anchored" data-anchor-id="zymo-mocks-2">Zymo mocks</h3>
<p>We can demo this very quickly using our Zymo mocks.</p>
<p>If batch effects were negligible, then we’d expect little separation by batch. In other words, we do not expect strong clustering by batch.</p>
<p>First we’ll use the Jaccard dissimilarity measure, which is a binary measure, meaning it considers only presence-absence. And as an ordination method we’ll use principal coordinate analysis (PCoA; also called multidimensional scaling, or MDS).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a>ps_clean <span class="sc">%&gt;%</span> </span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">subset_samples</span>(., amplicon_type <span class="sc">==</span> <span class="st">"zymo_mock"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb110-3"><a href="#cb110-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter_taxa</span>(., <span class="cf">function</span>(x) <span class="fu">sum</span>(x <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">&gt;</span> <span class="dv">0</span>, <span class="at">prune =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb110-4"><a href="#cb110-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ordinate</span>(., <span class="at">distance =</span> <span class="st">"jaccard"</span>, <span class="at">binary =</span> <span class="cn">TRUE</span>, <span class="at">method =</span> <span class="st">"MDS"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb110-5"><a href="#cb110-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_ordination</span>(ps_clean, ., <span class="at">type =</span> <span class="st">"samples"</span>, <span class="at">color =</span> <span class="st">"instrument_run"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-62-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Again, we see evidence of a batch effect. At this point, we’re not surprised - we’ve seen evidence of this in our bar plots and in our alpha diversity plots.</p>
<p>What happens to the plot above if we change the distance metric to Bray-Curtis (<code>distance = "bray"</code>), a quantitative measure, meaning that it takes abundance information into account? (Don’t forget to remove or set to false the <code>binary</code> argument.) Does the pattern change? Does our conclusion change?</p>
<p>You may have noticed several distance metrics listed under the function <code>vegan::vegdist()</code> that aren’t available in <code>phyloseq</code>. One of them is the robust Aitchison distance. This metric is a popular choice among microbiome researchers because of the way it handles the compositionality of our data using a centered log-ratio (CLR) transformation. This is especially important when we have relatively few taxa, such as after agglomeration (aggregation) at high taxonomic level (e.g., Phylum or Class). Here’s an example using this distance metric.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vegan)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the count table (what vegan calls the "community data matrix")</span></span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a>zym_tab <span class="ot">&lt;-</span> ps_clean <span class="sc">%&gt;%</span> </span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">subset_samples</span>(., amplicon_type <span class="sc">==</span> <span class="st">"zymo_mock"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb112-4"><a href="#cb112-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter_taxa</span>(., <span class="cf">function</span>(x) <span class="fu">sum</span>(x <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">&gt;</span> <span class="dv">0</span>, <span class="at">prune =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb112-5"><a href="#cb112-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">otu_table</span>() </span>
<span id="cb112-6"><a href="#cb112-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb112-7"><a href="#cb112-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Use vegan's vegdist() to calculate a robust Aitchison distance matrix</span></span>
<span id="cb112-8"><a href="#cb112-8" aria-hidden="true" tabindex="-1"></a>ait_dist_zym <span class="ot">&lt;-</span> <span class="fu">vegdist</span>(<span class="at">x =</span> zym_tab, <span class="at">method =</span> <span class="st">"robust.aitchison"</span>)</span>
<span id="cb112-9"><a href="#cb112-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb112-10"><a href="#cb112-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Input this distance matrix into phyloseq's ordinate() and plot</span></span>
<span id="cb112-11"><a href="#cb112-11" aria-hidden="true" tabindex="-1"></a>ps_clean <span class="sc">%&gt;%</span></span>
<span id="cb112-12"><a href="#cb112-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ordinate</span>(., <span class="at">distance =</span> ait_dist_zym, <span class="at">method =</span> <span class="st">"MDS"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb112-13"><a href="#cb112-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_ordination</span>(ps_clean, ., <span class="at">type =</span> <span class="st">"samples"</span>, <span class="at">color =</span> <span class="st">"instrument_run"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-64-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Looks … familiar.</p>
<p>Note that <code>plot_ordination()</code> includes an argument <code>justDF</code> that outputs the data underlying the plot rather than the plot itself. This can be helpful if you’d prefer to customize your plot using a different package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Highlight `justDF` argument</span></span>
<span id="cb113-2"><a href="#cb113-2" aria-hidden="true" tabindex="-1"></a>ps_clean <span class="sc">%&gt;%</span></span>
<span id="cb113-3"><a href="#cb113-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ordinate</span>(., <span class="at">distance =</span> ait_dist_zym, <span class="at">method =</span> <span class="st">"MDS"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb113-4"><a href="#cb113-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_ordination</span>(ps_clean, ., <span class="at">axes =</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>), <span class="at">justDF =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb113-5"><a href="#cb113-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(Axis<span class="fl">.1</span><span class="sc">:</span>reads_out) <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="at">n =</span> <span class="dv">3</span>) <span class="sc">%&gt;%</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<caption>Table continues below</caption>
<colgroup>
<col style="width: 38%">
<col style="width: 12%">
<col style="width: 13%">
<col style="width: 13%">
<col style="width: 21%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">&nbsp;</th>
<th style="text-align: center;">Axis.1</th>
<th style="text-align: center;">Axis.2</th>
<th style="text-align: center;">Axis.3</th>
<th style="text-align: center;">amplicon_type</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;"><strong>ZYMO.DNA.CTRL.P1.run1</strong></td>
<td style="text-align: center;">-7.108</td>
<td style="text-align: center;">-0.9808</td>
<td style="text-align: center;">-0.5162</td>
<td style="text-align: center;">zymo_mock</td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>ZYMO.DNA.CTRL.P2.run1</strong></td>
<td style="text-align: center;">-7.97</td>
<td style="text-align: center;">-4.122</td>
<td style="text-align: center;">0.02995</td>
<td style="text-align: center;">zymo_mock</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><strong>ZYMO.DNA.CTRL.P3.run1</strong></td>
<td style="text-align: center;">-4.194</td>
<td style="text-align: center;">-4.718</td>
<td style="text-align: center;">0.2392</td>
<td style="text-align: center;">zymo_mock</td>
</tr>
</tbody>
</table>
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 38%">
<col style="width: 23%">
<col style="width: 16%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">&nbsp;</th>
<th style="text-align: center;">instrument_run</th>
<th style="text-align: center;">reads_out</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;"><strong>ZYMO.DNA.CTRL.P1.run1</strong></td>
<td style="text-align: center;">run1</td>
<td style="text-align: center;">147436</td>
</tr>
<tr class="even">
<td style="text-align: center;"><strong>ZYMO.DNA.CTRL.P2.run1</strong></td>
<td style="text-align: center;">run1</td>
<td style="text-align: center;">144714</td>
</tr>
<tr class="odd">
<td style="text-align: center;"><strong>ZYMO.DNA.CTRL.P3.run1</strong></td>
<td style="text-align: center;">run1</td>
<td style="text-align: center;">114688</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="study-samples-2" class="level3">
<h3 class="anchored" data-anchor-id="study-samples-2">Study samples</h3>
<p>The next three plots should give a taste of how clustering patterns may change given, e.g., the choice of distance metric. The first is binary Jaccard, the second is Bray-Curtis, and the third is robust Aitchison.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a>ps_clean <span class="sc">%&gt;%</span> </span>
<span id="cb114-2"><a href="#cb114-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">subset_samples</span>(., <span class="sc">!</span><span class="fu">is.na</span>(sample_type)) <span class="sc">%&gt;%</span></span>
<span id="cb114-3"><a href="#cb114-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter_taxa</span>(., <span class="cf">function</span>(x) <span class="fu">sum</span>(x <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">&gt;</span> <span class="dv">0</span>, <span class="at">prune =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb114-4"><a href="#cb114-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ordinate</span>(., <span class="at">distance =</span> <span class="st">"jaccard"</span>, <span class="at">binary =</span> <span class="cn">TRUE</span>, <span class="at">method =</span> <span class="st">"MDS"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb114-5"><a href="#cb114-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_ordination</span>(ps_clean, ., <span class="at">type =</span> <span class="st">"samples"</span>, <span class="at">color =</span> <span class="st">"sample_type"</span>) <span class="sc">+</span></span>
<span id="cb114-6"><a href="#cb114-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"binary Jaccard"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-66-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a>ps_clean <span class="sc">%&gt;%</span> </span>
<span id="cb115-2"><a href="#cb115-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">subset_samples</span>(., <span class="sc">!</span><span class="fu">is.na</span>(sample_type)) <span class="sc">%&gt;%</span></span>
<span id="cb115-3"><a href="#cb115-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter_taxa</span>(., <span class="cf">function</span>(x) <span class="fu">sum</span>(x <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">&gt;</span> <span class="dv">0</span>, <span class="at">prune =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb115-4"><a href="#cb115-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ordinate</span>(., <span class="at">distance =</span> <span class="st">"bray"</span>, <span class="at">binary =</span> <span class="cn">FALSE</span>, <span class="at">method =</span> <span class="st">"MDS"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb115-5"><a href="#cb115-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_ordination</span>(ps_clean, ., <span class="at">type =</span> <span class="st">"samples"</span>, <span class="at">color =</span> <span class="st">"sample_type"</span>) <span class="sc">+</span></span>
<span id="cb115-6"><a href="#cb115-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Bray-Curtis"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-67-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a>sam_tab <span class="ot">&lt;-</span> ps_clean <span class="sc">%&gt;%</span> </span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">subset_samples</span>(., <span class="sc">!</span><span class="fu">is.na</span>(sample_type)) <span class="sc">%&gt;%</span></span>
<span id="cb116-3"><a href="#cb116-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter_taxa</span>(., <span class="cf">function</span>(x) <span class="fu">sum</span>(x <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">&gt;</span> <span class="dv">0</span>, <span class="at">prune =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb116-4"><a href="#cb116-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">otu_table</span>() </span>
<span id="cb116-5"><a href="#cb116-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-6"><a href="#cb116-6" aria-hidden="true" tabindex="-1"></a>ait_dist_sam <span class="ot">&lt;-</span> <span class="fu">vegdist</span>(<span class="at">x =</span> sam_tab, <span class="at">method =</span> <span class="st">"robust.aitchison"</span>)</span>
<span id="cb116-7"><a href="#cb116-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-8"><a href="#cb116-8" aria-hidden="true" tabindex="-1"></a>ps_clean <span class="sc">%&gt;%</span></span>
<span id="cb116-9"><a href="#cb116-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ordinate</span>(., <span class="at">distance =</span> ait_dist_sam, <span class="at">method =</span> <span class="st">"MDS"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb116-10"><a href="#cb116-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_ordination</span>(ps_clean, ., <span class="at">type =</span> <span class="st">"samples"</span>, <span class="at">color =</span> <span class="st">"sample_type"</span>) <span class="sc">+</span></span>
<span id="cb116-11"><a href="#cb116-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"robust Aitchison"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-68-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We see some different shapes. Nonetheless, all plots suggest that the primary source of variation in our data is by body site (gut versus vaginal). However, we can’t rule out that what we see here is driven by a batch effect. So to be sure, we’d need to address that issue – by correcting for the batch effect or by regenerating the data.</p>
<p>Try re-running the three plots with different taxa filtering or subsetting. Or using different distance measures. Or different ordination methods (e.g., NMDS).</p>
</section>
</section>
<section id="lets-practice" class="level2">
<h2 class="anchored" data-anchor-id="lets-practice">6. Let’s practice!</h2>
<p>Let’s practice what we’ve learned using some new data. I’ve provided two “practice” phyloseq objects, both based on published data.</p>
<p>One is gut (stool) microbiome data from mildly lactose-intolerant adults sampled over time as they eliminated and then reintroduced milk into their diet (<code>dairy.ps</code>).</p>
<p>The other is vaginal microbiome data from women sampled soon before and soon after giving birth, some via vaginal delivery and others via c-section (<code>delivery.ps</code>).</p>
<section id="manage-paths" class="level3">
<h3 class="anchored" data-anchor-id="manage-paths">Manage paths</h3>
<p>Show paths to the data we intend to analyze</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a>fs<span class="sc">::</span><span class="fu">dir_ls</span>(<span class="fu">here</span>(<span class="st">"materials/2-workshop2/4-16s-pt1/data"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>/Users/jelsherbini/dev/durban-data-science-for-biology/materials/2-workshop2/4-16s-pt1/data/instructional
/Users/jelsherbini/dev/durban-data-science-for-biology/materials/2-workshop2/4-16s-pt1/data/practice</code></pre>
</div>
</div>
<p>Create an object containing the path to the practice data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a>pt_path <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="fu">here</span>(<span class="st">"materials/2-workshop2/4-16s-pt1/data/practice/"</span>))</span>
<span id="cb119-2"><a href="#cb119-2" aria-hidden="true" tabindex="-1"></a>pt_path</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "/Users/jelsherbini/dev/durban-data-science-for-biology/materials/2-workshop2/4-16s-pt1/data/practice/"</code></pre>
</div>
</div>
<p>List the files we intend to load</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(pt_path)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "dairy_phyloseq.rds"    "delivery_phyloseq.rds"</code></pre>
</div>
</div>
</section>
<section id="load-data" class="level3">
<h3 class="anchored" data-anchor-id="load-data">Load data</h3>
<p>Let’s load the practice data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a>ps_dairy <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(pt_path, <span class="st">"dairy_phyloseq.rds"</span>))</span>
<span id="cb123-2"><a href="#cb123-2" aria-hidden="true" tabindex="-1"></a>ps_deliv <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(pt_path, <span class="st">"delivery_phyloseq.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="dairy-dataset" class="level3">
<h3 class="anchored" data-anchor-id="dairy-dataset">Dairy dataset</h3>
<p>Produce a concise summary of <code>ps_dairy</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a>ps_dairy</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 3001 taxa and 434 samples ]
sample_data() Sample Data:       [ 434 samples by 4 sample variables ]
tax_table()   Taxonomy Table:    [ 3001 taxa by 6 taxonomic ranks ]</code></pre>
</div>
</div>
<p>Access the sample variables for <code>ps_dairy</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_variables</span>(ps_dairy)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "participant_id" "previous_diet"  "study_day"      "study_phase"   </code></pre>
</div>
</div>
<p>Review the study design by making a plot; use the variables <code>participant_id</code>, <code>study_day</code>, and <code>study_phase</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a>ps_dairy <span class="sc">%&gt;%</span> </span>
<span id="cb128-2"><a href="#cb128-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_data</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb128-3"><a href="#cb128-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> study_day, <span class="at">y =</span> participant_id, <span class="at">color =</span> study_phase)) <span class="sc">+</span></span>
<span id="cb128-4"><a href="#cb128-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb128-5"><a href="#cb128-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"dairy study; stool samples"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-75-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Question: What is the primary source of variation among the samples in <code>ps_dairy</code>? Is it <code>participant_id</code> or <code>study_phase</code>? Hint: Begin with binary Jaccard</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a>ps_dairy <span class="sc">%&gt;%</span> </span>
<span id="cb129-2"><a href="#cb129-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ordinate</span>(., <span class="at">distance =</span> <span class="st">"jaccard"</span>, <span class="at">binary =</span> <span class="cn">TRUE</span>, <span class="at">method =</span> <span class="st">"MDS"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb129-3"><a href="#cb129-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_ordination</span>(ps_dairy, ., <span class="at">type =</span> <span class="st">"samples"</span>, <span class="at">color =</span> <span class="st">"participant_id"</span>) <span class="sc">+</span></span>
<span id="cb129-4"><a href="#cb129-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"binary Jaccard"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-76-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a>ps_dairy <span class="sc">%&gt;%</span> </span>
<span id="cb130-2"><a href="#cb130-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ordinate</span>(., <span class="at">distance =</span> <span class="st">"jaccard"</span>, <span class="at">binary =</span> <span class="cn">TRUE</span>, <span class="at">method =</span> <span class="st">"MDS"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb130-3"><a href="#cb130-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_ordination</span>(ps_dairy, ., <span class="at">type =</span> <span class="st">"samples"</span>, <span class="at">color =</span> <span class="st">"study_phase"</span>) <span class="sc">+</span></span>
<span id="cb130-4"><a href="#cb130-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"binary Jaccard"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-76-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Your answer?</p>
</section>
<section id="delivery-dataset" class="level3">
<h3 class="anchored" data-anchor-id="delivery-dataset">Delivery dataset</h3>
<p>Produce a concise summary of <code>ps_deliv</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a>ps_deliv</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 1551 taxa and 140 samples ]
sample_data() Sample Data:       [ 140 samples by 7 sample variables ]
tax_table()   Taxonomy Table:    [ 1551 taxa by 12 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 1551 tips and 1550 internal nodes ]</code></pre>
</div>
</div>
<p>Access the sample variables for <code>ps_deliv</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_variables</span>(ps_deliv)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pregnancy_id"       "day_vs_delivery"    "status"            
[4] "delivery_mode"      "sp_lacto_crispatus" "sp_lacto_gasseri"  
[7] "sp_lacto_iners"    </code></pre>
</div>
</div>
<p>Review the study design by making a plot; use the variables <code>pregnancy_id</code>, <code>day_vs_delivery</code>, and <code>status</code>; facet by <code>delivery_mode</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a>ps_deliv <span class="sc">%&gt;%</span> </span>
<span id="cb135-2"><a href="#cb135-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_data</span>() <span class="sc">%&gt;%</span></span>
<span id="cb135-3"><a href="#cb135-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> day_vs_delivery, <span class="at">y =</span> pregnancy_id, <span class="at">color =</span> <span class="fu">fct_rev</span>(status))) <span class="sc">+</span></span>
<span id="cb135-4"><a href="#cb135-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">0</span>, <span class="at">linetype =</span> <span class="dv">2</span>, <span class="at">linewidth =</span> <span class="fl">0.3</span>) <span class="sc">+</span></span>
<span id="cb135-5"><a href="#cb135-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb135-6"><a href="#cb135-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>delivery_mode, <span class="at">scales =</span> <span class="st">"free_y"</span>) <span class="sc">+</span></span>
<span id="cb135-7"><a href="#cb135-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"delivery study; vaginal samples"</span>,</span>
<span id="cb135-8"><a href="#cb135-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">color =</span> <span class="st">"status"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-79-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Question: Does the Shannon diversity index change with delivery? Does this depend on delivery mode?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate Shannon</span></span>
<span id="cb136-2"><a href="#cb136-2" aria-hidden="true" tabindex="-1"></a>shan_del <span class="ot">&lt;-</span> ps_deliv <span class="sc">%&gt;%</span> </span>
<span id="cb136-3"><a href="#cb136-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">estimate_richness</span>(<span class="at">measures =</span> <span class="st">"Shannon"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb136-4"><a href="#cb136-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"sample_id"</span>)</span>
<span id="cb136-5"><a href="#cb136-5" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb136-6"><a href="#cb136-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Access the sample data and attach Shannon</span></span>
<span id="cb136-7"><a href="#cb136-7" aria-hidden="true" tabindex="-1"></a>sam_shan_del <span class="ot">&lt;-</span> ps_deliv <span class="sc">%&gt;%</span> </span>
<span id="cb136-8"><a href="#cb136-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_data</span>() <span class="sc">%&gt;%</span> <span class="fu">data.frame</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb136-9"><a href="#cb136-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"sample_id"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb136-10"><a href="#cb136-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(shan_del, <span class="at">by =</span> <span class="st">"sample_id"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb136-11"><a href="#cb136-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(pregnancy_id) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(<span class="fu">n</span>() <span class="sc">==</span> <span class="dv">2</span>) <span class="sc">%&gt;%</span> <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb136-12"><a href="#cb136-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(pregnancy_id, status) </span>
<span id="cb136-13"><a href="#cb136-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb136-14"><a href="#cb136-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot</span></span>
<span id="cb136-15"><a href="#cb136-15" aria-hidden="true" tabindex="-1"></a>sam_shan_del <span class="sc">%&gt;%</span> </span>
<span id="cb136-16"><a href="#cb136-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">fct_rev</span>(status), <span class="at">y =</span> Shannon)) <span class="sc">+</span></span>
<span id="cb136-17"><a href="#cb136-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">alpha =</span> <span class="fl">0.4</span>, <span class="at">outlier.shape =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb136-18"><a href="#cb136-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_quasirandom</span>(<span class="at">stroke =</span> <span class="cn">FALSE</span>, <span class="at">width =</span> <span class="fl">0.3</span>) <span class="sc">+</span></span>
<span id="cb136-19"><a href="#cb136-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>delivery_mode) <span class="sc">+</span></span>
<span id="cb136-20"><a href="#cb136-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_compare_means</span>(<span class="at">method =</span> <span class="st">"wilcox.test"</span>, <span class="at">paired =</span> <span class="cn">TRUE</span>) <span class="sc">+</span> </span>
<span id="cb136-21"><a href="#cb136-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Pregnancy status"</span>,</span>
<span id="cb136-22"><a href="#cb136-22" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Shannon diversity index"</span>) <span class="sc">+</span></span>
<span id="cb136-23"><a href="#cb136-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-80-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Your answer?</p>
<p>Bonus question: Try ordinating the samples in <code>ps_deliv</code> using Bray-Curtis distances. Color by the relative abundance of <em>Lactobacillus crispatus</em>, which I’ve included as a variable within the sample data (<code>sp_lacto_crispatus</code>), and also facet by <code>status</code>.</p>
<p>Note that dominant taxa are strong drivers of clustering among vaginal samples. Can we “turn down the volume” on these dominant taxa so we can see more of what’s going on here? Try applying a square-root transformation before the ordination step.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a>ps_deliv <span class="sc">%&gt;%</span> </span>
<span id="cb137-2"><a href="#cb137-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">transform_sample_counts</span>(., <span class="cf">function</span>(x) x<span class="sc">^</span>(<span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb137-3"><a href="#cb137-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ordinate</span>(., <span class="at">distance =</span> <span class="st">"bray"</span>, <span class="at">binary =</span> <span class="cn">FALSE</span>, <span class="at">method =</span> <span class="st">"MDS"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb137-4"><a href="#cb137-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_ordination</span>(ps_deliv, ., <span class="at">type =</span> <span class="st">"samples"</span>, <span class="at">color =</span> <span class="st">"sp_lacto_crispatus"</span>) <span class="sc">+</span></span>
<span id="cb137-5"><a href="#cb137-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span><span class="fu">fct_rev</span>(status))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="16S_part1_code_files/figure-html/unnamed-chunk-81-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Just as we transferred our alpha diversity measures to a dataframe, one can also transfer any number of taxon relative abundances to a dataframe as well.</p>
<p>That’s it for now. Nice work!! 🤓</p>
</section>
</section>
<section id="reproducibility-receipt" class="level2">
<h2 class="anchored" data-anchor-id="reproducibility-receipt">7. Reproducibility receipt</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.3 (2024-02-29)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.4.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Africa/Johannesburg
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] vegan_2.6-4      lattice_0.22-6   permute_0.9-7    ggpubr_0.6.0    
 [5] ggbeeswarm_0.7.2 here_1.0.1       pander_0.6.5     rstatix_0.7.2   
 [9] magrittr_2.0.3   lubridate_1.9.3  forcats_1.0.0    stringr_1.5.1   
[13] dplyr_1.1.4      purrr_1.0.2      readr_2.1.5      tidyr_1.3.1     
[17] tibble_3.2.1     ggplot2_3.5.1    tidyverse_2.0.0  phyloseq_1.46.0 

loaded via a namespace (and not attached):
 [1] ade4_1.7-22             tidyselect_1.2.1        vipor_0.4.7            
 [4] farver_2.1.1            Biostrings_2.70.3       bitops_1.0-7           
 [7] fastmap_1.1.1           RCurl_1.98-1.14         digest_0.6.35          
[10] timechange_0.3.0        lifecycle_1.0.4         cluster_2.1.6          
[13] survival_3.5-8          compiler_4.3.3          rlang_1.1.5            
[16] tools_4.3.3             igraph_2.0.3            utf8_1.2.4             
[19] yaml_2.3.8              data.table_1.15.4       ggsignif_0.6.4         
[22] knitr_1.49              labeling_0.4.3          htmlwidgets_1.6.4      
[25] plyr_1.8.9              abind_1.4-5             withr_3.0.0            
[28] BiocGenerics_0.48.1     grid_4.3.3              stats4_4.3.3           
[31] fansi_1.0.6             multtest_2.58.0         biomformat_1.30.0      
[34] colorspace_2.1-1        Rhdf5lib_1.24.2         scales_1.3.0           
[37] iterators_1.0.14        MASS_7.3-60.0.1         cli_3.6.3              
[40] rmarkdown_2.29          crayon_1.5.2            generics_0.1.3         
[43] rstudioapi_0.16.0       reshape2_1.4.4          tzdb_0.4.0             
[46] ape_5.8                 rhdf5_2.46.1            zlibbioc_1.48.2        
[49] splines_4.3.3           parallel_4.3.3          XVector_0.42.0         
[52] vctrs_0.6.5             Matrix_1.6-5            carData_3.0-5          
[55] jsonlite_1.8.8          car_3.1-2               IRanges_2.36.0         
[58] hms_1.1.3               S4Vectors_0.40.2        beeswarm_0.4.0         
[61] foreach_1.5.2           glue_1.8.0              codetools_0.2-20       
[64] stringi_1.8.3           gtable_0.3.4            GenomeInfoDb_1.38.8    
[67] munsell_0.5.1           pillar_1.9.0            htmltools_0.5.8.1      
[70] rhdf5filters_1.14.1     GenomeInfoDbData_1.2.11 R6_2.5.1               
[73] rprojroot_2.0.4         evaluate_0.23           Biobase_2.62.0         
[76] backports_1.4.1         broom_1.0.5             Rcpp_1.0.12            
[79] nlme_3.1-164            mgcv_1.9-1              xfun_0.50              
[82] fs_1.6.5                pkgconfig_2.0.3        </code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/elsherbini\.github\.io\/durban-data-science-for-biology\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>© 2023. Huge thanks to <a href="https://github.com/andrewpbray">Andrew P. Bray</a> and <a href="https://github.com/clauswilke">Claus O. Wilke</a> for open source code and sharing content.</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>This page is built <a href="https://quarto.org/">Quarto</a>.</p>
</div>
  </div>
</footer>




</body></html>