<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.551">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Data Science for Biology Workshop Series - Hypothesis testing, data transformation, compositional data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../">
<script src="../../../site_libs/quarto-html/quarto.js"></script>
<script src="../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<link href="../../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../../index.html">
    <span class="navbar-title">Data Science for Biology Workshop Series</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../../instance_ips.html"> 
<span class="menu-text">AWS Instance IPs</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../installation-instructions.html"> 
<span class="menu-text">Installation</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../datasets.html"> 
<span class="menu-text">Datasets</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../resources.html"> 
<span class="menu-text">Resources</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../past_workshops.html"> 
<span class="menu-text">Past Workshops</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/elsherbini/durban-data-science-for-biology"> <i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <a class="flex-grow-1 no-decor" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
          <h1 class="quarto-secondary-nav-title">Hypothesis testing, data transformation, compositional data</h1>
        </a>     
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false">
 <span class="menu-text"><em>1. Reproducible Data Analysis with R</em></span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/1-workshop1/0-welcome/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 0:</b> Welcome to the workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/1-workshop1/1-data-generation/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 1</b>: How data analysis is informed by data generation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/1-workshop1/2-intro-to-r/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 2</b>: Intro to R, Rstudio, and Quarto</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/1-workshop1/3-tidyverse-101/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 3</b>: Intro to data visualization and data wrangling with the <code>tidyverse</code></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/1-workshop1/4-tidyverse-201/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 4</b>: Data wrangling and more data visualization</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/1-workshop1/5-tableone/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 5</b>: <code>tableone</code> and its Basic Statistics</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/1-workshop1/6-hypothesis-testing/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 6</b>: Data transformations and more statistics</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/1-workshop1/7-custom-data-visualizations/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 7</b>: Customizing data visualizations using <code>colorspace</code>, <code>ggplot2</code>, and <code>patchwork</code></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/1-workshop1/8-group-projects/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Group Project</b> -Applying what you’ve learned to new data</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false">
 <span class="menu-text"><b>2. Working with High Dimensional Data in R - The Human Microbiome</b></span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/2-workshop2/0-welcome/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 0:</b> Welcome to the workshop</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/2-workshop2/1-library-prep/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 1</b>: Sequencing Library Preparation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/2-workshop2/2-r-refresher/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 2</b>: R Refresher</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/2-workshop2/3-16s_processing/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 3</b>: 16S Data Processing</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/2-workshop2/4-16s-pt1/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 4</b>: Exploring High Dimensional Data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/2-workshop2/5-16s-pt2/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 5</b>: Correlations and Ordinations</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/2-workshop2/6-virgo/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 6</b>: Shotgun Metagenomics and VIRGO</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/2-workshop2/7-keynote/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Module 7</b>: Keynote</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/2-workshop2/8-group-project/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Group Activity:</b> Applying what you’ve learned to new data</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <span class="sidebar-item-text sidebar-link text-start">
 <span class="menu-text"><em>3. Bioinformatics and Transcriptomics Data Analysis - Coming October 2024</em></span></span>
  </li>
    </ul>
    </div>
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#hypothesis-testing" id="toc-hypothesis-testing" class="nav-link active" data-scroll-target="#hypothesis-testing">Hypothesis testing</a>
  <ul class="collapse">
  <li><a href="#t-tests" id="toc-t-tests" class="nav-link" data-scroll-target="#t-tests"><span class="math inline">\(t\)</span>-tests</a>
  <ul class="collapse">
  <li><a href="#qq-plots-checking-compatibility-with-a-normal-distribution" id="toc-qq-plots-checking-compatibility-with-a-normal-distribution" class="nav-link" data-scroll-target="#qq-plots-checking-compatibility-with-a-normal-distribution">QQ-plots: checking compatibility with a (normal) distribution</a></li>
  </ul></li>
  <li><a href="#impact-of-sample-size-on-p-values" id="toc-impact-of-sample-size-on-p-values" class="nav-link" data-scroll-target="#impact-of-sample-size-on-p-values">Impact of sample size on p-values</a>
  <ul class="collapse">
  <li><a href="#small-dataset" id="toc-small-dataset" class="nav-link" data-scroll-target="#small-dataset">Small dataset</a></li>
  <li><a href="#large-datasets" id="toc-large-datasets" class="nav-link" data-scroll-target="#large-datasets">Large datasets</a></li>
  <li><a href="#clinical-vs-statistical-significance" id="toc-clinical-vs-statistical-significance" class="nav-link" data-scroll-target="#clinical-vs-statistical-significance">Clinical <em>vs</em> Statistical significance</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#data-transformation" id="toc-data-transformation" class="nav-link" data-scroll-target="#data-transformation">Data transformation</a>
  <ul class="collapse">
  <li><a href="#cytokine-data-exploration" id="toc-cytokine-data-exploration" class="nav-link" data-scroll-target="#cytokine-data-exploration">Cytokine data exploration</a></li>
  <li><a href="#log-transformation" id="toc-log-transformation" class="nav-link" data-scroll-target="#log-transformation">Log-transformation</a></li>
  <li><a href="#association-with-bv" id="toc-association-with-bv" class="nav-link" data-scroll-target="#association-with-bv">Association with BV</a></li>
  </ul></li>
  <li><a href="#non-parametric-tests" id="toc-non-parametric-tests" class="nav-link" data-scroll-target="#non-parametric-tests">Non-parametric tests</a>
  <ul class="collapse">
  <li><a href="#the-wilcoxon-rank-sum-test" id="toc-the-wilcoxon-rank-sum-test" class="nav-link" data-scroll-target="#the-wilcoxon-rank-sum-test">The Wilcoxon rank-sum test</a></li>
  </ul></li>
  <li><a href="#multiple-testing" id="toc-multiple-testing" class="nav-link" data-scroll-target="#multiple-testing">Multiple testing</a>
  <ul class="collapse">
  <li><a href="#simulations" id="toc-simulations" class="nav-link" data-scroll-target="#simulations">Simulations</a></li>
  <li><a href="#cytokines" id="toc-cytokines" class="nav-link" data-scroll-target="#cytokines">Cytokines</a></li>
  </ul></li>
  <li><a href="#displaying-longitudinal-data" id="toc-displaying-longitudinal-data" class="nav-link" data-scroll-target="#displaying-longitudinal-data">Displaying longitudinal data</a>
  <ul class="collapse">
  <li><a href="#paired-tests" id="toc-paired-tests" class="nav-link" data-scroll-target="#paired-tests">Paired tests</a></li>
  </ul></li>
  <li><a href="#compositional-data" id="toc-compositional-data" class="nav-link" data-scroll-target="#compositional-data">Compositional data</a>
  <ul class="collapse">
  <li><a href="#flow-cytometry-data" id="toc-flow-cytometry-data" class="nav-link" data-scroll-target="#flow-cytometry-data">Flow cytometry data</a></li>
  </ul></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">Hypothesis testing, data transformation, compositional data</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse) <span class="co"># loads the packages from the tidyverse suite</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pander) <span class="co"># allows to display pretty tables </span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork) <span class="co"># allows to combine ggplot (see Module 7)</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_light</span>())</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>) <span class="co"># set the "random seed" for reproducibility</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="hypothesis-testing" class="level1">
<h1>Hypothesis testing</h1>
<p>Before testing hypotheses on real data, let’s develop our intuition on data we simulate ourselves.</p>
<p>For this first example, let’s simulate 3 datasets by sampling 50 values from 2 normal populations with different means but same variance (So two datasets are drawn from population 1 and one is drawn from population 2). Let’s use <span class="math inline">\(\mu_1 = 0\)</span>, <span class="math inline">\(\mu_2 = 2\)</span>, and <span class="math inline">\(\sigma^2 = 4\)</span>.</p>
<div class="callout callout-style-simple callout-tip no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<p>To simulate normally distributed data, check the <code>rnorm</code> function by typing <code>?rnorm</code> in the console.</p>
<p>We put the results of our simulations in a variable <code>X</code> that will be a “long” tibble with the following two columns: <code>dataset</code>, <code>value</code>.</p>
</div>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>n_samples <span class="ot">&lt;-</span> <span class="dv">50</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>common_sd <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="co"># square root of 4 (we said)</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>sample_1 <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(n_samples, <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> common_sd) </span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>sample_2 <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(n_samples, <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> common_sd) </span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>sample_3 <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(n_samples, <span class="at">mean =</span> <span class="dv">2</span>, <span class="at">sd =</span> common_sd)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>X <span class="ot">&lt;-</span> </span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_rows</span>(</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">tibble</span>(<span class="at">dataset =</span> <span class="dv">1</span>, <span class="at">value =</span> sample_1),</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">tibble</span>(<span class="at">dataset =</span> <span class="dv">2</span>, <span class="at">value =</span> sample_2),</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">tibble</span>(<span class="at">dataset =</span> <span class="dv">3</span>, <span class="at">value =</span> sample_3)</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">dataset =</span> dataset <span class="sc">|&gt;</span> <span class="fu">factor</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>If we display the top 6 raws, we have:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># the head function takes the top x rows of a table</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co"># by default, the first 6 rows</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co"># pander just prints the result in a pretty way</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>X <span class="sc">|&gt;</span> <span class="fu">head</span>() <span class="sc">|&gt;</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 13%">
<col style="width: 13%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">dataset</th>
<th style="text-align: center;">value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">1</td>
<td style="text-align: center;">-1.121</td>
</tr>
<tr class="even">
<td style="text-align: center;">1</td>
<td style="text-align: center;">-0.4604</td>
</tr>
<tr class="odd">
<td style="text-align: center;">1</td>
<td style="text-align: center;">3.117</td>
</tr>
<tr class="even">
<td style="text-align: center;">1</td>
<td style="text-align: center;">0.141</td>
</tr>
<tr class="odd">
<td style="text-align: center;">1</td>
<td style="text-align: center;">0.2586</td>
</tr>
<tr class="even">
<td style="text-align: center;">1</td>
<td style="text-align: center;">3.43</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>and the bottom 6 raws are:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># tail prints the last 6 rows</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>X <span class="sc">|&gt;</span> <span class="fu">tail</span>() <span class="sc">|&gt;</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 13%">
<col style="width: 13%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">dataset</th>
<th style="text-align: center;">value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">3</td>
<td style="text-align: center;">-1.203</td>
</tr>
<tr class="even">
<td style="text-align: center;">3</td>
<td style="text-align: center;">0.9382</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3</td>
<td style="text-align: center;">-0.9235</td>
</tr>
<tr class="even">
<td style="text-align: center;">3</td>
<td style="text-align: center;">3.376</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3</td>
<td style="text-align: center;">6.2</td>
</tr>
<tr class="even">
<td style="text-align: center;">3</td>
<td style="text-align: center;">-0.5741</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>To validate that we’ve simulated our data properly, let’s compute the mean and variance of our data and display a histogram for each dataset.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>X <span class="sc">%&gt;%</span> </span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(dataset) <span class="sc">%&gt;%</span> </span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">mean =</span> <span class="fu">mean</span>(value), <span class="at">var =</span> <span class="fu">var</span>(value)) <span class="sc">|&gt;</span> </span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 13%">
<col style="width: 13%">
<col style="width: 13%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">dataset</th>
<th style="text-align: center;">mean</th>
<th style="text-align: center;">var</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">1</td>
<td style="text-align: center;">0.06881</td>
<td style="text-align: center;">3.429</td>
</tr>
<tr class="even">
<td style="text-align: center;">2</td>
<td style="text-align: center;">0.2928</td>
<td style="text-align: center;">3.279</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3</td>
<td style="text-align: center;">1.492</td>
<td style="text-align: center;">3.915</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Do these values make sense?</p>
<p>We can display the distributions we just simulated using <code>ggplot</code> and <code>geom_histogram</code>.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(X, <span class="fu">aes</span>(<span class="at">x =</span> value, <span class="at">fill =</span> dataset)) <span class="sc">+</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(dataset <span class="sc">~</span> ., <span class="at">labeller =</span> label_both) <span class="sc">+</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">fill =</span> <span class="st">"none"</span>) <span class="sc">+</span> <span class="co"># we don't really need a fill legend </span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">strip.text.y =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">0</span>, <span class="at">hjust =</span> <span class="dv">0</span>)) <span class="co"># sometimes it is nicer to rotate the panel labels so they are horizontal and more easily readable</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="module-6-hands-on_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Visually, it looks like the <strong>means</strong> of datasets 1 and 2 are similar, but the <strong>mean</strong> of dataset 3 is larger (expected based on how we simulated the data). However, because there is large variability in the data and a lot of overlap between the values of these datasets, we want to make sure that what we think we observe is not due to chance.</p>
<p>This what statistical tests help us do. Feel free to review the workshop slides if you need a reminder of what statistical tests are.</p>
<p>Here, we want to do a test on the <strong>means</strong> of the datasets and test if the means of datasets 2 or 3 are different from the mean of dataset 1.</p>
<section id="t-tests" class="level2">
<h2 class="anchored" data-anchor-id="t-tests"><span class="math inline">\(t\)</span>-tests</h2>
<p>Because we have more than 40 samples in each dataset, we can use a <span class="math inline">\(t\)</span>-test.</p>
<p>In <code>R</code>, <span class="math inline">\(t\)</span>-tests can be done with the <code>t.test</code> function. Note that there are several ways to use the <code>t.test</code> function and several options we need to be careful about. To read more about these options, type <code>?t.test</code> in your console.</p>
<p>Specifically, we need to be careful about specifying what our Null and alternative hypotheses are.</p>
<p>This is specified using the <code>alternative</code> option of the <code>t.test</code> function.</p>
<p>The option corresponding to</p>
<span class="math display">\[\begin{align*}
H_0:&amp; \ \mu_1 = \mu_2 \\
H_a:&amp; \ \mu_1 \neq \mu_2
\end{align*}\]</span>
<p>is <code>alternative = "two-sided"</code> which is the default option of <code>t.test</code> (that is, if we don’t specify the <code>alternative</code>, the function automatically assumes that we want to perform a “two-sided” test). It is <code>"two-sided"</code> because <span class="math inline">\(\mu_2\)</span> can be smaller OR larger in the alternative hypothesis.</p>
<p>Let’s say that we do not have any <em>a priori</em> on the alternative and perform a two-sided test on the means of datasets 1 and 2.<br>
</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(value <span class="sc">~</span> dataset, <span class="at">data =</span> X <span class="sc">|&gt;</span> <span class="fu">filter</span>(dataset <span class="sc">%in%</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
    Welch Two Sample t-test

data:  value by dataset
t = -0.61157, df = 97.951, p-value = 0.5422
alternative hypothesis: true difference in means between group 1 and group 2 is not equal to 0
95 percent confidence interval:
 -0.9508970  0.5028781
sample estimates:
mean in group 1 mean in group 2 
      0.0688071       0.2928165 </code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this is the same as :</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="co"># t.test(x = X$value[X$dataset == 1], y = X$value[X$dataset == 2])</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We see that the probability to observe the <span class="math inline">\(t\)</span> value is quite large. So, we do NOT reject the null hypothesis that the two means are the same.</p>
<p>Now, if we do this test comparing the means of datasets 1 and <strong>3</strong>, what do we get? What do you conclude?</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(value <span class="sc">~</span> dataset, <span class="at">data =</span> X <span class="sc">|&gt;</span> <span class="fu">filter</span>(dataset <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
    Welch Two Sample t-test

data:  value by dataset
t = -3.714, df = 97.572, p-value = 0.0003399
alternative hypothesis: true difference in means between group 1 and group 3 is not equal to 0
95 percent confidence interval:
 -2.1839829 -0.6628012
sample estimates:
mean in group 1 mean in group 3 
      0.0688071       1.4921991 </code></pre>
</div>
</div>
<p>As explained above, the tests we have done so far are “two-sided”. That means that the null hypothesis is that the two means are the same, and the alternative is that they are different.</p>
<p><span class="math display">\[
H_0: \mu_1 = \mu_2
\]</span></p>
<p><span class="math display">\[
H_a: \mu_1 \neq \mu_2
\]</span></p>
<p>But sometimes, we have an <em>a priori</em> that the alternative is that one of the two means is larger. This is where we need a “one-sided” test.</p>
<p>We do this in <code>R</code> with the <code>alternative</code> option of the <code>t.test</code> function.</p>
<p>Remember: to obtain the documentation about the <code>t.test</code> function, you can type <code>?t.test</code> in the console.</p>
<p>How do we need to change the <code>alternative</code> option to perform the test corresponding to this pair of hypotheses ?</p>
<span class="math display">\[\begin{align*}
H_0:&amp;\  \mu_1 \geq \mu_3 \\
H_a:&amp;\  \mu_1 &lt; \mu_3 ( \iff \mu_1 - \mu_3 &lt; 0 )
\end{align*}\]</span>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(value <span class="sc">~</span> dataset, <span class="at">data =</span> X <span class="sc">|&gt;</span> <span class="fu">filter</span>(dataset <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>)),</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>       <span class="at">alternative =</span> <span class="st">"less"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
    Welch Two Sample t-test

data:  value by dataset
t = -3.714, df = 97.572, p-value = 0.0001699
alternative hypothesis: true difference in means between group 1 and group 3 is less than 0
95 percent confidence interval:
       -Inf -0.7869575
sample estimates:
mean in group 1 mean in group 3 
      0.0688071       1.4921991 </code></pre>
</div>
</div>
<p>How do the <span class="math inline">\(t\)</span> and <span class="math inline">\(p\)</span> values compare in the two-sided <em>vs</em> one-sided test?<br>
Why?</p>
<section id="qq-plots-checking-compatibility-with-a-normal-distribution" class="level3">
<h3 class="anchored" data-anchor-id="qq-plots-checking-compatibility-with-a-normal-distribution">QQ-plots: checking compatibility with a (normal) distribution</h3>
<p>Remember that the <span class="math inline">\(t\)</span>-test requires, for small sample sizes, for the data to be drawn from a normal distribution. Usually, with real data, we don’t always know what distribution the data are samples from.</p>
<p>We can always check that our data is <em>compatible</em> with a normal distribution by performing a QQ-plot:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>example_data <span class="ot">&lt;-</span> </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tibble</span>(<span class="at">norm =</span> <span class="fu">rnorm</span>(<span class="dv">20</span>), <span class="at">lnorm =</span> <span class="fu">rlnorm</span>(<span class="dv">20</span>))</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="co"># just like `rnorm` samples from a normal distribution, </span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="co"># `rlnorm` samples from a log-normal distribution</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>g_norm <span class="ot">&lt;-</span> </span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(example_data, <span class="fu">aes</span>(<span class="at">sample =</span> norm)) <span class="sc">+</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_qq</span>() <span class="sc">+</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_qq_line</span>() <span class="sc">+</span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Sampled from a</span><span class="sc">\n</span><span class="st">normal distribution"</span>)</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>g_lnorm <span class="ot">&lt;-</span> </span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(example_data, <span class="fu">aes</span>(<span class="at">sample =</span> lnorm)) <span class="sc">+</span></span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_qq</span>() <span class="sc">+</span></span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_qq_line</span>() <span class="sc">+</span></span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Sampled from a</span><span class="sc">\n</span><span class="st">*log*normal distribution"</span>)</span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>g_norm <span class="sc">+</span> g_lnorm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="module-6-hands-on_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We see that, on the left plot, the dots are close to the line, and on the right plot, some dots are very far from the line. These dots far from the line are a warning that the distribution is probably not normal.</p>
</section>
</section>
<section id="impact-of-sample-size-on-p-values" class="level2">
<h2 class="anchored" data-anchor-id="impact-of-sample-size-on-p-values">Impact of sample size on p-values</h2>
<section id="small-dataset" class="level3">
<h3 class="anchored" data-anchor-id="small-dataset">Small dataset</h3>
<p>Let’s re-do our analysis but decrease the sample size to 10 samples per dataset.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># another way to simulate data is to do the following</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="co"># check what the expand_grid and rowwise function do</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>X_small_sample <span class="ot">&lt;-</span> </span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expand_grid</span>(</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">dataset =</span> (<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>) <span class="sc">|&gt;</span> <span class="fu">factor</span>(), </span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">sample =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">pop_true_mean =</span> <span class="fu">ifelse</span>(dataset <span class="sc">==</span> <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">0</span>),</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">pop_true_var =</span> <span class="dv">2</span></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rowwise</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">value =</span> </span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>      <span class="fu">rnorm</span>(<span class="dv">1</span>, <span class="at">mean =</span> pop_true_mean, <span class="at">sd =</span> <span class="fu">sqrt</span>(pop_true_var))</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>()</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>X_small_sample <span class="sc">|&gt;</span> <span class="fu">head</span>() <span class="sc">|&gt;</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 13%">
<col style="width: 12%">
<col style="width: 22%">
<col style="width: 20%">
<col style="width: 13%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">dataset</th>
<th style="text-align: center;">sample</th>
<th style="text-align: center;">pop_true_mean</th>
<th style="text-align: center;">pop_true_var</th>
<th style="text-align: center;">value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">1</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">0.3033</td>
</tr>
<tr class="even">
<td style="text-align: center;">1</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">-0.4592</td>
</tr>
<tr class="odd">
<td style="text-align: center;">1</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">0.1338</td>
</tr>
<tr class="even">
<td style="text-align: center;">1</td>
<td style="text-align: center;">4</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">-1.266</td>
</tr>
<tr class="odd">
<td style="text-align: center;">1</td>
<td style="text-align: center;">5</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">-1.854</td>
</tr>
<tr class="even">
<td style="text-align: center;">1</td>
<td style="text-align: center;">6</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">2.824</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(X_small_sample, <span class="fu">aes</span>(<span class="at">x =</span> value, <span class="at">fill =</span> dataset)) <span class="sc">+</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(dataset <span class="sc">~</span> ., <span class="at">labeller =</span> label_both) <span class="sc">+</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">fill =</span> <span class="st">"none"</span>) <span class="sc">+</span> </span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">strip.text.y =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">0</span>, <span class="at">hjust =</span> <span class="dv">0</span>)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="module-6-hands-on_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can still use a <span class="math inline">\(t\)</span>-test because we <em>know</em> that our samples are drawn from a Normal distribution.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(value <span class="sc">~</span> dataset, </span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>       <span class="at">data =</span> X_small_sample <span class="sc">|&gt;</span> <span class="fu">filter</span>(dataset <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
    Welch Two Sample t-test

data:  value by dataset
t = -3.2548, df = 17.988, p-value = 0.004402
alternative hypothesis: true difference in means between group 1 and group 3 is not equal to 0
95 percent confidence interval:
 -3.4653702 -0.7465168
sample estimates:
mean in group 1 mean in group 3 
      -0.377852        1.728091 </code></pre>
</div>
</div>
<p>We see that the <span class="math inline">\(p\)</span>-value is now much larger because, with the small datasets, we have a lot more uncertainty on the true means of the populations.</p>
</section>
<section id="large-datasets" class="level3">
<h3 class="anchored" data-anchor-id="large-datasets">Large datasets</h3>
<p>Let’s now redo the same again but with a very large sample size for each dataset. For example N = 1000.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>X_large_sample <span class="ot">&lt;-</span> </span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expand_grid</span>(</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">dataset =</span> (<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>) <span class="sc">|&gt;</span> <span class="fu">factor</span>(), </span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">sample =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">1000</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span> </span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">pop_true_mean =</span> <span class="fu">ifelse</span>(dataset <span class="sc">==</span> <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">0</span>),</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">pop_true_var =</span> <span class="dv">2</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rowwise</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">value =</span> </span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>      <span class="fu">rnorm</span>(<span class="dv">1</span>, <span class="at">mean =</span> pop_true_mean, <span class="at">sd =</span> <span class="fu">sqrt</span>(pop_true_var))</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>()</span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a>X_large_sample <span class="sc">|&gt;</span> <span class="fu">head</span>() <span class="sc">|&gt;</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 13%">
<col style="width: 12%">
<col style="width: 22%">
<col style="width: 20%">
<col style="width: 13%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">dataset</th>
<th style="text-align: center;">sample</th>
<th style="text-align: center;">pop_true_mean</th>
<th style="text-align: center;">pop_true_var</th>
<th style="text-align: center;">value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">1</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">-0.8117</td>
</tr>
<tr class="even">
<td style="text-align: center;">1</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">0.874</td>
</tr>
<tr class="odd">
<td style="text-align: center;">1</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">1.57</td>
</tr>
<tr class="even">
<td style="text-align: center;">1</td>
<td style="text-align: center;">4</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">1.001</td>
</tr>
<tr class="odd">
<td style="text-align: center;">1</td>
<td style="text-align: center;">5</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">-0.5143</td>
</tr>
<tr class="even">
<td style="text-align: center;">1</td>
<td style="text-align: center;">6</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">0.0845</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(X_large_sample, <span class="fu">aes</span>(<span class="at">x =</span> value, <span class="at">fill =</span> dataset)) <span class="sc">+</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(dataset <span class="sc">~</span> ., <span class="at">labeller =</span> label_both) <span class="sc">+</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">fill =</span> <span class="st">"none"</span>) <span class="sc">+</span> </span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">strip.text.y =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">0</span>, <span class="at">hjust =</span> <span class="dv">0</span>)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="module-6-hands-on_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(value <span class="sc">~</span> dataset, </span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>       <span class="at">data =</span> X_large_sample <span class="sc">|&gt;</span> <span class="fu">filter</span>(dataset <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
    Welch Two Sample t-test

data:  value by dataset
t = -30.732, df = 1997.9, p-value &lt; 2.2e-16
alternative hypothesis: true difference in means between group 1 and group 3 is not equal to 0
95 percent confidence interval:
 -2.075287 -1.826308
sample estimates:
mean in group 1 mean in group 3 
     0.03031352      1.98111096 </code></pre>
</div>
</div>
<p>The <span class="math inline">\(p\)</span>-value is as small as it can be.</p>
<p>Let’s also re-do the test comparing datasets 1 and 2</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(value <span class="sc">~</span> dataset, </span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>       <span class="at">data =</span> X_large_sample <span class="sc">|&gt;</span> <span class="fu">filter</span>(dataset <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
    Welch Two Sample t-test

data:  value by dataset
t = 0.075717, df = 1996.9, p-value = 0.9397
alternative hypothesis: true difference in means between group 1 and group 2 is not equal to 0
95 percent confidence interval:
 -0.1186693  0.1282006
sample estimates:
mean in group 1 mean in group 2 
     0.03031352      0.02554787 </code></pre>
</div>
</div>
<p>We still don’t reject the null hypothesis, which is what we expect.</p>
</section>
<section id="clinical-vs-statistical-significance" class="level3">
<h3 class="anchored" data-anchor-id="clinical-vs-statistical-significance">Clinical <em>vs</em> Statistical significance</h3>
<p>However, sometimes, we need to be careful with large sample sizes because tiny effects can still lead to very small p-values. However, these tiny effects don’t have much but these do not have much <em>clinical</em> relevance.</p>
<p>Let’s verify that with a difference in means of 0.2</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">rnorm</span>(<span class="dv">1000</span>, <span class="at">mean =</span> <span class="fl">0.0</span>, <span class="at">sd =</span> <span class="dv">2</span>),</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> <span class="fu">rnorm</span>(<span class="dv">1000</span>, <span class="at">mean =</span> <span class="fl">0.2</span>, <span class="at">sd =</span> <span class="dv">2</span>)</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
    Welch Two Sample t-test

data:  rnorm(1000, mean = 0, sd = 2) and rnorm(1000, mean = 0.2, sd = 2)
t = -2.18, df = 1998, p-value = 0.02937
alternative hypothesis: true difference in means is not equal to 0
95 percent confidence interval:
 -0.36547886 -0.01931448
sample estimates:
  mean of x   mean of y 
-0.02210611  0.17029056 </code></pre>
</div>
</div>
<p>The <span class="math inline">\(p\)</span>-value is lower than 1/20, but the effect (<em>i.e.</em>, the mean in differences), compared to the standard deviations is very small.</p>
<p>This is what we are talking about when discussing the “clinical” <em>vs</em> “statistical” significance.</p>
</section>
</section>
</section>
<section id="data-transformation" class="level1">
<h1>Data transformation</h1>
<p>Now, let’s play with the datasets provided on the workshop website. Specifically, we are interested in analyzing the cytokines data.</p>
<section id="cytokine-data-exploration" class="level2">
<h2 class="anchored" data-anchor-id="cytokine-data-exploration">Cytokine data exploration</h2>
<p>Our first task will be to display the distribution of the “IL-1<span class="math inline">\(\beta\)</span>” cytokine.</p>
<p>To do so, we load the cytokine data and filter for the cytokine of interest.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># notice where I had stored my files.</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Make sure to modify the path to your files</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>elisa <span class="ot">&lt;-</span> </span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">read_csv</span>(</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="st">"data/03_elisa_cytokines_UKZN_workshop_2023.csv"</span>,</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_col_types =</span> <span class="cn">FALSE</span></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>IL1b <span class="ot">&lt;-</span> elisa <span class="sc">|&gt;</span> <span class="fu">filter</span>(cytokine <span class="sc">==</span> <span class="st">"IL-1b"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We display its distribution with the <code>geom_histogram</code> function, and we can use the color (<code>fill</code>) option to flag whether the concentration value was within or out of the limits of detection.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>cytokine_lim_cols <span class="ot">&lt;-</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">"within limits"</span> <span class="ot">=</span> <span class="st">"navy"</span>, <span class="st">"out of range"</span> <span class="ot">=</span> <span class="st">"indianred1"</span>)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(IL1b, <span class="fu">aes</span>(<span class="at">x =</span> conc, <span class="at">fill =</span> limits)) <span class="sc">+</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"IL-1b concentrations"</span>) <span class="sc">+</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> cytokine_lim_cols)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="module-6-hands-on_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>What do we observe?</p>
<p>Do we think that the distribution of IL-1<span class="math inline">\(\beta\)</span> concentations follow a normal distribution?</p>
<p>Let’s still check with a QQ-plot:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  IL1b, </span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aes</span>(<span class="at">sample =</span> conc)</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_qq_line</span>() <span class="sc">+</span></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_qq</span>(<span class="at">size =</span> <span class="fl">0.75</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="module-6-hands-on_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Look back at the QQ-plots we did earlier on simulated data. Does this QQ-plot look like one of the simulated one?</p>
</section>
<section id="log-transformation" class="level2">
<h2 class="anchored" data-anchor-id="log-transformation">Log-transformation</h2>
<p>Let’s now repeat the last two displays, using the log of the concentration instead of the concentration itself. Remember, we can use the <code>mutate</code> function to add a new variable to our <code>data.frame</code>.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>IL1b <span class="ot">&lt;-</span> IL1b <span class="sc">|&gt;</span> <span class="fu">mutate</span>(<span class="at">logconc =</span> <span class="fu">log10</span>(conc))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(IL1b, <span class="fu">aes</span>(<span class="at">x =</span> logconc, <span class="at">fill =</span> limits)) <span class="sc">+</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"IL-1b concentrations (log 10)"</span>) <span class="sc">+</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"log10(conc)"</span>) <span class="sc">+</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> cytokine_lim_cols)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="module-6-hands-on_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Now, the “within-range” data is more compatible with a Normal distribution.</p>
<p>Let’s check with a QQ-plot:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>  IL1b, </span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aes</span>(<span class="at">sample =</span> logconc)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_qq_line</span>() <span class="sc">+</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_qq</span>(<span class="at">size =</span> <span class="fl">0.75</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="module-6-hands-on_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>It is not perfect, but it is much better.</p>
</section>
<section id="association-with-bv" class="level2">
<h2 class="anchored" data-anchor-id="association-with-bv">Association with BV</h2>
<p>Now, we are interested in testing whether a BV (Bacterial Vaginosis) diagnosis is associated with different mean concentrations of IL-1<span class="math inline">\(\beta\)</span>.</p>
<p>Remember, the BV diagnosis is contained in the Clinical Measurements table. So we first need to load that table in <code>R</code>.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>clin <span class="ot">&lt;-</span> </span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">read_csv</span>(</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"data/02_visit_clinical_measurements_UKZN_workshop_2023.csv"</span>,</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_col_types =</span> <span class="cn">FALSE</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>The first 6 rows of the clinical data are:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>clin <span class="sc">|&gt;</span> <span class="fu">head</span>() <span class="sc">|&gt;</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 12%">
<col style="width: 18%">
<col style="width: 13%">
<col style="width: 20%">
<col style="width: 16%">
<col style="width: 8%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">pid</th>
<th style="text-align: center;">time_point</th>
<th style="text-align: center;">arm</th>
<th style="text-align: center;">nugent_score</th>
<th style="text-align: center;">crp_blood</th>
<th style="text-align: center;">ph</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">pid_01</td>
<td style="text-align: center;">baseline</td>
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">8</td>
<td style="text-align: center;">0.44</td>
<td style="text-align: center;">5.7</td>
</tr>
<tr class="even">
<td style="text-align: center;">pid_01</td>
<td style="text-align: center;">week_1</td>
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">7</td>
<td style="text-align: center;">1.66</td>
<td style="text-align: center;">5.2</td>
</tr>
<tr class="odd">
<td style="text-align: center;">pid_01</td>
<td style="text-align: center;">week_7</td>
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">7</td>
<td style="text-align: center;">1.44</td>
<td style="text-align: center;">5.4</td>
</tr>
<tr class="even">
<td style="text-align: center;">pid_02</td>
<td style="text-align: center;">baseline</td>
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">7</td>
<td style="text-align: center;">1.55</td>
<td style="text-align: center;">5.2</td>
</tr>
<tr class="odd">
<td style="text-align: center;">pid_02</td>
<td style="text-align: center;">week_1</td>
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">7</td>
<td style="text-align: center;">0.75</td>
<td style="text-align: center;">4.8</td>
</tr>
<tr class="even">
<td style="text-align: center;">pid_02</td>
<td style="text-align: center;">week_7</td>
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">4</td>
<td style="text-align: center;">1.17</td>
<td style="text-align: center;">4.2</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>There are several ways to diagnose BV. One of them is to check whether a person has a Nugent score of 7 or more.</p>
<p>Let’s create a new <code>BV</code> variable that says whether a person was diagnosed with BV at each visit.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>clin <span class="ot">&lt;-</span> </span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>  clin <span class="sc">|&gt;</span> </span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">BV =</span> <span class="fu">ifelse</span>(nugent_score <span class="sc">&gt;=</span> <span class="dv">7</span>, <span class="st">"BV"</span>, <span class="st">"Healthy"</span>)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Since we want to compare the IL-1<span class="math inline">\(\beta\)</span> concentrations of individuals with or without BV, we need to join the clinical data with the IL-1<span class="math inline">\(\beta\)</span> concentration data.</p>
<p>Hint: to do so, we need a table describing how the sample IDs and participant x visit IDs are linked together - this info is in the “Sample ID” table.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>sample_info <span class="ot">&lt;-</span> </span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">read_csv</span>(</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="st">"data/00_sample_ids_UKZN_workshop_2023.csv"</span>,</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_col_types =</span> <span class="cn">FALSE</span></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>The first 6 rows of the sample info table are:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>sample_info <span class="sc">|&gt;</span> <span class="fu">head</span>() <span class="sc">|&gt;</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 12%">
<col style="width: 18%">
<col style="width: 16%">
<col style="width: 16%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">pid</th>
<th style="text-align: center;">time_point</th>
<th style="text-align: center;">arm</th>
<th style="text-align: center;">sample_id</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">pid_17</td>
<td style="text-align: center;">baseline</td>
<td style="text-align: center;">treatment</td>
<td style="text-align: center;">SAMP001</td>
</tr>
<tr class="even">
<td style="text-align: center;">pid_22</td>
<td style="text-align: center;">week_1</td>
<td style="text-align: center;">treatment</td>
<td style="text-align: center;">SAMP002</td>
</tr>
<tr class="odd">
<td style="text-align: center;">pid_25</td>
<td style="text-align: center;">week_1</td>
<td style="text-align: center;">treatment</td>
<td style="text-align: center;">SAMP003</td>
</tr>
<tr class="even">
<td style="text-align: center;">pid_41</td>
<td style="text-align: center;">week_1</td>
<td style="text-align: center;">treatment</td>
<td style="text-align: center;">SAMP004</td>
</tr>
<tr class="odd">
<td style="text-align: center;">pid_44</td>
<td style="text-align: center;">week_7</td>
<td style="text-align: center;">treatment</td>
<td style="text-align: center;">SAMP005</td>
</tr>
<tr class="even">
<td style="text-align: center;">pid_16</td>
<td style="text-align: center;">week_1</td>
<td style="text-align: center;">treatment</td>
<td style="text-align: center;">SAMP006</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We see that we can use the <code>sample_id</code> column to bind the cytokine concentrations with the <code>sample_info</code> table, then bind on the participant and visit ID with the clinical data.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>IL1b <span class="ot">&lt;-</span> </span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  IL1b <span class="sc">|&gt;</span> </span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(sample_info, <span class="at">by =</span> <span class="fu">join_by</span>(sample_id)) <span class="sc">|&gt;</span> </span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(clin, <span class="at">by =</span> <span class="fu">join_by</span>(pid, time_point, arm))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Now that our data are joined, let’s display the histogram of IL-1<span class="math inline">\(\beta\)</span> concentrations, including those out-of-range, colored by BV diagnosis.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>BV_colors <span class="ot">&lt;-</span> </span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">"BV"</span> <span class="ot">=</span> <span class="st">"darkgoldenrod1"</span>, <span class="st">"Healthy"</span> <span class="ot">=</span> <span class="st">"cornflowerblue"</span>)</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  IL1b, <span class="co">#|&gt; filter(limits == "within limits"),</span></span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aes</span>(<span class="at">x =</span> logconc, <span class="at">fill =</span> BV)</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">30</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">position =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"log10(conc)"</span>) <span class="sc">+</span></span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"IL-1b concentration by BV diagnosis"</span>) <span class="sc">+</span></span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> BV_colors)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="module-6-hands-on_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>From this visualization, it looks like the distribution is lower in healthy participants than in participants with BV.</p>
<p>What do we think of the “out-of-range” values? Should we include them in our analysis?</p>
<p>Let’s do a statistical test to see if these differences in concentrations could have happened by chance.</p>
<p>Since our data, once log-transformed, look normal, (with the exclusion of the out-of-range samples) it makes sense to use a <span class="math inline">\(t\)</span>-test. We can also check how many samples we have in each group to check if, regardless of the underlying distribution, there are enough samples to use a <span class="math inline">\(t\)</span>-test.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>IL1b <span class="sc">|&gt;</span> </span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(BV) <span class="sc">|&gt;</span> </span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">table</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>BV
     BV Healthy 
     46      86 </code></pre>
</div>
</div>
<p>We have many samples, so we could, regardless of the underlying distribution, use a <span class="math inline">\(t\)</span>-test.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(logconc <span class="sc">~</span> BV, <span class="at">data =</span> IL1b)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
    Welch Two Sample t-test

data:  logconc by BV
t = 6.5449, df = 116.4, p-value = 1.672e-09
alternative hypothesis: true difference in means between group BV and group Healthy is not equal to 0
95 percent confidence interval:
 0.5332052 0.9959381
sample estimates:
     mean in group BV mean in group Healthy 
            1.1245273             0.3599557 </code></pre>
</div>
</div>
<p>What do you conclude?</p>
</section>
</section>
<section id="non-parametric-tests" class="level1">
<h1>Non-parametric tests</h1>
<p>So far, we could use the <span class="math inline">\(t\)</span>-test to make inference on the means of different populations because we either had more than 40 samples in each group or we observed or <em>knew</em> that the samples were drawn from a normal distribution.</p>
<p>So, when our data is small and does not appear to be drawn from a normal distribution, we cannot use the <span class="math inline">\(t\)</span>-test that assumes normality of the underlying data or requires large enough sample sizes.</p>
<p>The <span class="math inline">\(t\)</span>-test is part of the family of <em>parametric</em> tests because they assume that the underlying data follows a specific distribution which can be characterized by <em>parameters</em>. For example, the parameters of a normal distribution are the mean and the variance.</p>
<p>There are also <em>non-parametric</em> tests which makes no assumption on the distribution of the data.</p>
<section id="the-wilcoxon-rank-sum-test" class="level2">
<h2 class="anchored" data-anchor-id="the-wilcoxon-rank-sum-test">The Wilcoxon rank-sum test</h2>
<p>The Wilcoxon rank-sum test is a non-parametric test to compare two independent datasets. The null hypothesis is that, for randomly selected values <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span> from two populations, the probability of <span class="math inline">\(X\)</span> being greater than <span class="math inline">\(Y\)</span> is equal to the probability of <span class="math inline">\(Y\)</span> being greater than <span class="math inline">\(X\)</span> (see <a href="https://en.wikipedia.org/wiki/Mann–Whitney_U_test">Wikipedia</a>), regardless of the distribution <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span> are drawn from.</p>
<p>Let’s try the test on our small dataset:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wilcox.test</span>(value <span class="sc">~</span> dataset, </span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">data =</span> X_small_sample <span class="sc">|&gt;</span> <span class="fu">filter</span>(dataset <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
    Wilcoxon rank sum exact test

data:  value by dataset
W = 13, p-value = 0.003886
alternative hypothesis: true location shift is not equal to 0</code></pre>
</div>
</div>
<p>And compare it to the <span class="math inline">\(t\)</span>-test (we can because we know this data is drawn from a Normal).</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(value <span class="sc">~</span> dataset, </span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>       <span class="at">data =</span> X_small_sample <span class="sc">|&gt;</span> <span class="fu">filter</span>(dataset <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
    Welch Two Sample t-test

data:  value by dataset
t = -3.2548, df = 17.988, p-value = 0.004402
alternative hypothesis: true difference in means between group 1 and group 3 is not equal to 0
95 percent confidence interval:
 -3.4653702 -0.7465168
sample estimates:
mean in group 1 mean in group 3 
      -0.377852        1.728091 </code></pre>
</div>
</div>
<p>We see that both provides small probabilites to observe the data under the null hypothesis.</p>
<p>So, why not always using a non-parametric test if they work in all situations and not worry about normality of the data?</p>
<p>Because, non-parametric tests have <strong>less power to detect small effects</strong>.</p>
<p>To check that, we can rely on simulations.<br>
Here, we simulate 1000 times two small datasets of 10 samples with a relatively small difference between their means (relative to the standard deviation) and perform both a <span class="math inline">\(t\)</span> test and a Wilcoxon rank sum test.</p>
<p>We then count how many times each test rejected the null (as they should) and see if one of the two tests reject more often.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>simulate_and_run_both_test <span class="ot">&lt;-</span> <span class="cf">function</span>(){</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>  x1 <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">10</span>, <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> <span class="dv">2</span>)</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>  x2 <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">10</span>, <span class="at">mean =</span> <span class="fl">1.5</span>, <span class="at">sd =</span> <span class="dv">2</span>)</span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>  ttest <span class="ot">&lt;-</span> <span class="fu">t.test</span>(x1, x2)</span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a>  wtest <span class="ot">&lt;-</span> <span class="fu">wilcox.test</span>(x1, x2)</span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tibble</span>(<span class="at">ttest_pvalue =</span> ttest<span class="sc">$</span>p.value, <span class="at">wtest_pvalue =</span> wtest<span class="sc">$</span>p.value)</span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb48-10"><a href="#cb48-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-11"><a href="#cb48-11" aria-hidden="true" tabindex="-1"></a>simulation_results <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">1000</span>, <span class="fu">simulate_and_run_both_test</span>())</span>
<span id="cb48-12"><a href="#cb48-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-13"><a href="#cb48-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-14"><a href="#cb48-14" aria-hidden="true" tabindex="-1"></a><span class="fu">apply</span>(simulation_results <span class="sc">&lt;=</span> <span class="fl">0.05</span>, <span class="at">MARGIN =</span> <span class="dv">2</span>, <span class="at">FUN =</span> mean)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>ttest_pvalue wtest_pvalue 
       0.354        0.319 </code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># apply is a function that allows to "apply" a given function (here the function "mean") to all rows or columns of a table.</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="co"># To chose if to apply to rows or to columns, we use 1 (rows) or 2 (columns) in the 2nd argument of the apply function. The first argument is the table we apply the function to.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We see that the <span class="math inline">\(t\)</span>-test more frequently detected that the two samples were drawn from distribution with different means/locations.</p>
</section>
</section>
<section id="multiple-testing" class="level1">
<h1>Multiple testing</h1>
<section id="simulations" class="level2">
<h2 class="anchored" data-anchor-id="simulations">Simulations</h2>
<p>As we discussed above, the <span class="math inline">\(p\)</span>-value of a test is the probability, under the null hypothesis, to observe a value of the test statistics as extreme as the one we observe. Usually, if that probability is less than 0.05 (we have less than one/20 chances to observe that value), we reject the null hypothesis.</p>
<p>So, if we repeat a test many many times on data generated under the null hypothesis (so we should <em>not</em> reject it), we will obtain a small <span class="math inline">\(p\)</span>-value a few times.</p>
<p>To verify this, let’s do a simulation experiment and repeat the following procedure a 1000 times: we draw two small datasets from that same population (= the null hypothesis is true) and perform a <span class="math inline">\(t\)</span>-test. In theory, we should get 0.05 x 1000 = 50 experiments with a <span class="math inline">\(p\)</span>-value smaller than 0.05.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>simulate_and_test <span class="ot">&lt;-</span> <span class="cf">function</span>(){</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>  x1 <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">10</span>)</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>  x2 <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">10</span>) <span class="co"># same mean</span></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>  ttest <span class="ot">&lt;-</span> <span class="fu">t.test</span>(x1, x2)</span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>  ttest<span class="sc">$</span>p.value <span class="co"># we return the p-value</span></span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a>null_p.values <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">1000</span>, <span class="fu">simulate_and_test</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Let’s count how many times we have a p-value smaller than 0.05 under the null hypothesis:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(null_p.values <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 52</code></pre>
</div>
</div>
<p>Which corresponds to a rate of</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(null_p.values <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.052</code></pre>
</div>
</div>
<p>which is, indeed, very close to the <span class="math inline">\(p\)</span>-value threshold that we’ve selected.</p>
<p>This is because, under the null hypothesis, the distribution of <span class="math inline">\(p\)</span>-values is uniform:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>null_p.values_tbl <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">p_values =</span> null_p.values)</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(null_p.values_tbl, <span class="fu">aes</span>(<span class="at">x =</span> p_values)) <span class="sc">+</span></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="module-6-hands-on_files/figure-html/unnamed-chunk-40-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>This is a good opportunity to learn how to perform a QQ-plot for another distribution than the Normal distribution. Here, to check that the p-values follow a uniform distribution, we can use the <code>distribution</code> argument from the <code>geom_qq</code> functions and do the following QQ-plot:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(null_p.values_tbl, <span class="fu">aes</span>(<span class="at">sample =</span> p_values)) <span class="sc">+</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_qq_line</span>(<span class="at">distribution =</span> qunif) <span class="sc">+</span></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_qq</span>(<span class="at">distribution =</span> qunif, <span class="at">size =</span> <span class="fl">0.5</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="module-6-hands-on_files/figure-html/unnamed-chunk-41-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>This is very convincing that the <span class="math inline">\(p\)</span>-values under the null follow a uniform distribution.</p>
<p>Now, let’s come back to the interpretation/consequences of this uniform distribution: if the distribution of <span class="math inline">\(p\)</span>-values under the Null is uniform, this means that we will reject the null hypothesis with a rate of <span class="math inline">\(\alpha\)</span> if <span class="math inline">\(\alpha\)</span> is the rejection threshold.</p>
<p>This will also apply to real data if we perform the same test several times.</p>
<p>For example, let’s say that we were interested in not just IL-1<span class="math inline">\(\beta\)</span> but all cytokines and are to repeat the <span class="math inline">\(t\)</span>-test we did above for all cytokines, since we have 10 of them, it’s not unlikely that we’d have small <span class="math inline">\(p\)</span>-values just by chance.</p>
<p>So, whenever we do “multiple testing”, we need to adjust for that multiple testing.</p>
<p>There are several ways to do “multiple testing adjustments” but the explanations of these methods are outside the scope of this class. Many of these methods have conveniently been implemented in the <code>p.adjust</code> function.</p>
<p>One of these methods that “controls for the”false discovery rate” is the Benjamini-Hochberg adjustment.</p>
<p>Let’s apply it to our simulated <span class="math inline">\(p\)</span>-values and check now how many samples are still considered to have a significant adjusted <span class="math inline">\(p\)</span>-value:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>null_p.values_tbl <span class="ot">&lt;-</span> </span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>  null_p.values_tbl <span class="sc">|&gt;</span> </span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">q_values =</span> <span class="fu">p.adjust</span>(p_values, <span class="at">method =</span> <span class="st">"BH"</span>))</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(null_p.values_tbl<span class="sc">$</span>q_values <span class="sc">&lt;=</span> <span class="fl">0.05</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0</code></pre>
</div>
</div>
<p>None of them!</p>
<p>Which is what it should be as we simulated our data under the null hypothesis, we should never reject the Null.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(null_p.values_tbl, <span class="fu">aes</span>(<span class="at">x =</span> q_values)) <span class="sc">+</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">20</span>) <span class="sc">+</span></span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expand_limits</span>(<span class="at">x =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"BH adjusted p-values"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="module-6-hands-on_files/figure-html/unnamed-chunk-43-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="cytokines" class="level2">
<h2 class="anchored" data-anchor-id="cytokines">Cytokines</h2>
<p>Let’s now test which cytokines have different means in BV and non-BV study participants.</p>
<p>First, we need to take the log-concentration of all cytokines and join with the clinical data via the “sample info”.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>elisa <span class="ot">&lt;-</span> </span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>  elisa <span class="sc">|&gt;</span> </span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">logconc =</span> <span class="fu">log10</span>(conc)) <span class="sc">|&gt;</span> </span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(sample_info, <span class="at">by =</span> <span class="fu">join_by</span>(sample_id)) <span class="sc">|&gt;</span> </span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(clin, <span class="at">by =</span> <span class="fu">join_by</span>(pid, time_point, arm))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We can also display concentration by BV status for each cytokine:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>elisa <span class="sc">|&gt;</span> </span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> logconc, <span class="at">fill =</span> BV)) <span class="sc">+</span></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">30</span>, <span class="at">position =</span> <span class="st">"identity"</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(cytokine <span class="sc">~</span> .) <span class="sc">+</span></span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">'log10(concentrations)'</span>) <span class="sc">+</span></span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> BV_colors)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="module-6-hands-on_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Another way to look at this is to display the data as “boxplot”:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>elisa <span class="sc">|&gt;</span> </span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y =</span> logconc, <span class="at">fill =</span> BV, <span class="at">col =</span> BV, <span class="at">x =</span> cytokine)) <span class="sc">+</span></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">varwidth =</span> <span class="cn">TRUE</span>, <span class="at">outlier.size =</span> <span class="fl">0.5</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> BV_colors) <span class="sc">+</span></span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> BV_colors)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="module-6-hands-on_files/figure-html/unnamed-chunk-46-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Looking at these visual displays of the data. Do you think it makes sense to do the test for all cytokines? What about cytokines that have a lot of samples with concentrations lower than the limit of detection?</p>
<p>Should we exclude these cytokines? Remove the out-of-range values? Or keep them?</p>
<p>Ideally, we would only want to do the test for variables where we are confident that we have representative samples.</p>
<p>There is a little bit of subjectivity in terms of what we deem representative, but, for now, let’s say that we want the 1st and 3rd quartile to be within the range of detection for each cytokine and each group.</p>
<p>Let’s thus compute the 1st and 3rd quartiles (or, equivalently the 25th and 75th percentiles) for each cytokines and each group:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>interquartiles <span class="ot">&lt;-</span> </span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>  elisa <span class="sc">|&gt;</span> </span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># we also need to compute the LLOQ and ULOQ for each cytokine</span></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cytokine) <span class="sc">|&gt;</span> </span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">LLOQ =</span> <span class="fu">min</span>(logconc), <span class="at">ULOQ =</span> <span class="fu">max</span>(logconc)) <span class="sc">|&gt;</span></span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cytokine, LLOQ, ULOQ, BV) <span class="sc">|&gt;</span> </span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(</span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">1st quartile</span><span class="st">`</span> <span class="ot">=</span> <span class="fu">quantile</span>(logconc, <span class="fl">0.25</span>),</span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">3rd quartile</span><span class="st">`</span> <span class="ot">=</span> <span class="fu">quantile</span>(logconc, <span class="fl">0.75</span>)</span>
<span id="cb64-10"><a href="#cb64-10" aria-hidden="true" tabindex="-1"></a>  ) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>`summarise()` has grouped output by 'cytokine', 'LLOQ', 'ULOQ'. You can
override using the `.groups` argument.</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>interquartiles <span class="sc">|&gt;</span> </span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 15%">
<col style="width: 13%">
<col style="width: 11%">
<col style="width: 13%">
<col style="width: 20%">
<col style="width: 20%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">cytokine</th>
<th style="text-align: center;">LLOQ</th>
<th style="text-align: center;">ULOQ</th>
<th style="text-align: center;">BV</th>
<th style="text-align: center;">1st quartile</th>
<th style="text-align: center;">3rd quartile</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">IFN-Y</td>
<td style="text-align: center;">-0.4641</td>
<td style="text-align: center;">1.369</td>
<td style="text-align: center;">BV</td>
<td style="text-align: center;">-0.4641</td>
<td style="text-align: center;">-0.1024</td>
</tr>
<tr class="even">
<td style="text-align: center;">IFN-Y</td>
<td style="text-align: center;">-0.4641</td>
<td style="text-align: center;">1.369</td>
<td style="text-align: center;">Healthy</td>
<td style="text-align: center;">-0.4641</td>
<td style="text-align: center;">-0.4641</td>
</tr>
<tr class="odd">
<td style="text-align: center;">IL-10</td>
<td style="text-align: center;">-0.1152</td>
<td style="text-align: center;">1.482</td>
<td style="text-align: center;">BV</td>
<td style="text-align: center;">-0.1152</td>
<td style="text-align: center;">0.3345</td>
</tr>
<tr class="even">
<td style="text-align: center;">IL-10</td>
<td style="text-align: center;">-0.1152</td>
<td style="text-align: center;">1.482</td>
<td style="text-align: center;">Healthy</td>
<td style="text-align: center;">-0.1152</td>
<td style="text-align: center;">-0.1152</td>
</tr>
<tr class="odd">
<td style="text-align: center;">IL-1a</td>
<td style="text-align: center;">0.143</td>
<td style="text-align: center;">3.011</td>
<td style="text-align: center;">BV</td>
<td style="text-align: center;">1.681</td>
<td style="text-align: center;">2.238</td>
</tr>
<tr class="even">
<td style="text-align: center;">IL-1a</td>
<td style="text-align: center;">0.143</td>
<td style="text-align: center;">3.011</td>
<td style="text-align: center;">Healthy</td>
<td style="text-align: center;">1.067</td>
<td style="text-align: center;">1.976</td>
</tr>
<tr class="odd">
<td style="text-align: center;">IL-1b</td>
<td style="text-align: center;">-0.6012</td>
<td style="text-align: center;">2.347</td>
<td style="text-align: center;">BV</td>
<td style="text-align: center;">0.8021</td>
<td style="text-align: center;">1.334</td>
</tr>
<tr class="even">
<td style="text-align: center;">IL-1b</td>
<td style="text-align: center;">-0.6012</td>
<td style="text-align: center;">2.347</td>
<td style="text-align: center;">Healthy</td>
<td style="text-align: center;">-0.2347</td>
<td style="text-align: center;">0.9864</td>
</tr>
<tr class="odd">
<td style="text-align: center;">IL-6</td>
<td style="text-align: center;">-1.029</td>
<td style="text-align: center;">2.238</td>
<td style="text-align: center;">BV</td>
<td style="text-align: center;">0.08762</td>
<td style="text-align: center;">0.7945</td>
</tr>
<tr class="even">
<td style="text-align: center;">IL-6</td>
<td style="text-align: center;">-1.029</td>
<td style="text-align: center;">2.238</td>
<td style="text-align: center;">Healthy</td>
<td style="text-align: center;">-0.364</td>
<td style="text-align: center;">0.6946</td>
</tr>
<tr class="odd">
<td style="text-align: center;">IL-8</td>
<td style="text-align: center;">-0.3279</td>
<td style="text-align: center;">3.398</td>
<td style="text-align: center;">BV</td>
<td style="text-align: center;">2.164</td>
<td style="text-align: center;">2.916</td>
</tr>
<tr class="even">
<td style="text-align: center;">IL-8</td>
<td style="text-align: center;">-0.3279</td>
<td style="text-align: center;">3.398</td>
<td style="text-align: center;">Healthy</td>
<td style="text-align: center;">1.575</td>
<td style="text-align: center;">2.71</td>
</tr>
<tr class="odd">
<td style="text-align: center;">IP-10</td>
<td style="text-align: center;">-0.5157</td>
<td style="text-align: center;">3.699</td>
<td style="text-align: center;">BV</td>
<td style="text-align: center;">-0.5157</td>
<td style="text-align: center;">0.06446</td>
</tr>
<tr class="even">
<td style="text-align: center;">IP-10</td>
<td style="text-align: center;">-0.5157</td>
<td style="text-align: center;">3.699</td>
<td style="text-align: center;">Healthy</td>
<td style="text-align: center;">-0.1079</td>
<td style="text-align: center;">1.945</td>
</tr>
<tr class="odd">
<td style="text-align: center;">MIG</td>
<td style="text-align: center;">0.1872</td>
<td style="text-align: center;">4.398</td>
<td style="text-align: center;">BV</td>
<td style="text-align: center;">0.1872</td>
<td style="text-align: center;">1.07</td>
</tr>
<tr class="even">
<td style="text-align: center;">MIG</td>
<td style="text-align: center;">0.1872</td>
<td style="text-align: center;">4.398</td>
<td style="text-align: center;">Healthy</td>
<td style="text-align: center;">0.6586</td>
<td style="text-align: center;">2.023</td>
</tr>
<tr class="odd">
<td style="text-align: center;">MIP-3a</td>
<td style="text-align: center;">0.7137</td>
<td style="text-align: center;">2.516</td>
<td style="text-align: center;">BV</td>
<td style="text-align: center;">0.7137</td>
<td style="text-align: center;">0.7137</td>
</tr>
<tr class="even">
<td style="text-align: center;">MIP-3a</td>
<td style="text-align: center;">0.7137</td>
<td style="text-align: center;">2.516</td>
<td style="text-align: center;">Healthy</td>
<td style="text-align: center;">0.7137</td>
<td style="text-align: center;">1.168</td>
</tr>
<tr class="odd">
<td style="text-align: center;">TNFa</td>
<td style="text-align: center;">-0.327</td>
<td style="text-align: center;">1.631</td>
<td style="text-align: center;">BV</td>
<td style="text-align: center;">-0.327</td>
<td style="text-align: center;">0.5051</td>
</tr>
<tr class="even">
<td style="text-align: center;">TNFa</td>
<td style="text-align: center;">-0.327</td>
<td style="text-align: center;">1.631</td>
<td style="text-align: center;">Healthy</td>
<td style="text-align: center;">-0.327</td>
<td style="text-align: center;">0.1446</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Let’s filter for cytokines with an interquartile range within the limits of quantification for both groups:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>interquartiles <span class="sc">|&gt;</span> </span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">1st quartile</span><span class="st">`</span> <span class="sc">&gt;</span> LLOQ,</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">3rd quartile</span><span class="st">`</span> <span class="sc">&lt;</span> ULOQ</span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cytokine) <span class="sc">|&gt;</span></span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">n =</span> <span class="fu">n</span>(), <span class="at">.groups =</span> <span class="st">"drop"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(n <span class="sc">==</span> <span class="dv">2</span>) <span class="sc">|&gt;</span></span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 15%">
<col style="width: 5%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">cytokine</th>
<th style="text-align: center;">n</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">IL-1a</td>
<td style="text-align: center;">2</td>
</tr>
<tr class="even">
<td style="text-align: center;">IL-1b</td>
<td style="text-align: center;">2</td>
</tr>
<tr class="odd">
<td style="text-align: center;">IL-6</td>
<td style="text-align: center;">2</td>
</tr>
<tr class="even">
<td style="text-align: center;">IL-8</td>
<td style="text-align: center;">2</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>That only leaves us with 4 cytokines. Our criteria might be a little too stringent, and, as long as you justify it properly, you could loosen these criteria. We just want to make sure that you always take a critical look at your data before running a test.</p>
<p>We will now filter our data to only keep these 4 cytokines and perform a statistical test for each of theses.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>ttest_res <span class="ot">&lt;-</span> </span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>  elisa <span class="sc">|&gt;</span> </span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cytokine) <span class="sc">|&gt;</span> </span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(</span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>    cytokine <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"IL-1a"</span>,<span class="st">"IL-1b"</span>, <span class="st">"IL-6"</span>, <span class="st">"IL-8"</span>)</span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span></span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(</span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">p-value</span><span class="st">`</span> <span class="ot">=</span> <span class="fu">t.test</span>(logconc <span class="sc">~</span> BV)<span class="sc">$</span>p.value,</span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">p &lt; 0.05</span><span class="st">`</span> <span class="ot">=</span> <span class="fu">ifelse</span>(<span class="st">`</span><span class="at">p-value</span><span class="st">`</span> <span class="sc">&lt;</span> <span class="fl">0.05</span>, <span class="st">"yes"</span>, <span class="st">""</span>)</span>
<span id="cb68-10"><a href="#cb68-10" aria-hidden="true" tabindex="-1"></a>  ) </span>
<span id="cb68-11"><a href="#cb68-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-12"><a href="#cb68-12" aria-hidden="true" tabindex="-1"></a>ttest_res <span class="sc">|&gt;</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 15%">
<col style="width: 16%">
<col style="width: 16%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">cytokine</th>
<th style="text-align: center;">p-value</th>
<th style="text-align: center;">p &lt; 0.05</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">IL-1a</td>
<td style="text-align: center;">8.492e-08</td>
<td style="text-align: center;">yes</td>
</tr>
<tr class="even">
<td style="text-align: center;">IL-1b</td>
<td style="text-align: center;">1.672e-09</td>
<td style="text-align: center;">yes</td>
</tr>
<tr class="odd">
<td style="text-align: center;">IL-6</td>
<td style="text-align: center;">0.08282</td>
<td style="text-align: center;"></td>
</tr>
<tr class="even">
<td style="text-align: center;">IL-8</td>
<td style="text-align: center;">0.00825</td>
<td style="text-align: center;">yes</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We see that 3/4 cytokines have p-values smaller than 0.05. But remember, we need to adjust for multiple testing.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>ttest_res <span class="ot">&lt;-</span> </span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>  ttest_res <span class="sc">|&gt;</span> </span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">adj. p-value</span><span class="st">`</span> <span class="ot">=</span> <span class="fu">p.adjust</span>(<span class="st">`</span><span class="at">p-value</span><span class="st">`</span>, <span class="at">method =</span> <span class="st">"BH"</span>),</span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">statistical</span><span class="sc">\n</span><span class="at">significance</span><span class="st">`</span> <span class="ot">=</span> </span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a>      <span class="fu">ifelse</span>(<span class="st">`</span><span class="at">adj. p-value</span><span class="st">`</span> <span class="sc">&lt;</span> <span class="fl">0.05</span>, <span class="st">"yes"</span>,<span class="st">""</span>)</span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-9"><a href="#cb69-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-10"><a href="#cb69-10" aria-hidden="true" tabindex="-1"></a>ttest_res <span class="sc">|&gt;</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 14%">
<col style="width: 15%">
<col style="width: 14%">
<col style="width: 19%">
<col style="width: 35%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">cytokine</th>
<th style="text-align: center;">p-value</th>
<th style="text-align: center;">p &lt; 0.05</th>
<th style="text-align: center;">adj. p-value</th>
<th style="text-align: center;">statistical significance</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">IL-1a</td>
<td style="text-align: center;">8.492e-08</td>
<td style="text-align: center;">yes</td>
<td style="text-align: center;">1.698e-07</td>
<td style="text-align: center;">yes</td>
</tr>
<tr class="even">
<td style="text-align: center;">IL-1b</td>
<td style="text-align: center;">1.672e-09</td>
<td style="text-align: center;">yes</td>
<td style="text-align: center;">6.689e-09</td>
<td style="text-align: center;">yes</td>
</tr>
<tr class="odd">
<td style="text-align: center;">IL-6</td>
<td style="text-align: center;">0.08282</td>
<td style="text-align: center;"></td>
<td style="text-align: center;">0.08282</td>
<td style="text-align: center;"></td>
</tr>
<tr class="even">
<td style="text-align: center;">IL-8</td>
<td style="text-align: center;">0.00825</td>
<td style="text-align: center;">yes</td>
<td style="text-align: center;">0.011</td>
<td style="text-align: center;">yes</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The three cytokines that had a p-value smaller than 0.05 are still significant after adjusting for multiple testing.</p>
<p>Redo this analysis, but with a less stringent criteria for representativeness. For example, let’s only request from our cytokines that their 1st and 3rd quartiles are different.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>selected_cytokines <span class="ot">&lt;-</span> </span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>  interquartiles <span class="sc">|&gt;</span> </span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(</span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">1st quartile</span><span class="st">`</span> <span class="sc">&lt;</span> <span class="st">`</span><span class="at">3rd quartile</span><span class="st">`</span></span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cytokine) <span class="sc">|&gt;</span></span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">n =</span> <span class="fu">n</span>(), <span class="at">.groups =</span> <span class="st">"drop"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb70-8"><a href="#cb70-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(n <span class="sc">==</span> <span class="dv">2</span>) <span class="sc">|&gt;</span></span>
<span id="cb70-9"><a href="#cb70-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(cytokine) <span class="sc">|&gt;</span> <span class="fu">unlist</span>()</span>
<span id="cb70-10"><a href="#cb70-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-11"><a href="#cb70-11" aria-hidden="true" tabindex="-1"></a>selected_cytokines</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>cytokine1 cytokine2 cytokine3 cytokine4 cytokine5 cytokine6 cytokine7 
  "IL-1a"   "IL-1b"    "IL-6"    "IL-8"   "IP-10"     "MIG"    "TNFa" </code></pre>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>ttest_res <span class="ot">&lt;-</span> </span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>  elisa <span class="sc">|&gt;</span> </span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cytokine <span class="sc">%in%</span> selected_cytokines) <span class="sc">|&gt;</span> </span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cytokine) <span class="sc">|&gt;</span></span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(</span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">p-value</span><span class="st">`</span> <span class="ot">=</span> <span class="fu">t.test</span>(logconc <span class="sc">~</span> BV)<span class="sc">$</span>p.value,</span>
<span id="cb72-7"><a href="#cb72-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">p &lt; 0.05</span><span class="st">`</span> <span class="ot">=</span> <span class="fu">ifelse</span>(<span class="st">`</span><span class="at">p-value</span><span class="st">`</span> <span class="sc">&lt;</span> <span class="fl">0.05</span>, <span class="st">"yes"</span>, <span class="st">""</span>)</span>
<span id="cb72-8"><a href="#cb72-8" aria-hidden="true" tabindex="-1"></a>  )  <span class="sc">|&gt;</span> </span>
<span id="cb72-9"><a href="#cb72-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb72-10"><a href="#cb72-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">adj. p-value</span><span class="st">`</span> <span class="ot">=</span> <span class="fu">p.adjust</span>(<span class="st">`</span><span class="at">p-value</span><span class="st">`</span>, <span class="at">method =</span> <span class="st">"BH"</span>),</span>
<span id="cb72-11"><a href="#cb72-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">statistical</span><span class="sc">\n</span><span class="at">significance</span><span class="st">`</span> <span class="ot">=</span> </span>
<span id="cb72-12"><a href="#cb72-12" aria-hidden="true" tabindex="-1"></a>      <span class="fu">ifelse</span>(<span class="st">`</span><span class="at">adj. p-value</span><span class="st">`</span> <span class="sc">&lt;</span> <span class="fl">0.05</span>, <span class="st">"yes"</span>,<span class="st">""</span>)</span>
<span id="cb72-13"><a href="#cb72-13" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb72-14"><a href="#cb72-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-15"><a href="#cb72-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-16"><a href="#cb72-16" aria-hidden="true" tabindex="-1"></a>ttest_res <span class="sc">|&gt;</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 14%">
<col style="width: 15%">
<col style="width: 14%">
<col style="width: 19%">
<col style="width: 35%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">cytokine</th>
<th style="text-align: center;">p-value</th>
<th style="text-align: center;">p &lt; 0.05</th>
<th style="text-align: center;">adj. p-value</th>
<th style="text-align: center;">statistical significance</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">IL-1a</td>
<td style="text-align: center;">8.492e-08</td>
<td style="text-align: center;">yes</td>
<td style="text-align: center;">1.486e-07</td>
<td style="text-align: center;">yes</td>
</tr>
<tr class="even">
<td style="text-align: center;">IL-1b</td>
<td style="text-align: center;">1.672e-09</td>
<td style="text-align: center;">yes</td>
<td style="text-align: center;">5.853e-09</td>
<td style="text-align: center;">yes</td>
</tr>
<tr class="odd">
<td style="text-align: center;">IL-6</td>
<td style="text-align: center;">0.08282</td>
<td style="text-align: center;"></td>
<td style="text-align: center;">0.08282</td>
<td style="text-align: center;"></td>
</tr>
<tr class="even">
<td style="text-align: center;">IL-8</td>
<td style="text-align: center;">0.00825</td>
<td style="text-align: center;">yes</td>
<td style="text-align: center;">0.01155</td>
<td style="text-align: center;">yes</td>
</tr>
<tr class="odd">
<td style="text-align: center;">IP-10</td>
<td style="text-align: center;">1.974e-12</td>
<td style="text-align: center;">yes</td>
<td style="text-align: center;">1.381e-11</td>
<td style="text-align: center;">yes</td>
</tr>
<tr class="even">
<td style="text-align: center;">MIG</td>
<td style="text-align: center;">4.902e-08</td>
<td style="text-align: center;">yes</td>
<td style="text-align: center;">1.144e-07</td>
<td style="text-align: center;">yes</td>
</tr>
<tr class="odd">
<td style="text-align: center;">TNFa</td>
<td style="text-align: center;">0.02701</td>
<td style="text-align: center;">yes</td>
<td style="text-align: center;">0.03151</td>
<td style="text-align: center;">yes</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>What do you conclude from this analysis?</p>
<p>As an exercise, you can also re-do the analyses using a non-parametric test and discuss the differences in the results.</p>
</section>
</section>
<section id="displaying-longitudinal-data" class="level1">
<h1>Displaying longitudinal data</h1>
<p>So far, we have ignore the fact that we have longitudinal data.</p>
<p>Remember the study design:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/study_design.png" class="img-fluid quarto-figure quarto-figure-center figure-img" alt="The study design" width="546"></p>
</figure>
</div>
<p>We have 3 time points for each patient. We can thus look at the evolution of the cytokine concentrations over time.</p>
<p>We have already joined the <code>elisa</code> data with the <code>sample_info</code> data so we already have the participant id and the visit id for all cytokine samples:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>elisa <span class="sc">|&gt;</span>  <span class="fu">head</span>() <span class="sc">|&gt;</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<caption>Table continues below</caption>
<colgroup>
<col style="width: 15%">
<col style="width: 13%">
<col style="width: 10%">
<col style="width: 20%">
<col style="width: 12%">
<col style="width: 11%">
<col style="width: 16%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">sample_id</th>
<th style="text-align: center;">cytokine</th>
<th style="text-align: center;">conc</th>
<th style="text-align: center;">limits</th>
<th style="text-align: center;">logconc</th>
<th style="text-align: center;">pid</th>
<th style="text-align: center;">time_point</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">SAMP094</td>
<td style="text-align: center;">IL-1a</td>
<td style="text-align: center;">173.9</td>
<td style="text-align: center;">within limits</td>
<td style="text-align: center;">2.24</td>
<td style="text-align: center;">pid_01</td>
<td style="text-align: center;">baseline</td>
</tr>
<tr class="even">
<td style="text-align: center;">SAMP094</td>
<td style="text-align: center;">IL-10</td>
<td style="text-align: center;">0.767</td>
<td style="text-align: center;">out of range</td>
<td style="text-align: center;">-0.1152</td>
<td style="text-align: center;">pid_01</td>
<td style="text-align: center;">baseline</td>
</tr>
<tr class="odd">
<td style="text-align: center;">SAMP094</td>
<td style="text-align: center;">IL-1b</td>
<td style="text-align: center;">5.39</td>
<td style="text-align: center;">within limits</td>
<td style="text-align: center;">0.7316</td>
<td style="text-align: center;">pid_01</td>
<td style="text-align: center;">baseline</td>
</tr>
<tr class="even">
<td style="text-align: center;">SAMP094</td>
<td style="text-align: center;">IL-8</td>
<td style="text-align: center;">48.29</td>
<td style="text-align: center;">within limits</td>
<td style="text-align: center;">1.684</td>
<td style="text-align: center;">pid_01</td>
<td style="text-align: center;">baseline</td>
</tr>
<tr class="odd">
<td style="text-align: center;">SAMP094</td>
<td style="text-align: center;">IL-6</td>
<td style="text-align: center;">5.07</td>
<td style="text-align: center;">within limits</td>
<td style="text-align: center;">0.705</td>
<td style="text-align: center;">pid_01</td>
<td style="text-align: center;">baseline</td>
</tr>
<tr class="even">
<td style="text-align: center;">SAMP094</td>
<td style="text-align: center;">TNFa</td>
<td style="text-align: center;">0.471</td>
<td style="text-align: center;">out of range</td>
<td style="text-align: center;">-0.327</td>
<td style="text-align: center;">pid_01</td>
<td style="text-align: center;">baseline</td>
</tr>
</tbody>
</table>
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 13%">
<col style="width: 20%">
<col style="width: 16%">
<col style="width: 8%">
<col style="width: 8%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">arm</th>
<th style="text-align: center;">nugent_score</th>
<th style="text-align: center;">crp_blood</th>
<th style="text-align: center;">ph</th>
<th style="text-align: center;">BV</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">8</td>
<td style="text-align: center;">0.44</td>
<td style="text-align: center;">5.7</td>
<td style="text-align: center;">BV</td>
</tr>
<tr class="even">
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">8</td>
<td style="text-align: center;">0.44</td>
<td style="text-align: center;">5.7</td>
<td style="text-align: center;">BV</td>
</tr>
<tr class="odd">
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">8</td>
<td style="text-align: center;">0.44</td>
<td style="text-align: center;">5.7</td>
<td style="text-align: center;">BV</td>
</tr>
<tr class="even">
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">8</td>
<td style="text-align: center;">0.44</td>
<td style="text-align: center;">5.7</td>
<td style="text-align: center;">BV</td>
</tr>
<tr class="odd">
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">8</td>
<td style="text-align: center;">0.44</td>
<td style="text-align: center;">5.7</td>
<td style="text-align: center;">BV</td>
</tr>
<tr class="even">
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">8</td>
<td style="text-align: center;">0.44</td>
<td style="text-align: center;">5.7</td>
<td style="text-align: center;">BV</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Before jumping into the visualization, let’s do some data wrangling.</p>
<p>First, let’s do some cosmetic changes to the <code>time_point</code> column so that the visits labels print nicely when using <code>ggplot</code>.</p>
<p>To replace the “_” with a space, we can use the <code>str_replace</code> function from the <code>stringr</code> package.</p>
<p>Before:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>elisa<span class="sc">$</span>time_point <span class="sc">|&gt;</span> <span class="fu">unique</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "baseline" "week_1"   "week_7"  </code></pre>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>elisa <span class="ot">&lt;-</span> </span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>  elisa <span class="sc">|&gt;</span> </span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">time_point =</span> time_point <span class="sc">|&gt;</span> <span class="fu">str_replace</span>(<span class="st">"_"</span>, <span class="st">" "</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>After:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co"># let's check that it worked</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>elisa<span class="sc">$</span>time_point <span class="sc">|&gt;</span> <span class="fu">unique</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "baseline" "week 1"   "week 7"  </code></pre>
</div>
</div>
<p>Then, let’s transform all the variables that are <code>character</code> but should be <code>factors</code> to … <code>factor</code>.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>elisa <span class="ot">&lt;-</span> </span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>  elisa <span class="sc">|&gt;</span> </span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">sample_id =</span> sample_id <span class="sc">|&gt;</span> <span class="fu">factor</span>(),</span>
<span id="cb79-5"><a href="#cb79-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">cytokine =</span> cytokine <span class="sc">|&gt;</span> <span class="fu">factor</span>(),</span>
<span id="cb79-6"><a href="#cb79-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">pid =</span> pid <span class="sc">|&gt;</span> <span class="fu">factor</span>(),</span>
<span id="cb79-7"><a href="#cb79-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">time_point =</span>time_point <span class="sc">|&gt;</span> <span class="fu">factor</span>(),</span>
<span id="cb79-8"><a href="#cb79-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">arm =</span> arm <span class="sc">|&gt;</span> <span class="fu">factor</span>(),</span>
<span id="cb79-9"><a href="#cb79-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">limits =</span> limits <span class="sc">|&gt;</span> <span class="fu">factor</span>()</span>
<span id="cb79-10"><a href="#cb79-10" aria-hidden="true" tabindex="-1"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Now, we can display the IL-1<span class="math inline">\(\beta\)</span> log-concentrations over-time, connecting the concentrations of a given participant by a line.</p>
<p>Note the use of the <code>group</code> aesthetic in <code>geom_line</code> to tell <code>ggplot</code> to connect the points of the same participant.</p>
<p>Check what happens if you remove <code>aes(group = pid)</code> from <code>geom_line</code>.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>arm_color <span class="ot">&lt;-</span> </span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">"placebo"</span> <span class="ot">=</span> <span class="st">"deeppink"</span>, <span class="st">"treatment"</span> <span class="ot">=</span> <span class="st">"steelblue1"</span>)</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>elisa <span class="sc">|&gt;</span> </span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cytokine <span class="sc">==</span> <span class="st">"IL-1b"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb80-6"><a href="#cb80-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> time_point, <span class="at">y =</span> logconc, <span class="at">color =</span> arm)) <span class="sc">+</span></span>
<span id="cb80-7"><a href="#cb80-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">group =</span> pid), <span class="at">alpha =</span> <span class="fl">0.75</span>) <span class="sc">+</span></span>
<span id="cb80-8"><a href="#cb80-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb80-9"><a href="#cb80-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Visits"</span>) <span class="sc">+</span></span>
<span id="cb80-10"><a href="#cb80-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"log10(IL-1b concentration)"</span>) <span class="sc">+</span></span>
<span id="cb80-11"><a href="#cb80-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="st">"Arm"</span>, <span class="at">values =</span> arm_color)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="module-6-hands-on_files/figure-html/unnamed-chunk-58-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<section id="paired-tests" class="level2">
<h2 class="anchored" data-anchor-id="paired-tests">Paired tests</h2>
<p>One question that this study may have aimed to answer is whether treatment altered how the cytokine concentrations changed over time.</p>
<p>Specifically, we may have been interested in comparing the changes between the baseline visit and the week 7 visit.</p>
<p>Let’s do the same visualization, but excluding week 1:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>elisa <span class="sc">|&gt;</span> </span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cytokine <span class="sc">==</span> <span class="st">"IL-1b"</span>, time_point <span class="sc">!=</span> <span class="st">"week 1"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> time_point, <span class="at">y =</span> logconc, <span class="at">color =</span> arm)) <span class="sc">+</span></span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">group =</span> pid), <span class="at">alpha =</span> <span class="fl">0.75</span>) <span class="sc">+</span></span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb81-6"><a href="#cb81-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Visits"</span>) <span class="sc">+</span></span>
<span id="cb81-7"><a href="#cb81-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"log10(IL-1b concentration)"</span>) <span class="sc">+</span></span>
<span id="cb81-8"><a href="#cb81-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="st">"Arm"</span>, <span class="at">values =</span> arm_color)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="module-6-hands-on_files/figure-html/unnamed-chunk-59-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We see that everyone starts with a different level of IL-1b and the ranking in the baseline visit is somewhat similar to the ranking in the week 7 visit, especially in the treatment arm.</p>
<p>To test this specifically, we can do a “paired test”.</p>
<p>It is “paired” because for each observation at the baseline visit, there is a “sister” observation at the week 7 visit (= the sample from the same participant).</p>
<p>Here, because we have many samples, we can use a paired <span class="math inline">\(t\)</span>-test, but if we had only one sample per participant, we would have to use a paired Wilcoxon test.</p>
<p>In a “real” analysis, we would only do one of the two tests, but for demonstration purposes, let’s do both (without adjusting for multiple testing because it’s just illustrative).</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>elisa <span class="sc">|&gt;</span> </span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cytokine <span class="sc">==</span> <span class="st">"IL-1b"</span>, time_point <span class="sc">!=</span> <span class="st">"week 1"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(cytokine, arm, pid, time_point, logconc) <span class="sc">|&gt;</span></span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> time_point, <span class="at">values_from =</span> logconc) <span class="sc">|&gt;</span></span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(arm) <span class="sc">|&gt;</span> </span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(</span>
<span id="cb82-7"><a href="#cb82-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">p (t-test)</span><span class="st">`</span> <span class="ot">=</span> </span>
<span id="cb82-8"><a href="#cb82-8" aria-hidden="true" tabindex="-1"></a>      <span class="fu">t.test</span>(<span class="at">x =</span> baseline, <span class="at">y =</span> <span class="st">`</span><span class="at">week 7</span><span class="st">`</span>, <span class="at">paired =</span> <span class="cn">TRUE</span>)<span class="sc">$</span>p.value,</span>
<span id="cb82-9"><a href="#cb82-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">p (signed rank)</span><span class="st">`</span> <span class="ot">=</span> </span>
<span id="cb82-10"><a href="#cb82-10" aria-hidden="true" tabindex="-1"></a>      <span class="fu">wilcox.test</span>(<span class="at">x =</span> baseline, <span class="at">y =</span> <span class="st">`</span><span class="at">week 7</span><span class="st">`</span>, <span class="at">paired =</span> <span class="cn">TRUE</span>)<span class="sc">$</span>p.value</span>
<span id="cb82-11"><a href="#cb82-11" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb82-12"><a href="#cb82-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb82-13"><a href="#cb82-13" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">p &lt; 0.05 (t)</span><span class="st">`</span> <span class="ot">=</span> <span class="fu">ifelse</span>(<span class="st">`</span><span class="at">p (t-test)</span><span class="st">`</span> <span class="sc">&lt;</span> <span class="fl">0.05</span>, <span class="st">"yes"</span>, <span class="st">""</span>),</span>
<span id="cb82-14"><a href="#cb82-14" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">p &lt; 0.05 (w)</span><span class="st">`</span> <span class="ot">=</span> <span class="fu">ifelse</span>(<span class="st">`</span><span class="at">p (signed rank)</span><span class="st">`</span> <span class="sc">&lt;</span> <span class="fl">0.05</span>, <span class="st">"yes"</span>, <span class="st">""</span>)</span>
<span id="cb82-15"><a href="#cb82-15" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb82-16"><a href="#cb82-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: There was 1 warning in `summarize()`.
ℹ In argument: `p (signed rank) = wilcox.test(x = baseline, y = `week 7`,
  paired = TRUE)$p.value`.
ℹ In group 2: `arm = treatment`.
Caused by warning in `wilcox.test.default()`:
! cannot compute exact p-value with ties</code></pre>
</div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 16%">
<col style="width: 17%">
<col style="width: 24%">
<col style="width: 20%">
<col style="width: 20%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">arm</th>
<th style="text-align: center;">p (t-test)</th>
<th style="text-align: center;">p (signed rank)</th>
<th style="text-align: center;">p &lt; 0.05 (t)</th>
<th style="text-align: center;">p &lt; 0.05 (w)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">placebo</td>
<td style="text-align: center;">0.3998</td>
<td style="text-align: center;">0.41</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
</tr>
<tr class="even">
<td style="text-align: center;">treatment</td>
<td style="text-align: center;">0.0007528</td>
<td style="text-align: center;">0.001226</td>
<td style="text-align: center;">yes</td>
<td style="text-align: center;">yes</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We see that the two tests agree in this case (as they should because the effects are quite large in the treatment arm).</p>
<p>Exercise: repeat the same visualization and analysis for IL-6 and IL-8. Do you need to adjust for multiple testing?</p>
</section>
</section>
<section id="compositional-data" class="level1">
<h1>Compositional data</h1>
<p>Multivariate data are said to be “compositional” when they are expressed in terms of <em>proportion</em> of a sample.</p>
<p>The problem with compositional data is that they are constrained to sum to 1, so that the values of the different variables are not independent.</p>
<p>Which means that if only one variable changes in absolute value (unknown), in relative values (known), the others will change too.</p>
<p>For example, if we have 3 cell sub-types: A, B, and C.</p>
<p>In one sample, we have 102 cells A, 72 cells B, and 147 cell C.</p>
<p>In code, we would write:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>sample_1 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"A"</span> <span class="ot">=</span> <span class="dv">102</span>, <span class="st">"B"</span> <span class="ot">=</span> <span class="dv">72</span>, <span class="st">"C"</span> <span class="ot">=</span> <span class="dv">147</span>)</span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>sample_1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>  A   B   C 
102  72 147 </code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="co"># note: usually we define vectors with just their values, like this: c(102, 72, 147), but we can also give names to each element of the vector as we did above</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>In another sample, we have the same number of cells A and B, but we have 62 additional cells C.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>sample_2 <span class="ot">&lt;-</span> sample_1 </span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>sample_2[<span class="st">"C"</span>] <span class="ot">&lt;-</span> sample_2[<span class="st">"C"</span>] <span class="sc">+</span> <span class="dv">62</span></span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-4"><a href="#cb87-4" aria-hidden="true" tabindex="-1"></a><span class="co"># the two lines above do the following: we create sample_2 as a copy of sample_1. Then, we only modify the "C" element of sample_2 and we add 62 to the value that was already there.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Now, we can compute the proportions of each cell type in each sample:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>prop_1 <span class="ot">&lt;-</span> sample_1 <span class="sc">/</span> <span class="fu">sum</span>(sample_1)</span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a>prop_2 <span class="ot">&lt;-</span> sample_2 <span class="sc">/</span> <span class="fu">sum</span>(sample_2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>And display them:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>prop_1 <span class="sc">|&gt;</span> <span class="fu">round</span>(<span class="at">digits =</span> <span class="dv">2</span>) <span class="co"># the `round` function rounds the number. The `digit` argument specifies the number of digits to keep after the decimal point. Try with digits = 0 or digits = 4 to see the difference in output.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>   A    B    C 
0.32 0.22 0.46 </code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>prop_2 <span class="sc">|&gt;</span> <span class="fu">round</span>(<span class="at">digits =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>   A    B    C 
0.27 0.19 0.55 </code></pre>
</div>
</div>
<p>We see that the proportion of C has increased (because there are more cells) and consequently, the proportions of A and B have decreased (even though there were the same number of cells).</p>
<section id="flow-cytometry-data" class="level2">
<h2 class="anchored" data-anchor-id="flow-cytometry-data">Flow cytometry data</h2>
<p>For now, let’s explore the flow cytometry data.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>flow <span class="ot">&lt;-</span> </span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">read_csv</span>(</span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"data/04_flow_cytometry_UKZN_workshop_2023.csv"</span></span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 132 Columns: 10
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (1): sample_id
dbl (9): live_cd19_negative, cd45_negative, cd45_positive, neutrophils, non_...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
<p>We have one row per sample. And remember, we have three sample per patient (one for each visit).</p>
<p>One of the first things we observe is that all columns (except <code>sample_id</code>) are (large) integer numbers. We say that we have “count data”.</p>
<p>Remember, that these are the number of cells observed in each “gating” categories. Also, remember that some cell categories are subset of other categories.</p>
<p>For example, “live CD19-” cells are roughly divided into “CD45+” and “CD45-” cells.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(flow, </span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">y =</span> live_cd19_negative, </span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a>           <span class="at">x =</span> cd45_positive <span class="sc">+</span> cd45_negative)</span>
<span id="cb95-4"><a href="#cb95-4" aria-hidden="true" tabindex="-1"></a>       ) <span class="sc">+</span></span>
<span id="cb95-5"><a href="#cb95-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> <span class="dv">0</span>, <span class="at">slope =</span> <span class="dv">1</span>, <span class="at">col =</span> <span class="st">"purple1"</span>) <span class="sc">+</span></span>
<span id="cb95-6"><a href="#cb95-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb95-7"><a href="#cb95-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>() <span class="sc">+</span></span>
<span id="cb95-8"><a href="#cb95-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="module-6-hands-on_files/figure-html/unnamed-chunk-66-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>And “CD3+” are roughly divided into “CD4 T” and “CD8 T” cells.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(flow, </span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> cd3_positive, </span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>           <span class="at">y =</span> cd4_t_cells <span class="sc">+</span> cd8_t_cells)</span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a>       ) <span class="sc">+</span></span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> <span class="dv">0</span>, <span class="at">slope =</span> <span class="dv">1</span>, <span class="at">col =</span> <span class="st">"purple1"</span>) <span class="sc">+</span></span>
<span id="cb96-6"><a href="#cb96-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb96-7"><a href="#cb96-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>() <span class="sc">+</span></span>
<span id="cb96-8"><a href="#cb96-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="module-6-hands-on_files/figure-html/unnamed-chunk-67-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>So, if we show the fraction of “CD4 T” cells against that of “CD8 T” cells among all T cells (the “CD3+” cells).</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(flow, </span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> cd4_t_cells<span class="sc">/</span>cd3_positive, </span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a>           <span class="at">y =</span> cd8_t_cells<span class="sc">/</span>cd3_positive )</span>
<span id="cb97-4"><a href="#cb97-4" aria-hidden="true" tabindex="-1"></a>       ) <span class="sc">+</span></span>
<span id="cb97-5"><a href="#cb97-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb97-6"><a href="#cb97-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"CD4 T cells (fraction of all T cells)"</span>) <span class="sc">+</span></span>
<span id="cb97-7"><a href="#cb97-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"CD8 T cells (fraction of all T cells)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="module-6-hands-on_files/figure-html/unnamed-chunk-68-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We see that we have a negative correlation: when the proportion of CD8 T cells increases, the proportion of CD4 T cells decreases.</p>
<p>What are the consequences of that?</p>
<p>Let’s say that we are interested in looking at the treatment effect on the proportion of different T cell sub-types. If we were to forget that these were compositional data, we could do the same as what we did for the cytokines, and perform a test for each cell sub-type.</p>
<p>But since we know that they are anti-correlated, if there is an effect on one sub-type, there might be an <em>opposite</em> effect on the other sub-type.</p>
<p>Let’s see if that’s the case.</p>
<p>First, let’s create a simplified flow cytometry table with just the T cell data and compute the proportion of the CD4 and CD8 T cells.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>flow_T <span class="ot">&lt;-</span> </span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>  flow <span class="sc">|&gt;</span> </span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(sample_id, cd4_t_cells, cd8_t_cells, cd3_positive) <span class="sc">|&gt;</span></span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb98-5"><a href="#cb98-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">cd4_t_cells_f =</span> cd4_t_cells<span class="sc">/</span>cd3_positive,</span>
<span id="cb98-6"><a href="#cb98-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">cd8_t_cells_f =</span> cd8_t_cells<span class="sc">/</span>cd3_positive,</span>
<span id="cb98-7"><a href="#cb98-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">sum =</span> cd4_t_cells_f <span class="sc">+</span> cd8_t_cells_f,</span>
<span id="cb98-8"><a href="#cb98-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">remaining_t_cells_f =</span> <span class="dv">1</span> <span class="sc">-</span> sum</span>
<span id="cb98-9"><a href="#cb98-9" aria-hidden="true" tabindex="-1"></a>    ) </span>
<span id="cb98-10"><a href="#cb98-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-11"><a href="#cb98-11" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(flow_T) <span class="sc">|&gt;</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<caption>Table continues below</caption>
<colgroup>
<col style="width: 16%">
<col style="width: 19%">
<col style="width: 19%">
<col style="width: 20%">
<col style="width: 22%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">sample_id</th>
<th style="text-align: center;">cd4_t_cells</th>
<th style="text-align: center;">cd8_t_cells</th>
<th style="text-align: center;">cd3_positive</th>
<th style="text-align: center;">cd4_t_cells_f</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">SAMP094</td>
<td style="text-align: center;">409</td>
<td style="text-align: center;">444</td>
<td style="text-align: center;">988</td>
<td style="text-align: center;">0.414</td>
</tr>
<tr class="even">
<td style="text-align: center;">SAMP052</td>
<td style="text-align: center;">7126</td>
<td style="text-align: center;">6009</td>
<td style="text-align: center;">15078</td>
<td style="text-align: center;">0.4726</td>
</tr>
<tr class="odd">
<td style="text-align: center;">SAMP017</td>
<td style="text-align: center;">36212</td>
<td style="text-align: center;">13434</td>
<td style="text-align: center;">53758</td>
<td style="text-align: center;">0.6736</td>
</tr>
<tr class="even">
<td style="text-align: center;">SAMP046</td>
<td style="text-align: center;">2872</td>
<td style="text-align: center;">2629</td>
<td style="text-align: center;">6254</td>
<td style="text-align: center;">0.4592</td>
</tr>
<tr class="odd">
<td style="text-align: center;">SAMP025</td>
<td style="text-align: center;">6384</td>
<td style="text-align: center;">3039</td>
<td style="text-align: center;">11355</td>
<td style="text-align: center;">0.5622</td>
</tr>
<tr class="even">
<td style="text-align: center;">SAMP050</td>
<td style="text-align: center;">2405</td>
<td style="text-align: center;">1310</td>
<td style="text-align: center;">4076</td>
<td style="text-align: center;">0.59</td>
</tr>
</tbody>
</table>
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 22%">
<col style="width: 12%">
<col style="width: 30%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">cd8_t_cells_f</th>
<th style="text-align: center;">sum</th>
<th style="text-align: center;">remaining_t_cells_f</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">0.4494</td>
<td style="text-align: center;">0.8634</td>
<td style="text-align: center;">0.1366</td>
</tr>
<tr class="even">
<td style="text-align: center;">0.3985</td>
<td style="text-align: center;">0.8711</td>
<td style="text-align: center;">0.1289</td>
</tr>
<tr class="odd">
<td style="text-align: center;">0.2499</td>
<td style="text-align: center;">0.9235</td>
<td style="text-align: center;">0.07649</td>
</tr>
<tr class="even">
<td style="text-align: center;">0.4204</td>
<td style="text-align: center;">0.8796</td>
<td style="text-align: center;">0.1204</td>
</tr>
<tr class="odd">
<td style="text-align: center;">0.2676</td>
<td style="text-align: center;">0.8299</td>
<td style="text-align: center;">0.1701</td>
</tr>
<tr class="even">
<td style="text-align: center;">0.3214</td>
<td style="text-align: center;">0.9114</td>
<td style="text-align: center;">0.08857</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We also computed the sum of the fractions of the CD4 and CD8 T cells and see that this sum is not always 1. That means that some cells are not CD4 or CD8 T cells. The distribution of the sum of their fraction goes as:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(flow_T, <span class="fu">aes</span>(<span class="at">x =</span> sum)) <span class="sc">+</span></span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb99-3"><a href="#cb99-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">1</span>, <span class="at">col =</span> <span class="st">"red"</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb99-4"><a href="#cb99-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"sum of the fractions of CD4 and CD8 T cells"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="module-6-hands-on_files/figure-html/unnamed-chunk-70-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>As said above, one question we may have is whether the proportions of CD4 and CD8 T cells are different between the two arms at the end of the intervention.</p>
<p>To be able to answer this question, we first need to join the flow cytometry data with the sample information (the <code>sample_info</code> table) so we know which sample is from which visit and which participant and their arm.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>flow_T_with_info <span class="ot">&lt;-</span> </span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>  flow_T <span class="sc">|&gt;</span> </span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(sample_info, <span class="at">by =</span> <span class="st">"sample_id"</span>)</span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-5"><a href="#cb100-5" aria-hidden="true" tabindex="-1"></a>flow_T_with_info <span class="sc">|&gt;</span> <span class="fu">head</span>() <span class="sc">|&gt;</span> <span class="fu">pander</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<caption>Table continues below</caption>
<colgroup>
<col style="width: 16%">
<col style="width: 19%">
<col style="width: 19%">
<col style="width: 20%">
<col style="width: 22%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">sample_id</th>
<th style="text-align: center;">cd4_t_cells</th>
<th style="text-align: center;">cd8_t_cells</th>
<th style="text-align: center;">cd3_positive</th>
<th style="text-align: center;">cd4_t_cells_f</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">SAMP094</td>
<td style="text-align: center;">409</td>
<td style="text-align: center;">444</td>
<td style="text-align: center;">988</td>
<td style="text-align: center;">0.414</td>
</tr>
<tr class="even">
<td style="text-align: center;">SAMP052</td>
<td style="text-align: center;">7126</td>
<td style="text-align: center;">6009</td>
<td style="text-align: center;">15078</td>
<td style="text-align: center;">0.4726</td>
</tr>
<tr class="odd">
<td style="text-align: center;">SAMP017</td>
<td style="text-align: center;">36212</td>
<td style="text-align: center;">13434</td>
<td style="text-align: center;">53758</td>
<td style="text-align: center;">0.6736</td>
</tr>
<tr class="even">
<td style="text-align: center;">SAMP046</td>
<td style="text-align: center;">2872</td>
<td style="text-align: center;">2629</td>
<td style="text-align: center;">6254</td>
<td style="text-align: center;">0.4592</td>
</tr>
<tr class="odd">
<td style="text-align: center;">SAMP025</td>
<td style="text-align: center;">6384</td>
<td style="text-align: center;">3039</td>
<td style="text-align: center;">11355</td>
<td style="text-align: center;">0.5622</td>
</tr>
<tr class="even">
<td style="text-align: center;">SAMP050</td>
<td style="text-align: center;">2405</td>
<td style="text-align: center;">1310</td>
<td style="text-align: center;">4076</td>
<td style="text-align: center;">0.59</td>
</tr>
</tbody>
</table>
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 20%">
<col style="width: 11%">
<col style="width: 27%">
<col style="width: 11%">
<col style="width: 16%">
<col style="width: 12%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">cd8_t_cells_f</th>
<th style="text-align: center;">sum</th>
<th style="text-align: center;">remaining_t_cells_f</th>
<th style="text-align: center;">pid</th>
<th style="text-align: center;">time_point</th>
<th style="text-align: center;">arm</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">0.4494</td>
<td style="text-align: center;">0.8634</td>
<td style="text-align: center;">0.1366</td>
<td style="text-align: center;">pid_01</td>
<td style="text-align: center;">baseline</td>
<td style="text-align: center;">placebo</td>
</tr>
<tr class="even">
<td style="text-align: center;">0.3985</td>
<td style="text-align: center;">0.8711</td>
<td style="text-align: center;">0.1289</td>
<td style="text-align: center;">pid_01</td>
<td style="text-align: center;">week_1</td>
<td style="text-align: center;">placebo</td>
</tr>
<tr class="odd">
<td style="text-align: center;">0.2499</td>
<td style="text-align: center;">0.9235</td>
<td style="text-align: center;">0.07649</td>
<td style="text-align: center;">pid_01</td>
<td style="text-align: center;">week_7</td>
<td style="text-align: center;">placebo</td>
</tr>
<tr class="even">
<td style="text-align: center;">0.4204</td>
<td style="text-align: center;">0.8796</td>
<td style="text-align: center;">0.1204</td>
<td style="text-align: center;">pid_02</td>
<td style="text-align: center;">baseline</td>
<td style="text-align: center;">placebo</td>
</tr>
<tr class="odd">
<td style="text-align: center;">0.2676</td>
<td style="text-align: center;">0.8299</td>
<td style="text-align: center;">0.1701</td>
<td style="text-align: center;">pid_02</td>
<td style="text-align: center;">week_1</td>
<td style="text-align: center;">placebo</td>
</tr>
<tr class="even">
<td style="text-align: center;">0.3214</td>
<td style="text-align: center;">0.9114</td>
<td style="text-align: center;">0.08857</td>
<td style="text-align: center;">pid_02</td>
<td style="text-align: center;">week_7</td>
<td style="text-align: center;">placebo</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Let’s now display the proportions of these cells at the week 7 visit:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>flow_T_with_info_long <span class="ot">&lt;-</span> </span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a>  flow_T_with_info <span class="sc">|&gt;</span> </span>
<span id="cb101-3"><a href="#cb101-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(sample_id, pid, time_point, arm, <span class="fu">ends_with</span>(<span class="st">"_f"</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb101-4"><a href="#cb101-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb101-5"><a href="#cb101-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">cols =</span> <span class="fu">ends_with</span>(<span class="st">"_f"</span>),</span>
<span id="cb101-6"><a href="#cb101-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">"cell_type"</span>,</span>
<span id="cb101-7"><a href="#cb101-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_to =</span> <span class="st">"fraction of T cells"</span></span>
<span id="cb101-8"><a href="#cb101-8" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb101-9"><a href="#cb101-9" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb101-10"><a href="#cb101-10" aria-hidden="true" tabindex="-1"></a>flow_T_with_info_long <span class="sc">|&gt;</span> </span>
<span id="cb101-11"><a href="#cb101-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> cell_type, <span class="at">y =</span> <span class="st">`</span><span class="at">fraction of T cells</span><span class="st">`</span>,</span>
<span id="cb101-12"><a href="#cb101-12" aria-hidden="true" tabindex="-1"></a>             <span class="at">fill =</span> arm)) <span class="sc">+</span></span>
<span id="cb101-13"><a href="#cb101-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb101-14"><a href="#cb101-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> arm_color) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="module-6-hands-on_files/figure-html/unnamed-chunk-72-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>It looks like there might be some differences: a lower proportion of CD4 T cells in the intervention arm than in the placebo arm and and slightly higher proportions in the two others.</p>
<p>But from these compositional data, it is impossible to know if one of the cell type is driving these effects or if it is a combination of several changes in each cell types.</p>
<p>One could still do a test to see if the proportions are different in the two arms, but this is outside the scope of this workshop. Come back to the next ones for more :)</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/elsherbini\.github\.io\/durban-data-science-for-biology\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>© 2023. Huge thanks to <a href="https://github.com/andrewpbray">Andrew P. Bray</a> and <a href="https://github.com/clauswilke">Claus O. Wilke</a> for open source code and sharing content.</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>This page is built with <i class="fa-solid fa-heart" aria-label="heart"></i> and <a href="https://quarto.org/">Quarto</a>.</p>
</div>
  </div>
</footer>




</body></html>